---
title: "G816 PMT RNAseq analysis"
output: html_notebook
---


```{r import packages, message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(cowplot)
library(plotly)
library(readxl)
library(GSVA)
library(GSVAdata)
library(GSEABase)
library(limma)
library(msigdbr)
library(gage)
library(GO.db)
library(ggrepel)
library(ggsci)
library(paletteer)
library(parallel)
library(RColorBrewer)
library(ggrepel)
library(pheatmap)
library(umap)
library(fgsea)
library(pls)
library(matrixStats)
```

```{r Clear environment}
### Clear R workspace:
rm(list=ls())
```

```{r Check and set working directory}
### Check current working directory:
cwd <- getwd()
cwd
### Set working directory:
setwd(cwd) # Set current working directory to location of this notebook for easier handling and saving of data associated with the notebook
```
```{r Custom ggplot theme for plot background}
cust.theme <- function(ticks, axis, box, lnsz){
  if (missing(ticks)){
    ticks <- TRUE
  }
  if (missing(axis)){
    axis <- "y"
  }
  if (missing(box)){
    box <- TRUE
  }
  if (missing(lnsz)){
    lnsz <- 1
  }
  if (ticks == FALSE & axis == "y" ){
    theme_out <- theme(
      axis.text.y = element_blank(), axis.ticks.y = element_blank(), # Make y axis ticks blank
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = "black", fill = "white"),
      )
  }
  else if (ticks == FALSE & axis != "y"){
    theme_out <- theme(
      axis.text.x = element_blank(), axis.ticks.x = element_blank(), # Make x axis ticks blank
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = "black", fill = "white"),
      )
  }
  else {
     theme_out <- theme(
       axis.line.y = element_blank(),
       axis.line.x = element_blank(),
       panel.background = element_rect(size = lnsz, color = "black", fill = "white"),
       ) 
  }
  if (box != TRUE) {
    theme_out <- theme_out + theme(panel.background = element_blank(),
                                   axis.line.y = element_line(size = lnsz, color = "black"),
                                   axis.line.x = element_line(size = lnsz, color = "black")
                                   )
  } else {theme_out <- theme_out}
}
```

```{r import data set}
counts.import <- read_excel("WH3202_RNASeq_OverView.xlsx", sheet = "All.countsMatrix")
samples.import <- read_excel("WH3202_RNASeq_OverView.xlsx", sheet = "SeqReadQC", skip = 2)
```
##data clean-up
```{r cleaning up and filtering data}
counts.rna <- counts.import
samples.rna <- samples.import %>% 
  dplyr::slice(19:27)
samples <- c("Unt", "Unt", "Unt", "TT", "TT", "TT", "TMZ", "TMZ", "TMZ")

colnames(counts.rna)[1] <- "GENE"
colnames(counts.rna) <- gsub(".*_", "", colnames(counts.rna)) #removing extra stuff before sample numbers
samples.rna$`Sample Names` <- gsub(".*_", "", samples.rna$`Sample Names`) #removing extra stuff before sample numbers

counts.rna <- dplyr::select(counts.rna, GENE, samples.rna$`Sample Names`) %>% # filtering matrix ot only have the samples we care about - G816 samples
  distinct(GENE, .keep_all = T) %>% # get rid of duplicate genes 
  column_to_rownames(var="GENE")
colnames(counts.rna) <- samples

library(edgeR) # Contains functions for filtering and normalizing count data (from BioConductor)
#assign samples to groups: Unt, TT, TMZ
groups <- data.frame(
  samples = colnames(counts.rna),
  group = colnames(counts.rna) %>% factor(levels =c("Unt","TT", "TMZ"))
  )

keep.exprs <- filterByExpr(counts.rna, group = groups$group) # Identify lowly expressed genes 
counts.filt <- counts.rna[keep.exprs,, ] %>%  # Filter out lowly expressed genes
  data.frame()

#get ensembl IDs for our genes
genes <- rownames(counts.rna) %>% 
  mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL', multiVals = "first") %>% 
  data.frame() %>%
  rownames_to_column(var = "gene")

colnames(genes)[2] <- "ensembl"
genes <- genes %>% filter(!is.na(ensembl))

counts.filt %<>% filter(rownames(.) %in% genes$gene) # remove genes with no ensembl ID

# getting counts matrices with either gene names or ensembl IDs
counts.filt <- inner_join( 
  genes,
  counts.filt %>% rownames_to_column(var = "gene"),
  by = "gene"
)
counts.symbol <- counts.filt %>% dplyr::select(-ensembl) %>% column_to_rownames(var="gene")
counts.ensembl <- counts.filt %>% dplyr::select(-gene)%>% column_to_rownames(var="ensembl")


print(paste0("Number of unique genes with >0 expression: ", nrow(counts.rna[rowSums(counts.rna[])>0,])))
```

```{r Calculate normalized count data}

design <- data.frame(Unt = c(1, 1, 1, 0, 0, 0, 0, 0, 0),
                     TT = c(0,0,0,1,1,1,0,0,0),
                     TMZ = c(0,0,0,0,0,0,1,1,1))

voom(counts.ensembl, design, plot=T)
### Calculate log-scale CPM:
logcpm <- counts.ensembl
logcpm %<>% cpm(log=T)
colnames(logcpm) <- samples

## Transpose data (for upstream analyses) and assign column names using gene symbols:
data3 <- logcpm %>% t()



### Check if normalization of gene expression distributions is needed:
boxplot(data3 %>% t(), main = "Check gene expression distributions", ylab = "log-CPM")
```
##load gene sets
```{r Load packages and collections, message=F, warning=F}
## Load packages:
library(clusterProfiler)
library(org.Hs.eg.db)
library(msigdbr)

### Define MSigDB gene set collection(s) to use --> retrieve with 'msigdbr' package:
species = "Homo sapiens"

## Retrieve Hallmark and canonical pathways collections in the database:
hall = msigdbr(species = species, category = "H")
cp = msigdbr(species = species, category = "C2", subcategory = "CP")
cp.b = msigdbr(species = species, category = "C2", subcategory = "CP:BIOCARTA")
cp.r = msigdbr(species = species, category = "C2", subcategory = "CP:REACTOME")
cp.p = msigdbr(species = species, category = "C2", subcategory = "CP:PID")
cp.k = msigdbr(species = species, category = "C2", subcategory = "CP:KEGG")
cp.w = msigdbr(species = species, category = "C2", subcategory = "CP:WIKIPATHWAYS")
c6 = msigdbr(species = species, category = "C6")
gene_sets1 <- rbind(c6, hall, cp, cp.b, cp.r, cp.p, cp.k) %>% split(x = .$entrez_gene, f = .$gs_name)

## Go collections:
go.bp <- msigdbr(species = species, category = "C5", subcategory = "GO:BP")
go.cc <- msigdbr(species = species, category = "C5", subcategory = "GO:CC")
go.mf <- msigdbr(species = species, category = "C5", subcategory = "GO:MF")
gene_sets2 <- rbind(go.mf, go.cc, go.bp) %>% split(x = .$entrez_gene, f = .$gs_name)

## Transcription factor target collections:
go.tft <- msigdbr(species = species, category = "C3", subcategory = "TFT:GTRD")
go.tft_leg <- msigdbr(species = species, category = "C3", subcategory = "TFT:TFT_Legacy")
gene_sets3 <- rbind(go.tft, go.tft_leg) %>% split(x = .$entrez_gene, f = .$gs_name)

cgp.sig <- msigdbr(species = species, category = "C2", subcategory = "CGP") %>% 
  split(x = .$entrez_gene, f = .$gs_name)

phillips_mes <- read_excel("mmc4.xls", sheet = "MES") %>% distinct(`Gene Symbol`) %>% pull(`Gene Symbol`) %>% match(genes$gene) %>% genes$ensembl[.] %>% na.omit() %>% as.character()

phillips_pn <- read_excel("mmc4.xls", sheet = "PN") %>% distinct(`Gene Symbol`) %>% pull(`Gene Symbol`) %>%
  match(genes$gene) %>% genes$ensembl[.] %>% na.omit() %>% as.character()

phillips_mes_sig <- read_excel("mmc4.xls", sheet = "MES") %>% filter(`Signature gene` == "x") %>% distinct(`Gene Symbol`) %>% pull(`Gene Symbol`) 

phillips_pn_sig <- read_excel("mmc4.xls", sheet = "PN") %>% filter(`Signature gene` == "x") %>% distinct(`Gene Symbol`) %>% pull(`Gene Symbol`) 

segerman_mes <- c("DKK3", "SLC25A43", "FSTL1", "TBX15", "TRIM16", "ZHX3", "BLVRB", "PDCD1LG2", "CLMP", "NWD1", "TMBIM1", "IGDCC4", "LAMA5", "ABCC3", "DAB2", "CSF1", "SLC46A3", "OSMR", "STAMBPL1", "COPZ2", "FAP", "SLC9A9", "FN1", "MRO", "SLC2A10", "MVP", "HTATIP2", "CYP27A1", "SCPEP1", "PPIC", "TPCN1", "AHNAK", "PLBD2", "STOM", "RNLS", "SMPD1", "EMILIN1", "OSTF1", "PER1", "EPHX1", "EDIL3", "QSOX1", "RBMS3", "LOXL1", "C1S", "SHC1", "SERPINB1", "FUCA1", "MSRB3", "MT1G") %>% match(genes$gene) %>% genes$ensembl[.] %>% as.numeric() %>% na.omit()

pcEMT <- read_csv("pcEMT_sig.csv")
pcEMT_m <- pcEMT %>% filter(subtype == "M") %>% pull(var = "gene") %>% 
  match(genes$gene) %>% genes$ensembl[.] %>% na.omit() %>% as.character()
pcEMT_e <- pcEMT %>% filter(subtype == "E") %>% pull(var = "gene") %>% 
  match(genes$gene) %>% genes$ensembl[.] %>% na.omit() %>% as.character()

gene_sets_all <- append(gene_sets1, gene_sets2)# %>% append(gene_sets3)
#gene_sets_all <- gene_sets1

gene_sets5 <- read_tsv("IDHwt.GBM.MetaModules.tsv")
gene_sets5.mes1 <- pull(gene_sets5, MESlike1)
gene_sets5.mes2 <- pull(gene_sets5, MESlike2)
gene_sets5.mes <- append(gene_sets5.mes1, gene_sets5.mes2)

gene_sets5.pn <- gene_sets5 %>% pull(OPClike)
pmt.sig2 <- append(gene_sets5.mes, gene_sets5.pn)
pmt.sig2 <- pmt.sig2 %>% data.frame()
colnames(pmt.sig2)[1] <- "gene"

pmt.sig2 <- mutate(pmt.sig2, ID = pmt.sig2$gene %>% match(genes$gene) %>% genes$ensembl[.])

```
##overall gene set scores
```{r}
exp <- logcpm
exp.vars <- rowVars(exp)
exp <- exp %>% data.frame() %>% add_column(var = exp.vars) %>%
  filter(var > 0.5) %>% dplyr::select(-var) %>% data.matrix()
#exp %<>% t() %>% scale(center = T, scale = T) %>% t()
exp <- exp %>% data.frame() %>% rownames_to_column(var = "ensemblID")
gene_sets6 <- getGmt("msigdb.v7.0.entrez.gmt")
gene.ids <- geneIds(gene_sets6)

verhaak.mes <- gene.ids$VERHAAK_GLIOBLASTOMA_MESENCHYMAL

verhaak.mes.sc <- match(verhaak.mes, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(verhaak.mes.sc) <- NULL
verhaak.mes.sc %<>% column_to_rownames(var = "ensemblID")
gene_set_scores <- verhaak.mes.sc %>% colMeans() %>% data.frame(verhaak.mes=.)

verhaak.pn <- gene.ids$VERHAAK_GLIOBLASTOMA_PRONEURAL

verhaak.pn.sc <- match(verhaak.pn, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(verhaak.pn.sc) <- NULL
verhaak.pn.sc %<>% column_to_rownames(var = "ensemblID")
gene_set_scores <- gene_set_scores %>% mutate(
  verhaak.pn = colMeans(verhaak.pn.sc)
)

verhaak.cl <- gene.ids$VERHAAK_GLIOBLASTOMA_CLASSICAL

verhaak.cl.sc <- match(verhaak.cl, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(verhaak.cl.sc) <- NULL
verhaak.cl.sc %<>% column_to_rownames(var = "ensemblID")
gene_set_scores <- gene_set_scores %>% mutate(
  verhaak.cl = colMeans(verhaak.cl.sc)
)

hall.emt <- gene.ids$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION

hall.emt.sc <- match(hall.emt, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(hall.emt.sc) <- NULL
hall.emt.sc %<>% column_to_rownames(var = "ensemblID")
gene_set_scores <- gene_set_scores %>% mutate(
  hallmark.emt = colMeans(hall.emt.sc)
)

go.emt <- gene.ids$GO_EPITHELIAL_TO_MESENCHYMAL_TRANSITION

go.emt.sc <- match(go.emt, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(go.emt.sc) <- NULL
go.emt.sc %<>% column_to_rownames(var = "ensemblID")
gene_set_scores <- gene_set_scores %>% mutate(
  go.emt = colMeans(go.emt.sc)
)

neftel.mes1 <- gene_sets5.mes1 %>% match(genes$gene) %>% na.omit() %>% genes$ensembl[.] %>% as.numeric()

neftel.mes1.sc <- match(neftel.mes1, exp$ensemblID) %>% na.omit() %>% unique() %>% exp[.,] %>% data.frame()
rownames(neftel.mes1.sc) <- NULL
neftel.mes1.sc %<>% column_to_rownames(var = "ensemblID")
gene_set_scores <- gene_set_scores %>% mutate(
  neftel.mes1 = colMeans(neftel.mes1.sc)
)

neftel.mes2 <- gene_sets5.mes2 %>% match(genes$gene) %>% na.omit() %>% genes$ensembl[.] %>% as.numeric()

neftel.mes2.sc <- match(neftel.mes2, exp$ensemblID) %>% na.omit() %>% unique() %>% exp[.,] %>% data.frame()
rownames(neftel.mes2.sc) <- NULL
neftel.mes2.sc %<>% column_to_rownames(var = "ensemblID")
gene_set_scores <- gene_set_scores %>% mutate(
  neftel.mes2 = colMeans(neftel.mes2.sc)
)

neftel.opc <- gene_sets5.pn %>% match(genes$gene) %>% 
  na.omit() %>% genes$ensembl[.] %>% as.numeric()

neftel.opc.sc <- match(neftel.opc, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(neftel.opc.sc) <- NULL
neftel.opc.sc %<>% column_to_rownames(var = "ensemblID")
gene_set_scores <- gene_set_scores %>% mutate(
  neftel.opc = colMeans(neftel.opc.sc)
)

##plotting


#gene_set_scores %<>% scale(center = T, scale = T)

#(gene_set_scores)

gene_sets_all <- gene_sets_all %>% append(gene_sets6)
```

## DEA for Unt vs. TGFb+TNFa samples
```{r}
## Statistical thresholds:
adjPvalueCutoff1 <- 0.01 # Threshold for adjusted p-values (FDR) for GENE DEA
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)

## Gene expression data:
data_for_lm <- data.frame(logcpm)
colnames(data_for_lm) <- c("Unt", "Unt", "Unt", "TT", "TT", "TT", "TMZ", "TMZ", "TMZ")
#colnames(data_for_lm) <- data3 %>% rownames() # Col names = samples
#rownames(data_for_lm) <- colnames(data3) #%>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') # Row names = Entrez gene IDs

### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- model.matrix(~0 + groups$group)[1:9,1:3]
colnames(design1) <- c("Unt","TT", "TMZ")
fit1 <- lmFit(data_for_lm, design1) 

compare_group <- "TT" #change to TT or TMZ to compare with the Unt group
if(compare_group == "TT"){
  contr1 <- makeContrasts(TTvsUnt = TT - Unt, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
} else if(compare_group == "TMZ"){
  contr1 <- makeContrasts(TMZvsUnt = TMZ - Unt, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
}

## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes

###################################### Volcano plot --> GENES ######################################
txtsz <- 14
linesz <- 1
## Save list of ordered, differentially expressed genes:
allGenes <- mutate(allGenes, gene = allGenes$ID %>% match(genes$ensembl) %>% genes$gene[.])
DEAgenes.TT <- allGenes
#DEAgenes.TT$logFC %>% order() %>% DEAgenes.TT[.,] %>% write.csv("ordered_DEgenes_TGF+TNF.csv")


##PLOTTING
ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1 %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_point(data = function(x){subset(x, ID %in% "TT")},
    color = "black",
    shape = 15,
    # cex = 2.5
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
   geom_text_repel(
     data = function(x){subset(x, (adj.P.Val < adjPvalueCutoff1) & (abs(logFC) > 4))},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 15, max.iter = 100000
   ) +
  geom_text_repel(data = function(x){subset(x, ID == "TT")}) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TGF\u03B2+TNF\u03B1 vs. Untreated", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )
#ggsave("plots/DEA-genes_TGF+TNF.png", width = 6, height = 6)


#just verhaak genes
ggplot(subset(allGenes, ID %in% c(verhaak.mes, verhaak.pn)),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(data = function(x){subset(x, ID %in% verhaak.mes)},
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_point(data = function(x){subset(x, ID %in% verhaak.pn)},
    color = "darkred",
    shape = "triangle",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     data = function(x){subset(x, ID %in% verhaak.mes & abs(logFC) > logFCcutoff1)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
     geom_text_repel(
     data = function(x){subset(x, ID %in% verhaak.pn & abs(logFC) > logFCcutoff1)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkred"
   ) +
  geom_text(aes(x = -3, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TGF\u03B2+TNF\u03B1 vs. Untreated", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/G816_DEA_TGFb+TNFa_Verhaak.svg", width = 6, height = 6)
#ggsave("plots/DEA-genes_TGF+TNF_verhaak.png", width = 6, height = 6)


## printing stats for up- and down-regulated transcripts
diff_exp_lig.m <- allGenes %>% filter(ID %in% verhaak.mes)
paste("Mesenchymal Up:" , diff_exp_lig.m %>% 
        filter(adj.P.Val < adjPvalueCutoff1) %>% 
        filter(logFC > logFCcutoff1) %>% 
        nrow(), 
      "/",
      nrow(diff_exp_lig.m)
      )
paste("Mesenchymal Down:" , diff_exp_lig.m %>% 
        filter(adj.P.Val < adjPvalueCutoff1) %>% 
        filter(logFC < -logFCcutoff1) %>% 
        nrow(), 
      "/",
      nrow(diff_exp_lig.m)
      )

diff_exp_lig.p <- allGenes %>% filter(ID %in% verhaak.pn)
paste("Proneural Up:" , diff_exp_lig.p %>% 
        filter(adj.P.Val < adjPvalueCutoff1) %>% 
        filter(logFC > logFCcutoff1) %>% 
        nrow(), 
      "/",
      nrow(diff_exp_lig.p)
      )
paste("Proneural Down:" , diff_exp_lig.p %>% 
        filter(adj.P.Val < adjPvalueCutoff1) %>% 
        filter(logFC < -logFCcutoff1) %>% 
        nrow(), 
      "/",
      nrow(diff_exp_lig.p)
      )


#just TNFa/TGFb genes

tnfa_genes <- gene_sets_all$HALLMARK_TNFA_SIGNALING_VIA_NFKB
tgfb_genes <- gene_sets_all$HALLMARK_TGF_BETA_SIGNALING
ggplot(subset(allGenes, ID %in% tgfb_genes),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TGFb+TNFa vs. Untreated, Hallmark TGFb genes", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/G816_DEA_TGFb+TNFa_Hallmark-TGFb.svg", width = 6, height = 6)

#TNFa genes
ggplot(subset(allGenes, ID %in% tnfa_genes),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TGFb+TNFa vs. Untreated, Hallmark TNFa genes", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/G816_DEA_TGFb+TNFa_Hallmark-TNFa.svg", width = 6, height = 6)

```

```{r}
# plot the transcripts from qPCR
PCR_transcripts <- c("FN1", "CD44", "CHI3L1", "COL1A2", "OLIG2")

ggplot(subset(allGenes, gene %in% PCR_transcripts),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TGFb+TNFa vs. Untreated, genes measured by qPCR", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )


#phillips gene sets

ggplot(subset(allGenes, gene %in% c(phillips_mes_sig, phillips_pn_sig)),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(data = function(x){subset(x, gene %in% phillips_mes_sig)},
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_point(data = function(x){subset(x, gene %in% phillips_pn_sig)},
    color = "darkred",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     data = function(x){subset(x, gene %in% phillips_mes_sig)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
     geom_text_repel(
     data = function(x){subset(x, gene %in% phillips_pn_sig)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkred"
   ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TGFb+TNFa vs. Untreated, Phillips genes", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )
```



### GSOA
Now we'll analyze the two lists of differentially expressed genes -- up- and down-regulated due to PMT -- using gene set over-representation analysis (GSOA) using to see what gene sets are over-represented in each set of genes. We will accomplish this using functions from the clusterProfiler package and gene sets from the Molecular Signatures Database (MSigDB). Visualizations of the results will then be generated using functions from clusterProfiler and the GSOAP package.

### Down-regulated genes
Now perform the over-representation tests, starting with the list of down-regulated genes due to TGF+TNF-driven PMT
```{r Down-regulated TGF+TNF gene sets, eval=FALSE, include=FALSE}
#### Over-representation analyses: Down-regulated genes ###############################################
## Significance level:
alpha <- 0.05
p.cutoff <- 0.05 # P-value cutoff
txt.sz <- 6 # Text size for plots
num_cats <- 20 # Number of gene sets to show in dotplots

## Down-regulated genes:
down <- (DEgenes$logFC < 0) %>% 
  DEgenes$ID[.] %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'ENTREZID')

######### GO biological process:
#gobp_or.d <- enrichGO(gene = down,
#                      universe = genes$ensembl,
#                      OrgDb = org.Hs.eg.db, 
#                      ont = "BP", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
#                      qvalueCutoff = 0.05, readable = TRUE)
#p1 <- dotplot(gobp_or.d, 
#        # font.size = 8,
#        showCategory = num_cats) +
#  geom_segment(aes(xend=0, yend = Description), size = 1) +
#  labs(size = "Gene count", x = "Gene Ratio", title = "GO BP over-representation\n(down-regulated genes)")

######### GO cellular component:
#gocc_or.d <- enrichGO(gene = down,
#                      universe = genes$ensembl,
#                      OrgDb = org.Hs.eg.db, 
#                      ont = "CC", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
#                      qvalueCutoff = 0.05, readable = TRUE)
#p2 <- dotplot(gocc_or.d, 
#        # font.size = 8,
#        showCategory = num_cats) +
#  geom_segment(aes(xend=0, yend = Description), size = 1) +
#  labs(size = "Gene count", x = "Gene Ratio", title = "GO CC over-representation\n(down-regulated genes)")

######### GO molecular function:
gomf_or.d <- enrichGO(gene = down,
                        universe = as.character(genes$ensembl),
                        OrgDb = org.Hs.eg.db, ont = "MF", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                        qvalueCutoff = 0.05, readable = TRUE)
p3 <- dotplot(gomf_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "GO MF over-representation\n(down-regulated genes)")

######### KEGG:
kegg_or.d <- enrichKEGG(gene = down,
                        universe = as.character(genes$ensembl),
                        organism = "hsa", keyType = "kegg", 
                        pAdjustMethod = "BH", 
                        pvalueCutoff = p.cutoff, 
                        qvalueCutoff = 0.05
                        )
p4 <- dotplot(kegg_or.d,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "KEGG over-representation\n(down-regulated genes)")
p4
######## BIOCARTA:
biocarta_or.d <- enricher(down, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=cp.b %>% dplyr::select(gs_name, entrez_gene)
                          )
p5 <- dotplot(biocarta_or.d,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"
  #                    ) +
  labs(size = "Gene count", x = "Gene Ratio", title = "BIOCARTA over-representation\n(down-regulated genes)")

######### REACTOME:
reactome_or.d <- enricher(down, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                          TERM2GENE=cp.r %>% dplyr::select(gs_name, entrez_gene))
p6 <- dotplot(reactome_or.d, 
        font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"     
  #                    ) + 
  labs(size = "Gene count", x = "Gene Ratio", title = "REACTOME over-representation\n(down-regulated genes)")

######### PID:
pid_or.d <- enricher(down, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                     TERM2GENE=cp.p %>% dplyr::select(gs_name, entrez_gene))
p7 <- dotplot(pid_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x" 
  #                    ) +   
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "PID over-representation\n(down-regulated genes)")

######### Hallmark:
hallmark_or.d <- enricher(down, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=hall %>% dplyr::select(gs_name, entrez_gene))
p8 <- dotplot(hallmark_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + # USE VIRIDIS COLOR THEME
  # geom_label_repel(aes(label = Description, point.size = Count),  
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"  
  #                    ) +  
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "GSOA: Hallmark up-regulated gene sets\nTGFb+TNFa vs. Untreated")


# plot_grid(p1,p2,p3,p4,p5,p6,p7,p8, align = "h")
pg1 <- plot_grid(p3,p4,p6,p7,p8, align = "h",labels = "AUTO")
pg1
ggsave(plot = pg1, "Overrep_dotplots_downreg-genes_TGF+TNF.png", width = 35, height = 14)

```


```{r Up-regulated TGF+TNF gene sets, eval=FALSE, include=FALSE}
alpha <- 0.05
p.cutoff <- 0.05 # P-value cutoff
txt.sz <- 6 # Text size for plots
num_cats <- 20 # Number of gene sets to show in dotplots

## up-regulated genes:
up <- (DEgenes$logFC > 0) %>% 
  DEgenes$ID[.] %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'ENTREZID')

######### GO biological process:
#gobp_or.d <- enrichGO(gene = up,
#                      universe = genes$ensembl,
#                      OrgDb = org.Hs.eg.db, 
#                      ont = "BP", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
#                      qvalueCutoff = 0.05, readable = TRUE)
#p1 <- dotplot(gobp_or.d, 
#        # font.size = 8,
#        showCategory = num_cats) +
#  geom_segment(aes(xend=0, yend = Description), size = 1) +
#  labs(size = "Gene count", x = "Gene Ratio", title = "GO BP over-representation\n(up-regulated genes)")

######### GO cellular component:
#gocc_or.d <- enrichGO(gene = up,
#                      universe = genes$ensembl,
#                      OrgDb = org.Hs.eg.db, 
#                      ont = "CC", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
#                      qvalueCutoff = 0.05, readable = TRUE)
#p2 <- dotplot(gocc_or.d, 
#        # font.size = 8,
#        showCategory = num_cats) +
#  geom_segment(aes(xend=0, yend = Description), size = 1) +
#  labs(size = "Gene count", x = "Gene Ratio", title = "GO CC over-representation\n(up-regulated genes)")

######### GO molecular function:
gomf_or.d <- enrichGO(gene = up,
                        universe = as.character(genes$ensembl),
                        OrgDb = org.Hs.eg.db, ont = "MF", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                        qvalueCutoff = 0.05, readable = TRUE)
p3 <- dotplot(gomf_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "GO MF over-representation\n(up-regulated genes)")

######### KEGG:
kegg_or.d <- enrichKEGG(gene = up,
                        universe = as.character(genes$ensembl),
                        organism = "hsa", keyType = "kegg", 
                        pAdjustMethod = "BH", pvalueCutoff = p.cutoff, qvalueCutoff = 0.05,
                        )
p4 <- dotplot(kegg_or.d,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "KEGG over-representation\n(up-regulated genes)")
p4
######## BIOCARTA:
biocarta_or.d <- enricher(up, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=cp.b %>% dplyr::select(gs_name, entrez_gene)
                          )
p5 <- dotplot(biocarta_or.d,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"
  #                    ) +
  labs(size = "Gene count", x = "Gene Ratio", title = "BIOCARTA over-representation\n(up-regulated genes)")

######### REACTOME:
reactome_or.d <- enricher(up, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                          TERM2GENE=cp.r %>% dplyr::select(gs_name, entrez_gene))
p6 <- dotplot(reactome_or.d, 
        font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"     
  #                    ) + 
  labs(size = "Gene count", x = "Gene Ratio", title = "REACTOME over-representation\n(up-regulated genes)")

######### PID:
pid_or.d <- enricher(up, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                     TERM2GENE=cp.p %>% dplyr::select(gs_name, entrez_gene))
p7 <- dotplot(pid_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x" 
  #                    ) +   
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "PID over-representation\n(up-regulated genes)")

######### Hallmark:
hallmark_or.d <- enricher(up, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=hall %>% dplyr::select(gs_name, entrez_gene))
p8 <- dotplot(hallmark_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + # USE VIRIDIS COLOR THEME
  # geom_label_repel(aes(label = Description, point.size = Count),  
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"  
  #                    ) +  
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "GSOA: Hallmark up-regulated gene sets\nTGFb+TNFa vs. Untreated")
p8
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/TGF+TNF_GSOA.svg", width = 10, height = 6)

# plot_grid(p1,p2,p3,p4,p5,p6,p7,p8, align = "h")
#pg1 <- plot_grid(p3,p4,p6,p7,p8, align = "h",labels = "AUTO")
#pg1
#ggsave(plot = pg1, "Overrep_dotplots_upreg-genes_TGF+TNF.png", width = 35, height = 14)
```
##GSVA
```{r GSVA}
data_for_gsva <- logcpm
#rownames(data_for_gsva) <- exp.symbol$GeneID
force_calculate_gsva <- F
if(!file.exists("gsva_results.rds") | isTRUE(force_calculate_gsva)){
  gsva_res <- gsva(
    data_for_gsva, 
    gene_sets_all, verbose = F, # Change which gene sets are run
    # gene_sets4, verbose = T
    method = "gsva", 
#    kcdf = "Gaussian",
    # abs.ranking = F,
    # mx.diff = T,
    min.sz = 10, # Minimum number of genes required to include a gene set
    parallel.sz=detectCores()#-1
    )
  ## Save GSVA results for speed:
  saveRDS(gsva_res, file = "gsva_results.rds")
  
} else if(file.exists("gsva_results.rds") & !exists("gsva_res")){ 
  ## Load GSVA results, if they haven't been loaded/calculated already:
  gsva_res <- readRDS("gsva_results.rds") 
}
## Build data frame versions of GSVA results:
gsva_res.df <- gsva_res %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
gsva_res.df.t <- gsva_res %>% t() %>% data.frame(ID=rownames(.), .)
rownames(gsva_res.df) <- gsva_res.df$gene.set
```
###TGF+TNF
```{r plot GSVA results for TGF+TNF vs Unt samples}
gsva.res.tt <- gsva_res.df %>% dplyr::select(contains("Unt")|contains("TT"))

logFCcutoff1 <- 0 # Threshold for log2(fold-change)
adjPvalueCutoff <- 10^-2 # Threshold for adjusted p-values (FDR) - use for GO gene set
### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- model.matrix(~0 + groups$group)[1:6,1:2]
colnames(design1) <- c("Unt","TT")
fit1 <- lmFit(gsva.res.tt, design1) 
contr1 <- makeContrasts(TTvsUnt = TT - Unt, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff)
summary(res1)
DEgenes

DEgenes$logFC %>% order() %>% DEgenes[.,] %>% write.csv("DEgene_sets_TGF+TNF.csv")


#volcano plot for differentially expression gene sets
ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff),
                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
  geom_text_repel(
    data = function(x){subset(x, adj.P.Val < 10^-4 & logFC<logFCcutoff1)},
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff & logFC < 0)},
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff & ID %in% gene_sets_to_plot)},
    size = 3, alpha = 0.9,
    segment.size = 0.25, segment.alpha = .8, force=30, max.overlaps = 50
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 25),
    plot.title = element_text(size = 23),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "Gene Sets: Untreated vs. TGFb+TNFa",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits=c(0,5))+  scale_x_continuous(limits=c(-1.5, 1.5)) #full view
#  scale_y_continuous(limits=c(70,200))+  scale_x_continuous(limits=c(0, 5)) #zoomed view
#ggsave("GSVA_DE_TGF+TNF-neg.png", width = 16, height = 12)

sets.sub <- append(cp.p$gs_name, cp.b$gs_name) %>% append(hall$gs_name) #these gene sets are useful for actual pathways
sets.sub <- go.bp %>% filter(str_detect(gs_name, "SIGNALING|PHOSPHORYLATION")) %>% .$gs_name 
#sets.sub <- rbind(cp.p, cp.b, hall) %>% filter(str_detect(gs_name, "PATHWAY|SIGNALING")) %>% .$gs_name 

#sets.sub <- append(go.tft$gs_name, go.tft_leg$gs_name)
sets.sub.tf <- go.tft$gs_name #these gene sets are for transcription factors

allGenes %>% filter(ID %in% sets.sub)
sets.tt.up <- top_n(allGenes %>% filter(adj.P.Val<0.05, logFC>0) %>% 
                      filter(ID %in% sets.sub), 20, logFC)
sets.tt.dn <- top_n(allGenes %>% filter(adj.P.Val<0.05, logFC<0) %>% 
                      filter(ID %in% sets.sub), -20, logFC)

tf.sets.tt.up <- top_n(allGenes %>% filter(ID %in% sets.sub.tf), 20, logFC)
tf.sets.tt.dn <- top_n(allGenes %>% filter(ID %in% sets.sub.tf), -20, logFC)


```
###TMZ
```{r plot GSVA results for TMZ vs Unt samples}
gsva.res.tmz <- gsva_res.df %>% dplyr::select(contains("Unt")|contains("TMZ"))

logFCcutoff1 <- 0.5 # Threshold for log2(fold-change)
adjPvalueCutoff <- 10^-2 # Threshold for adjusted p-values (FDR) - use for GO gene set
### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- model.matrix(~0 + groups$group)[1:6,1:2]
colnames(design1) <- c("Unt","TMZ")
fit1 <- lmFit(gsva.res.tmz, design1) 
contr1 <- makeContrasts(TMZvsUnt = TMZ - Unt, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff)
summary(res1)
DEgenes

DEgenes$logFC %>% order() %>% DEgenes[.,] %>% write.csv("DEgene_sets_TMZ.csv")




#volcano plot for differentially expression gene sets
ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff),
                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
  geom_text_repel(
    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff & logFC<logFCcutoff1)},
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff & logFC < 0)},
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff & ID %in% gene_sets_to_plot)},
    size = 3, alpha = 0.9,
    segment.size = 0.25, segment.alpha = .8, force=50, max.overlaps = 30
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 25),
    plot.title = element_text(size = 23),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "Gene Sets: Untreated vs. TMZ",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits=c(0,5))+  scale_x_continuous(limits=c(-1.5, 1.5)) #full view
#  scale_y_continuous(limits=c(70,200))+  scale_x_continuous(limits=c(0, 5)) #zoomed view
#ggsave("GSVA_DE_TMZ-neg.png", width = 16, height = 12)

#sets.sub <- append(cp.p$gs_name, cp.b$gs_name)

#allGenes %>% filter(ID %in% sets.sub)
sets.tmz.up <- top_n(allGenes %>% filter(adj.P.Val<0.05, logFC>0) %>% 
                       filter(ID %in% sets.sub), 20, logFC)
sets.tmz.dn <- top_n(allGenes %>% filter(adj.P.Val<0.05, logFC<0) %>% 
                       filter(ID %in% sets.sub), -20, logFC)

tf.sets.tmz.up <- top_n(allGenes %>% filter(ID %in% sets.sub.tf), 20, logFC)
tf.sets.tmz.dn <- top_n(allGenes %>% filter(ID %in% sets.sub.tf), -20, logFC)
```

```{r barplots of GSVA results}
#plotting ligand samples
sets.tt.up %>%
  arrange(logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Up-regulated Gene Sets: TGF+TNF-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
ggsave("gsva_up_TT.png", width = 10, height = 8)

sets.tt.dn %>%
  arrange(-logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Down-regulated Gene Sets: TGF+TNF-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
#ggsave("gsva_dn_TT.png", width = 15, height = 6)

#plotting TMZ samples
sets.tmz.up %>%
  arrange(logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Up-regulated Gene Sets: TMZ-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
ggsave("gsva_up_TMZ.png", width = 10, height = 8)

sets.tmz.dn %>%
  arrange(-logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Down-regulated Gene Sets: TMZ-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
#ggsave("gsva_dn_TMZ.png", width = 15, height = 6)
```

```{r barplots of TF GSVA results}
#plotting ligand samples
tf.sets.tt.up %>%
  arrange(logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Up-regulated Gene Sets: TGF+TNF-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
ggsave("gsva_tf_up_TT_.png", width = 15, height = 6)

tf.sets.tt.dn %>%
  arrange(-logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Down-regulated Gene Sets: TGF+TNF-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
ggsave("gsva_tf_dn_TT.png", width = 15, height = 6)

#plotting TMZ samples
tf.sets.tmz.up %>%
  arrange(logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Up-regulated Gene Sets: TMZ-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
ggsave("gsva_tf_up_TMZ.png", width = 15, height = 6)

tf.sets.tmz.dn %>%
  arrange(-logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Down-regulated Gene Sets: TMZ-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
ggsave("gsva_tf_dn_TMZ.png", width = 15, height = 6)

```
## DEA for Unt vs. TMZ samples
```{r}
## Statistical thresholds:
adjPvalueCutoff1 <- 0.01 # Threshold for adjusted p-values (FDR) for GENE DEA
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)

## Gene expression data:
data_for_lm <- data.frame(logcpm) %>% dplyr::select(-contains("TT"))
colnames(data_for_lm) <- c("Unt", "Unt", "Unt", "TMZ", "TMZ", "TMZ")
#colnames(data_for_lm) <- data3 %>% rownames() # Col names = samples
#rownames(data_for_lm) <- colnames(data3) #%>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') # Row names = Entrez gene IDs

### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- model.matrix(~0 + groups$group)[1:6,1:2]
colnames(design1) <- c("Unt","TMZ")
fit1 <- lmFit(data_for_lm, design1) 
contr1 <- makeContrasts(TMZvsUnt = TMZ - Unt, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes


########## Various visualizations ########## 
## Gene LFC vs. average log-cpm:
plotMD(tmp1, column=1, status=res1[,1], main=colnames(tmp1)[1], 
       xlim=c(-8,13))

###################################### Volcano plot --> GENES ######################################
txtsz <- 14
linesz <- 1

allGenes <- mutate(allGenes, gene = allGenes$ID %>% match(genes$ensembl) %>% genes$gene[.])
DEAgenes.TMZ <- allGenes
DEAgenes.TMZ$logFC %>% order() %>% DEAgenes.TMZ[.,] %>% write.csv("ordered_DEgenes_TMZ.csv")
ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1 %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_point(data = function(x){subset(x, ID %in% "TMZ")},
    color = "black",
    shape = 15,
    # cex = 2.5
  ) +
    geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
   geom_text_repel(
     data = function(x){subset(x, (adj.P.Val < adjPvalueCutoff1) & (abs(logFC) > 2))},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 15, max.iter = 100000
   ) +
  geom_text_repel(data = function(x){subset(x, ID == "TMZ")}) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TMZ vs. Untreated", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )
ggsave("DEA-genes_TMZ.png", width = 6, height = 6)

#just verhaak genes
ggplot(subset(allGenes, ID %in% c(verhaak.mes, verhaak.pn)),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(data = function(x){subset(x, ID %in% verhaak.mes)},
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_point(data = function(x){subset(x, ID %in% verhaak.pn)},
    color = "darkred",
    shape = "triangle",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     data = function(x){subset(x, ID %in% verhaak.mes & abs(logFC) > logFCcutoff1 & adj.P.Val < adjPvalueCutoff1)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
     geom_text_repel(
     data = function(x){subset(x, ID %in% verhaak.pn & abs(logFC) > logFCcutoff1 & adj.P.Val < adjPvalueCutoff1)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkred"
   ) +
  geom_text(aes(x = -2, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TMZ vs. Untreated", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )
ggsave("DEA-genes_TMZ_verhaak.png", width = 6, height = 6)
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/G816_DEA_TMZ_Verhaak.svg", width = 6, height = 6)

## printing stats for up- and down-regulated transcripts
diff_exp_tmz.m <- allGenes %>% filter(ID %in% verhaak.mes)
paste("Mesenchymal Up:" , diff_exp_tmz.m %>% 
        filter(adj.P.Val < adjPvalueCutoff1) %>% 
        filter(logFC > logFCcutoff1) %>% 
        nrow(), 
      "/",
      nrow(diff_exp_tmz.m)
      )
paste("Mesenchymal Down:" , diff_exp_tmz.m %>% 
        filter(adj.P.Val < adjPvalueCutoff1) %>% 
        filter(logFC < -logFCcutoff1) %>% 
        nrow(), 
      "/",
      nrow(diff_exp_tmz.m)
      )

diff_exp_tmz.p <- allGenes %>% filter(ID %in% verhaak.pn)
paste("Proneural Up:" , diff_exp_tmz.p %>% 
        filter(adj.P.Val < adjPvalueCutoff1) %>% 
        filter(logFC > logFCcutoff1) %>% 
        nrow(), 
      "/",
      nrow(diff_exp_tmz.p)
      )
paste("Proneural Down:" , diff_exp_tmz.p %>% 
        filter(adj.P.Val < adjPvalueCutoff1) %>% 
        filter(logFC < -logFCcutoff1) %>% 
        nrow(), 
      "/",
      nrow(diff_exp_tmz.p)
      )


hall.tnfa <- gene_sets_all$HALLMARK_TNFA_SIGNALING_VIA_NFKB
hall.tgfb <- gene_sets_all$HALLMARK_TGF_BETA_SIGNALING

##plot genes that show up in TGFb, TNFa gene sets
ggplot(subset(allGenes, ID %in% hall.tgfb),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TMZ vs. Untreated", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )

```

```{r}
#plotting transcripts measured by qPCR
PCR_transcripts <- c("FN1", "CD44", "CHI3L1", "COL1A2", "OLIG2")

ggplot(subset(allGenes, gene %in% PCR_transcripts),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TMZ vs. Untreated, genes measured by qPCR", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )

#how about also plotting Phillips genes

ggplot(subset(allGenes, gene %in% c(phillips_mes_sig, phillips_pn_sig)),
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(data = function(x){subset(x, gene %in% phillips_mes_sig)},
    color = "darkblue",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_point(data = function(x){subset(x, gene %in% phillips_pn_sig)},
    color = "darkred",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text_repel(
     data = function(x){subset(x, gene %in% phillips_mes_sig)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
     geom_text_repel(
     data = function(x){subset(x, gene %in% phillips_pn_sig)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkred"
   ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA: TMZ vs. Untreated, Phillips genes", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )
```


```{r}
logFCcutoff1 <- 0 # Threshold for log2(fold-change)
adjPvalueCutoff <- 10^-2 # Threshold for adjusted p-values (FDR) - use for GO gene set
### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- model.matrix(~0 + groups$group)[1:9,1:3]
colnames(design1) <- c("Unt","TT", "TMZ")
fit1 <- lmFit(gsva_res.df[,-1], design1) 
contr1 <- makeContrasts(TTvsUnt = TT - Unt, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff)
summary(res1)
DEgenes

#DEgenes$logFC %>% order() %>% DEgenes[.,] %>% write.csv("DEgene_sets_TGF+TNF.csv")


#volcano plot for differentially expression genes
ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff),
                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
  geom_text_repel(
    data = function(x){subset(x, adj.P.Val < 10^-4 & logFC<logFCcutoff1)},
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff & logFC < 0)},
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff & ID %in% gene_sets_to_plot)},
    size = 3, alpha = 0.9,
    segment.size = 0.25, segment.alpha = .8, force=30, max.overlaps = 50
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 25),
    plot.title = element_text(size = 23),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "Gene Sets: Untreated vs. TGFb+TNFa",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits=c(0,5))+  scale_x_continuous(limits=c(-1.5, 1.5)) #full view
#  scale_y_continuous(limits=c(70,200))+  scale_x_continuous(limits=c(0, 5)) #zoomed view



```


### GSOA
Now we'll analyze the two lists of differentially expressed genes -- up- and down-regulated due to PMT -- using gene set over-representation analysis (GSOA) using to see what gene sets are over-represented in each set of genes. We will accomplish this using functions from the clusterProfiler package and gene sets from the Molecular Signatures Database (MSigDB). Visualizations of the results will then be generated using functions from clusterProfiler and the GSOAP package.

### Down-regulated genes
Now perform the over-representation tests, starting with the list of down-regulated genes due to TMZ-driven PMT
```{r Down-regulated TMZ gene sets, eval=FALSE, include=FALSE}
#### Over-representation analyses: Down-regulated genes ###############################################
## Significance level:
alpha <- 0.05
p.cutoff <- 0.05 # P-value cutoff
txt.sz <- 6 # Text size for plots
num_cats <- 20 # Number of gene sets to show in dotplots

## Down-regulated genes:
down <- (DEgenes$logFC < 0) %>% 
  DEgenes$ID[.] %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'ENTREZID')

######### GO biological process:
gobp_or.d <- enrichGO(gene = down,
                      universe = as.character(genes$ensembl),
                      OrgDb = org.Hs.eg.db, 
                      ont = "BP", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                      qvalueCutoff = 0.05, readable = TRUE)
p1 <- dotplot(gobp_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "GO BP over-representation\n(down-regulated genes)")

######### GO cellular component:
gocc_or.d <- enrichGO(gene = down,
                      universe = as.character(genes$ensembl),
                      OrgDb = org.Hs.eg.db, 
                      ont = "CC", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                      qvalueCutoff = 0.05, readable = TRUE)
p2 <- dotplot(gocc_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "GO CC over-representation\n(down-regulated genes)")

######### GO molecular function:
gomf_or.d <- enrichGO(gene = down,
                        universe = as.character(genes$ensembl),
                        OrgDb = org.Hs.eg.db, ont = "MF", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                        qvalueCutoff = 0.05, readable = TRUE)
p3 <- dotplot(gomf_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "GO MF over-representation\n(down-regulated genes)")

######### KEGG:
kegg_or.d <- enrichKEGG(gene = down,
                        universe = as.character(genes$ensembl),
                        organism = "hsa", keyType = "kegg", 
                        pAdjustMethod = "BH", pvalueCutoff = p.cutoff, qvalueCutoff = 0.05,
                        )
p4 <- dotplot(kegg_or.d,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "KEGG over-representation\n(down-regulated genes)")

######## BIOCARTA:
biocarta_or.d <- enricher(down, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=cp.b %>% dplyr::select(gs_name, entrez_gene)
                          )
p5 <- dotplot(biocarta_or.d,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"
  #                    ) +
  labs(size = "Gene count", x = "Gene Ratio", title = "BIOCARTA over-representation\n(down-regulated genes)")

######### REACTOME:
reactome_or.d <- enricher(down, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                          TERM2GENE=cp.r %>% dplyr::select(gs_name, entrez_gene))
p6 <- dotplot(reactome_or.d, 
        font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"     
  #                    ) + 
  labs(size = "Gene count", x = "Gene Ratio", title = "REACTOME over-representation\n(down-regulated genes)")

######### PID:
pid_or.d <- enricher(down, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                     TERM2GENE=cp.p %>% dplyr::select(gs_name, entrez_gene))
p7 <- dotplot(pid_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x" 
  #                    ) +   
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "PID over-representation\n(down-regulated genes)")

######### Hallmark:
hallmark_or.d <- enricher(down, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=hall %>% dplyr::select(gs_name, entrez_gene))
p8 <- dotplot(hallmark_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + # USE VIRIDIS COLOR THEME
  # geom_label_repel(aes(label = Description, point.size = Count),  
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"  
  #                    ) +  
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "Hallmark over-representation\n(down-regulated genes)")


# plot_grid(p1,p2,p3,p4,p5,p6,p7,p8, align = "h")
pg1 <- plot_grid(p3,p4,p6,p7,p8, align = "h",labels = "AUTO")
pg1
#ggsave(plot = pg1, "Overrep_dotplots_downreg-genes_TMZ.png", width = 35, height = 14)

```

```{r Up-regulated TMZ gene sets, eval=FALSE, include=FALSE}
alpha <- 0.05
p.cutoff <- 0.05 # P-value cutoff
txt.sz <- 6 # Text size for plots
num_cats <- 20 # Number of gene sets to show in dotplots

## up-regulated genes:
up <- (DEgenes$logFC > 0) %>% 
  DEgenes$ID[.] %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'ENTREZID')

######### GO biological process:
gobp_or.d <- enrichGO(gene = up,
                      universe = as.character(genes$ensembl),
                      OrgDb = org.Hs.eg.db, 
                      ont = "BP", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                      qvalueCutoff = 0.05, readable = TRUE)
p1 <- dotplot(gobp_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "GO BP over-representation\n(up-regulated genes)")

######### GO cellular component:
gocc_or.d <- enrichGO(gene = up,
                      universe = as.character(genes$ensembl),
                      OrgDb = org.Hs.eg.db, 
                      ont = "CC", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                      qvalueCutoff = 0.05, readable = TRUE)
p2 <- dotplot(gocc_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "GO CC over-representation\n(up-regulated genes)")

######### GO molecular function:
gomf_or.d <- enrichGO(gene = up,
                        universe = as.character(genes$ensembl),
                        OrgDb = org.Hs.eg.db, ont = "MF", pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                        qvalueCutoff = 0.05, readable = TRUE)
p3 <- dotplot(gomf_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "GO MF over-representation\n(up-regulated genes)")

######### KEGG:
kegg_or.d <- enrichKEGG(gene = up,
                        universe = as.character(genes$ensembl),
                        organism = "hsa", keyType = "kegg", 
                        pAdjustMethod = "BH", pvalueCutoff = p.cutoff, qvalueCutoff = 0.05,
                        )
p4 <- dotplot(kegg_or.d,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "KEGG over-representation\n(up-regulated genes)")

######## BIOCARTA:
biocarta_or.d <- enricher(up, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=cp.b %>% dplyr::select(gs_name, entrez_gene)
                          )
p5 <- dotplot(biocarta_or.d,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"
  #                    ) +
  labs(size = "Gene count", x = "Gene Ratio", title = "BIOCARTA over-representation\n(up-regulated genes)")

######### REACTOME:
reactome_or.d <- enricher(up, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                          TERM2GENE=cp.r %>% dplyr::select(gs_name, entrez_gene))
p6 <- dotplot(reactome_or.d, 
        font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"     
  #                    ) + 
  labs(size = "Gene count", x = "Gene Ratio", title = "REACTOME over-representation\n(up-regulated genes)")

######### PID:
pid_or.d <- enricher(up, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                     TERM2GENE=cp.p %>% dplyr::select(gs_name, entrez_gene))
p7 <- dotplot(pid_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x" 
  #                    ) +   
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "PID over-representation\n(up-regulated genes)")

######### Hallmark:
hallmark_or.d <- enricher(up, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=hall %>% dplyr::select(gs_name, entrez_gene))
p8 <- dotplot(hallmark_or.d, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + # USE VIRIDIS COLOR THEME
  # geom_label_repel(aes(label = Description, point.size = Count),  
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"  
  #                    ) +  
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "GSOA: Hallmark up-regulated gene sets\nTMZ vs. Untreated")
p8
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/TMZ_GSOA.svg", width = 10, height = 6)


# plot_grid(p1,p2,p3,p4,p5,p6,p7,p8, align = "h")
pg1 <- plot_grid(p3,p4,p6,p7,p8, align = "h",labels = "AUTO")
pg1
#ggsave(plot = pg1, "Overrep_dotplots_upreg-genes_TMZ.png", width = 35, height = 14)
```


## PCA - Verhaak gene sets
```{r Verhaak gene set analysis}
species = "Homo sapiens"
cgp.sig <- msigdbr(species = species, category = "C2", subcategory = "CGP") %>% 
  split(x = .$entrez_gene, f = .$gs_name)

mes.sig <- cgp.sig$VERHAAK_GLIOBLASTOMA_MESENCHYMAL
pn.sig <- cgp.sig$VERHAAK_GLIOBLASTOMA_PRONEURAL
pmt.sig <- append(mes.sig, pn.sig)

cpm.pmt <- logcpm %>% data.frame() %>%
  rownames_to_column(var = "gene") %>%
  filter(gene %in% pmt.sig) %>%
  column_to_rownames(var = "gene")
```

```{r PCA based on Verhaak PMT sig}
library(ggfortify)
library(matrixStats)
set.seed(100)
pcamat.pmt <- cpm.pmt %>% data.matrix()
colnames(pcamat.pmt) <- samples
pcamat.all <- logcpm %>% data.matrix()

rownames(pcamat.pmt) %<>% match(genes$ensembl) %>% genes$gene[.]
pcamat.vars <- rowVars(pcamat.pmt)

#pcamat.pmt %<>% data.frame() %>% select(-contains("TT")) %>% data.matrix()

pcamat.pmt <- pcamat.pmt %>% data.frame() %>% add_column(var = pcamat.vars) %>%
  filter(var > 0.5) %>% dplyr::select(-var) %>% data.matrix()

#ggsave("PCA-all-genes.png")
pcamat.t <- t(pcamat.pmt) 
pcamat.z <- pcamat.t %>% scale(center = T, scale = T) %>% t()

pca.pmt.3 <- prcomp(pcamat.t)
pca.scores <- pca.pmt.3$x
pca.loadings <- pca.pmt.3$rotation

samples.1 <- rownames(pca.scores)

pca.loadings <- pca.loadings%>% data.frame() %>% 
  rownames_to_column(var = "gene") %>% mutate(
    subtype = case_when(
      gene %>% match(genes$gene) %>% genes$ensembl[.] %in% pn.sig ~ "PN",
      gene %>% match(genes$gene) %>% genes$ensembl[.] %in% mes.sig ~ "MES",
      TRUE ~ "wut"
    )
  )


plt.loadings <- ggplot(data = pca.loadings,
       aes(x=PC1, y=PC2, color = subtype, label = gene))+
  geom_text()+
  theme_cowplot()+
  geom_hline(yintercept = 0, linetype = "dashed", alpha = 0.5)+
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)+
  labs(title = "PCA - loadings")+
  theme(legend.position = "none")
plt.loadings

pca.scores %<>% data.frame() %>% rownames_to_column(var = "sample")
plt.scores <- ggplot(data = pca.scores,
       aes(x=PC1, y=PC2, label = samples.1))+
  geom_point()+
  geom_text_repel()+
  theme_cowplot()+
  geom_hline(yintercept = 0, linetype = "dashed", alpha = 0.5)+
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)+
  labs(title = "PCA - scores")
plt.scores

plot_grid(plt.scores, plt.loadings)
#ggsave("varhaak-pca-loadings+scores.png", width = 12, height = 5)


##running UMAP as an alternative to PCA
embedding1 <- umap::umap(pcamat.pmt)

## Plot the result:
umap.res <- embedding1$layout %>%
  data.frame() %>% mutate(ID = rownames(.))%>% mutate(
    subtype = case_when(
      ID %>% match(genes$gene) %>% genes$ensembl[.] %in% pn.sig ~ "PN",
      ID %>% match(genes$gene) %>% genes$ensembl[.] %in% mes.sig ~ "MES",
      TRUE ~ "??"
    )
  )


ggplot(data = umap.res, aes(X1,X2, label = ID, color = subtype)) + 
  geom_text() + 
  #geom_point()+
  theme_cowplot(12) + labs(title="UMAP of RNA-seq data", x = "UMAP1", y = "UMAP2")+
  theme(legend.position = "none")
```
##PCA - Neftel gene sets
```{r PCA based on Neftel et al ssRNAseq gene sets, message=FALSE}
gene_sets5 <- read_tsv("IDHwt.GBM.MetaModules.tsv")
gene_sets5.mes1 <- pull(gene_sets5, MESlike1)
gene_sets5.mes2 <- pull(gene_sets5, MESlike2)
gene_sets5.mes <- append(gene_sets5.mes1, gene_sets5.mes2)

gene_sets5.pn <- gene_sets5 %>% pull(OPClike)
pmt.sig2 <- append(gene_sets5.mes, gene_sets5.pn)
pmt.sig2 <- pmt.sig2 %>% data.frame()
colnames(pmt.sig2)[1] <- "gene"

pmt.sig2 <- mutate(pmt.sig2, ID = pmt.sig2$gene %>% match(genes$gene) %>% genes$ensembl[.])

cpm.pmt <- logcpm %>% data.frame() %>%
  rownames_to_column(var = "gene") %>%
  filter(gene %in% pmt.sig2$ID) %>%
  column_to_rownames(var = "gene")
```

```{r PCA based on Suva PMT gene sets}
pcamat.pmt <- cpm.pmt %>% data.matrix()
colnames(pcamat.pmt) <- samples

rownames(pcamat.pmt) %<>% match(genes$ensembl) %>% genes$gene[.]
pcamat.vars <- rowVars(pcamat.pmt)

#pcamat.pmt %<>% data.frame() %>% select(-contains("TMZ")) %>% data.matrix()

pcamat.pmt <- pcamat.pmt %>% data.frame() %>% add_column(var = pcamat.vars) %>%
  filter(var > 0.5) %>% dplyr::select(-var) %>% data.matrix()

pcamat.t <- t(pcamat.pmt)
pcamat.z <- pcamat.t %>% scale(center = T, scale = T) %>% t()

pca.pmt.3 <- prcomp(pcamat.t)
pca.scores <- pca.pmt.3$x
pca.loadings <- pca.pmt.3$rotation

samples.1 <- rownames(pca.scores)

pca.loadings <- pca.loadings%>% data.frame() %>% 
  rownames_to_column(var = "gene") %>% mutate(
    subtype = case_when(
      gene %in% gene_sets5.pn ~ "PN",
      gene %in% gene_sets5.mes ~ "MES",
      TRUE ~ "??"
    )
  )


plt.loadings <- ggplot(data = pca.loadings,
       aes(x=PC1, y=PC2, color = subtype, label = gene))+
  geom_text()+
  theme_cowplot()+
  geom_hline(yintercept = 0, linetype = "dashed", alpha = 0.5)+
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)+
  labs(title = "PCA - loadings")+
  theme(legend.position = "none")
plt.loadings

pca.scores %<>% data.frame() %>% rownames_to_column(var = "sample")
plt.scores <- ggplot(data = pca.scores,
       aes(x=PC1, y=PC2, label = samples.1))+
  geom_point()+
  geom_text_repel()+
  theme_cowplot()+
  geom_hline(yintercept = 0, linetype = "dashed", alpha = 0.5)+
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)+
  labs(title = "PCA - scores")
plt.scores

plot_grid(plt.scores, plt.loadings)
#ggsave("suva-pca-loadings+scores.png", width = 12, height = 5)


##running UMAP as an alternative to PCA
embedding1 <- umap::umap(pcamat.pmt)

## Plot the result:
umap.res <- embedding1$layout %>%
  data.frame() %>% mutate(ID = rownames(.)) %>% mutate(
    subtype = case_when(
      ID %in% gene_sets5.pn ~ "PN",
      ID %in% gene_sets5.mes ~ "MES",
      TRUE ~ "??"
    )
  )

ggplot(data = umap.res, aes(X1,X2, label = ID, color = subtype)) + 
  #geom_point()+
  #geom_text_repel(max.overlaps = 40) + 
  geom_text()+
  theme_cowplot(18) + labs(title="UMAP of RNA-seq data", x = "UMAP1", y = "UMAP2")+
  theme(legend.position = "none")
```
##PCA - GO EMT gene set
```{r GO EMT gene set analysis}
go.emt.sig <- go.bp %>% filter(gs_name == "GO_EPITHELIAL_TO_MESENCHYMAL_TRANSITION")

cpm.pmt <- logcpm %>% data.frame() %>%
  rownames_to_column(var = "gene") %>%
  filter(gene %in% go.emt.sig$entrez_gene) %>%
  column_to_rownames(var = "gene")
```

```{r PCA based on GO EMT sig, eval=FALSE, include=FALSE}
library(ggfortify)
library(matrixStats)
pcamat.pmt <- cpm.pmt %>% data.matrix()
colnames(pcamat.pmt) <- samples
pcamat.all <- logcpm %>% data.matrix()

rownames(pcamat.pmt) %<>% match(genes$ensembl) %>% genes$gene[.]
pcamat.vars <- rowVars(pcamat.pmt)
#pcamat.pmt <- pcamat.pmt %>% data.frame() %>% add_column(var = pcamat.vars) %>%
#  filter(var > 0.3) %>% select(-var) %>% data.matrix()
#write.csv(pcamat.pmt, file = "g816_go_emt_genes_all.csv")
#ggsave("PCA-all-genes.png")
pcamat.t <- t(pcamat.pmt)


## PMT gene expression heatmap
exp.pmt.z <- pcamat.t %>% scale(center=T, scale=T) %>% t()
exp.genes <- rownames(exp.pmt.z) %>% data.frame()
colnames(exp.genes)[1] <- "gene"

pca.pmt.3 <- prcomp(pcamat.t)
pca.scores <- pca.pmt.3$x
pca.loadings <- pca.pmt.3$rotation
autoplot(pca.pmt.3, label = T,
         loadings.label = T, loadings.label.size = 3, 
         loadings.colour = "white", loadings.label.colour = "blue")+
  theme_cowplot()

samples.1 <- rownames(pca.scores)

pca.loadings <- pca.loadings%>% data.frame() %>% 
  rownames_to_column(var = "gene")


plt.loadings <- ggplot(data = pca.loadings,
       aes(x=PC1, y=PC2, label = gene))+
  geom_text()+
  theme_cowplot()+
  geom_hline(yintercept = 0, linetype = "dashed", alpha = 0.5)+
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)+
  labs(title = "PCA - loadings")+
  theme(legend.position = "none")
plt.loadings

pca.scores %<>% data.frame() %>% rownames_to_column(var = "sample")
plt.scores <- ggplot(data = pca.scores,
       aes(x=PC1, y=PC2, label = samples))+
  geom_point()+
  geom_text_repel()+
  theme_cowplot()+
  geom_hline(yintercept = 0, linetype = "dashed", alpha = 0.5)+
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)+
  labs(title = "PCA - scores")
plt.scores

plot_grid(plt.scores, plt.loadings)
#ggsave("GO_EMT-pca-loadings+scores.png", width = 12, height = 5)
```

##PCA - Hallmark EMT
```{r hallmark EMT gene set analysis}
hall.emt.sig <- hall %>% filter(gs_name == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION")

cpm.pmt <- logcpm %>% data.frame() %>%
  rownames_to_column(var = "gene") %>%
  filter(gene %in% hall.emt.sig$entrez_gene) %>%
  column_to_rownames(var = "gene")
```

```{r PCA based on hallmark EMT sig}
library(ggfortify)
library(matrixStats)
pcamat.pmt <- cpm.pmt %>% data.matrix()
colnames(pcamat.pmt) <- samples
pcamat.all <- logcpm %>% data.matrix()

rownames(pcamat.pmt) %<>% match(genes$ensembl) %>% genes$gene[.]
pcamat.vars <- rowVars(pcamat.pmt)
pcamat.pmt <- pcamat.pmt %>% data.frame() %>% add_column(var = pcamat.vars) %>%
  filter(var > 0.5) %>% dplyr::select(-var) %>% data.matrix()

#ggsave("PCA-all-genes.png")
pcamat.t <- t(pcamat.pmt)


## PMT gene expression heatmap
exp.pmt.z <- pcamat.t %>% scale(center=T, scale=T) %>% t()
exp.genes <- rownames(exp.pmt.z) %>% data.frame()
colnames(exp.genes)[1] <- "gene"

pca.pmt.3 <- prcomp(pcamat.t)
pca.scores <- pca.pmt.3$x
pca.loadings <- pca.pmt.3$rotation
autoplot(pca.pmt.3, label = T,
         loadings.label = T, loadings.label.size = 3, 
         loadings.colour = "white", loadings.label.colour = "blue")+
  theme_cowplot()

samples.1 <- rownames(pca.scores)

pca.loadings <- pca.loadings%>% data.frame() %>% 
  rownames_to_column(var = "gene")


plt.loadings <- ggplot(data = pca.loadings,
       aes(x=PC1, y=PC2, label = gene))+
  geom_text()+
  theme_cowplot()+
  geom_hline(yintercept = 0, linetype = "dashed", alpha = 0.5)+
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)+
  labs(title = "PCA - loadings")+
  theme(legend.position = "none")
plt.loadings

pca.scores %<>% data.frame() %>% rownames_to_column(var = "sample")
plt.scores <- ggplot(data = pca.scores,
       aes(x=PC1, y=PC2, label = samples))+
  geom_point()+
  geom_text_repel()+
  theme_cowplot()+
  geom_hline(yintercept = 0, linetype = "dashed", alpha = 0.5)+
  geom_vline(xintercept = 0, linetype = "dashed", alpha = 0.5)+
  labs(title = "PCA - scores")
plt.scores

plot_grid(plt.scores, plt.loadings)
#ggsave("hallmark_EMT-pca-loadings+scores.png", width = 12, height = 5)
```


```{r}
## Statistical thresholds:
adjPvalueCutoff1 <- 0.01 # Threshold for adjusted p-values (FDR) for GENE DEA
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)

## Gene expression data:
data_for_lm <- data.frame(logcpm) %>% dplyr::select(-contains("TMZ"))
colnames(data_for_lm) <- c("Unt", "Unt", "Unt", "TT", "TT", "TT")
#colnames(data_for_lm) <- data3 %>% rownames() # Col names = samples
#rownames(data_for_lm) <- colnames(data3) #%>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') # Row names = Entrez gene IDs

### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- model.matrix(~0 + groups$group)[1:6,1:2]
colnames(design1) <- c("Unt","TT")
fit1 <- lmFit(data_for_lm, design1) 
contr1 <- makeContrasts(TTvsUnt = TT - Unt, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes

## Save list of ordered, differentially expressed genes:
#DEgenes$logFC %>% order() %>% DEgenes[.,] %>% write.csv("ordered_DEgenes_TGF+TNF.csv")

########## Various visualizations ########## 
## Gene LFC vs. average log-cpm:
#plotMD(tmp1, column=1, status=res1[,1], main=colnames(tmp1)[1], 
#       xlim=c(-8,13))

###################################### Volcano plot --> GENES ######################################
txtsz <- 14
linesz <- 1
allGenes <- mutate(allGenes, gene = allGenes$ID %>% match(genes$ensembl) %>% genes$gene[.])
DEAgenes.TT <- allGenes

ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1 %>% as.numeric()
        ),
  # color = "red",
    alpha = 0.2,
    cex = 2.5,
  ) +
   geom_text_repel(
     data = function(x){subset(x, ID %in% phillips_pn & adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "black"
   ) +
#     geom_text(
#     data = function(x){subset(x, ID %in% phillips_pn)},
#     position = "dodge",
#     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
#     max.overlaps = 25, max.iter = 5000, color = "darkred"
#   ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA genes (Phillips PN): TGFb+TNFa vs. Untreated", x = "Expression log2FC", y = "-log10(adj. p-value)"
  ) +
  guides(
    color = FALSE
  )
#ggsave("DEA-genes_TGF+TNF.png", width = 12, height = 8)
```

```{r}
adjPvalueCutoff1 <- 0.01 # Threshold for adjusted p-values (FDR) for GENE DEA
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)

## Gene expression data:
data_for_lm <- data.frame(logcpm) %>% dplyr::select(-contains("TT"))
colnames(data_for_lm) <- c("Unt", "Unt", "Unt", "TMZ", "TMZ", "TMZ")
#colnames(data_for_lm) <- data3 %>% rownames() # Col names = samples
#rownames(data_for_lm) <- colnames(data3) #%>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') # Row names = Entrez gene IDs

### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- model.matrix(~0 + groups$group)[1:6,1:2]
colnames(design1) <- c("Unt","TMZ")
fit1 <- lmFit(data_for_lm, design1) 
contr1 <- makeContrasts(TMZvsUnt = TMZ - Unt, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes

########## Various visualizations ########## 
## Gene LFC vs. average log-cpm:
#plotMD(tmp1, column=1, status=res1[,1], main=colnames(tmp1)[1], 
#       xlim=c(-8,13))

###################################### Volcano plot --> GENES ######################################
txtsz <- 14
linesz <- 1
allGenes <- mutate(allGenes, gene = allGenes$ID %>% match(genes$ensembl) %>% genes$gene[.])
DEAgenes.TT <- allGenes

ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1 %>% as.numeric()
        ),
  # color = "red",
    alpha = 0.2,
    cex = 2.5,
  ) +
   geom_text(
     data = function(x){subset(x, ID %in% phillips_pn & adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "black"
   ) +
#     geom_text(
#     data = function(x){subset(x, ID %in% pcEMT_e)},
#     position = "dodge",
#     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
#     max.overlaps = 25, max.iter = 5000, color = "darkred"
#   ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA genes (Phillips PN): TMZ vs. Untreated", x = "Expression log2FC", y = "-log10(adj. p-value)"
  ) +
  guides(
    color = FALSE
  )
#ggsave("DEA-genes_TGF+TNF.png", width = 12, height = 8)
```

##GSEA NES
```{r NES, eval=FALSE, include=FALSE}
library(grid)
##TGFb+TNFa
TT.degenes <- read_csv("ordered_DEgenes_TGF+TNF.csv")
NEFTEL_MES <- append(neftel.mes1, neftel.mes2) %>% as.integer()
NEFTEL_PN <- neftel.opc %>% as.integer()

TT.ranks <- TT.degenes$logFC
names(TT.ranks) <- TT.degenes$ID

sets.gsea <- append(gene_sets1, cgp.sig["VERHAAK_GLIOBLASTOMA_MESENCHYMAL"]) %>%
  append(cgp.sig["VERHAAK_GLIOBLASTOMA_PRONEURAL"])
sets.gsea[['NEFTEL_PN']] <- NEFTEL_PN
sets.gsea[['NEFTEL_MES']] <- NEFTEL_MES
sets.gsea[["PHILLIPS_MES"]] <- phillips_mes
sets.gsea[["PHILLIPS_PN"]] <- phillips_pn
sets.gsea[["SEGERMAN_PMT"]] <- segerman_mes
sets.gsea[["PCEMT_M"]] <- pcEMT_m
sets.gsea[["PCEMT_E"]] <- pcEMT_e

##GSEA for TGFb+TNFa samples
fgseaRes <- fgsea(pathways = sets.gsea, stats = TT.ranks, minSize=5, maxSize = 500, nperm=1000)

##HALLMARK TGFB
nes.score <- fgseaRes %>% filter(pathway == "HALLMARK_TGF_BETA_SIGNALING") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "HALLMARK_TGF_BETA_SIGNALING") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=10)))
plotEnrichment(sets.gsea$HALLMARK_TGF_BETA_SIGNALING, TT.ranks) +
  labs(title = bquote(paste("HALLMARK TGFb"))) +
  annotation_custom(annot)
# ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_HALLMARK-TGFb_TT.svg", width = 6, height = 3, units = "in")

##HALLMARK TNFA
nes.score <- fgseaRes %>% filter(pathway == "HALLMARK_TNFA_SIGNALING_VIA_NFKB") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "HALLMARK_TNFA_SIGNALING_VIA_NFKB") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=10)))
plotEnrichment(sets.gsea$HALLMARK_TNFA_SIGNALING_VIA_NFKB, TT.ranks) +
  labs(title = bquote(paste("HALLMARK TNFa"))) +
  annotation_custom(annot)
# ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_HALLMARK-TNFa_TT.svg", width = 6, height = 3, units = "in")

##VERHAAK MES
nes.score <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$VERHAAK_GLIOBLASTOMA_MESENCHYMAL, TT.ranks) +
  labs(title = bquote(paste("Verhaak MES - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##VERHAAK PN
nes.score <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_PRONEURAL") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_PRONEURAL") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$VERHAAK_GLIOBLASTOMA_PRONEURAL, TT.ranks) +
  labs(title = bquote(paste("Verhaak PN - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##HALLMARK EMT
nes.score <- fgseaRes %>% filter(pathway == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION, TT.ranks) +
  labs(title = bquote(paste("HALLMARK EMT - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##NEFTEL MES
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_MES") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "NEFTEL_MES") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$NEFTEL_MES, TT.ranks) +
  labs(title = bquote(paste("Neftel MES - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##NEFTEL PN
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% .$NES %>% round(2)
plotEnrichment(sets.gsea$NEFTEL_PN, TT.ranks) +
  labs(title = bquote(paste("Neftel PN - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##PHILLIPS MES
nes.score <- fgseaRes %>% filter(pathway == "PHILLIPS_MES") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PHILLIPS_MES") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PHILLIPS_MES, TT.ranks) +
  labs(title = bquote(paste("Phillips MES - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##PHILLIPS PN
nes.score <- fgseaRes %>% filter(pathway == "PHILLIPS_PN") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PHILLIPS_PN") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PHILLIPS_PN, TT.ranks) +
  labs(title = bquote(paste("Phillips PN - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##SEGERMAN PMT
nes.score <- fgseaRes %>% filter(pathway == "SEGERMAN_PMT") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "SEGERMAN_PMT") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$SEGERMAN_PMT, TT.ranks) +
  labs(title = bquote(paste("Segerman PMT - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##pcEMT Mes
nes.score <- fgseaRes %>% filter(pathway == "PCEMT_M") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PCEMT_M") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PCEMT_M, TT.ranks) +
  labs(title = bquote(paste("pcEMT-M - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

##pcEMT Ep
nes.score <- fgseaRes %>% filter(pathway == "PCEMT_E") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PCEMT_E") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PCEMT_E, TT.ranks) +
  labs(title = bquote(paste("pcEMT-E - TGF", beta, "+TNF", alpha))) +
  annotation_custom(annot)

#topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
#topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
#topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
#plotGseaTable(sets.gsea[topPathways], TT.ranks, fgseaRes, gseaParam = 0.5) 
#data.table::fwrite(fgseaRes, file="fgseaRes_TT.txt", sep="\t", sep2=c("", " ", ""))


##TMZ
TMZ.degenes <- read_csv("ordered_DEgenes_TMZ.csv")
TMZ.ranks <- TMZ.degenes$logFC
names(TMZ.ranks) <- TMZ.degenes$ID


fgseaRes <- fgsea(pathways = sets.gsea, stats = TMZ.ranks, minSize=15, maxSize = 500, nperm=1000)

##VERHAAK MES
nes.score <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$VERHAAK_GLIOBLASTOMA_MESENCHYMAL, TMZ.ranks) +
  labs(title = "Verhaak MES - TMZ") +
  annotation_custom(annot)

##VERHAAK PN
nes.score <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_PRONEURAL") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_PRONEURAL") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$VERHAAK_GLIOBLASTOMA_PRONEURAL, TMZ.ranks) +
  labs(title = "Verhaak PN - TMZ")+
  annotation_custom(annot)

#HALLMARK EMT
nes.score <- fgseaRes %>% filter(pathway == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION, TMZ.ranks) +
  labs(title = "Hallmark EMT - TMZ")+
  annotation_custom(annot)

##NEFTEL MES
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_MES") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "NEFTEL_MES") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$NEFTEL_MES, TMZ.ranks) +
  labs(title = "Neftel MES - TMZ")+
  annotation_custom(annot)

##NEFTEL PN
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% .$NES %>% round(2)
plotEnrichment(sets.gsea$NEFTEL_PN, TMZ.ranks) +
  labs(title = "Neftel PN - TMZ") +
  annotation_custom(annot)

##PHILLIPS MES
nes.score <- fgseaRes %>% filter(pathway == "PHILLIPS_MES") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PHILLIPS_MES") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PHILLIPS_MES, TMZ.ranks) +
  labs(title = "Phillips MES - TMZ") +
  annotation_custom(annot)

##PHILLIPS PN
nes.score <- fgseaRes %>% filter(pathway == "PHILLIPS_PN") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PHILLIPS_PN") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PHILLIPS_PN, TMZ.ranks) +
  labs(title = "Phillips PN - TMZ") +
  annotation_custom(annot)

##SEGERMAN PMT
nes.score <- fgseaRes %>% filter(pathway == "SEGERMAN_PMT") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "SEGERMAN_PMT") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$SEGERMAN_PMT, TMZ.ranks) +
  labs(title = "Segerman PMT - TMZ") +
  annotation_custom(annot)

##pcEMT - M
nes.score <- fgseaRes %>% filter(pathway == "PCEMT_M") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PCEMT_M") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PCEMT_M, TMZ.ranks) +
  labs(title = "pcEMT-M - TMZ") +
  annotation_custom(annot)

##pcEMT - E
nes.score <- fgseaRes %>% filter(pathway == "PCEMT_E") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PCEMT_E") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PCEMT_E, TMZ.ranks) +
  labs(title = "pcEMT-E - TMZ") +
  annotation_custom(annot)

##HALLMARK TGFB
nes.score <- fgseaRes %>% filter(pathway == "HALLMARK_TGF_BETA_SIGNALING") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "HALLMARK_TGF_BETA_SIGNALING") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=10)))
plotEnrichment(sets.gsea$HALLMARK_TGF_BETA_SIGNALING, TMZ.ranks) +
  labs(title = bquote(paste("HALLMARK TGFb"))) +
  annotation_custom(annot)
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_HALLMARK-TGFb_TMZ.svg", width = 6, height = 3, units = "in")

##HALLMARK TNFA
nes.score <- fgseaRes %>% filter(pathway == "HALLMARK_TNFA_SIGNALING_VIA_NFKB") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "HALLMARK_TNFA_SIGNALING_VIA_NFKB") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=10)))
plotEnrichment(sets.gsea$HALLMARK_TNFA_SIGNALING_VIA_NFKB, TMZ.ranks) +
  labs(title = bquote(paste("HALLMARK TNFa"))) +
  annotation_custom(annot)
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_HALLMARK-TNFa_TMZ.svg", width = 6, height = 3, units = "in")
```
###TGFb+TNFa
```{r}
library(DOSE)
library(enrichplot)
library(grid)

plot_theme <- theme(text = element_text(size = 7),
                    axis.text = element_text(size = 2),
                    axis.title = element_text(size = 10),
                    axis.text.x = element_text(size = 8),
                    axis.text.y = element_text(size = 8))
TT.degenes <- read_csv("ordered_DEgenes_TGF+TNF.csv")
# NEFTEL_MES <- append(neftel.mes1, neftel.mes2) %>% as.integer()
# NEFTEL_PN <- neftel.opc %>% as.integer()

sets.gsea1 <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CGP") %>% 
  dplyr::select(gs_name, entrez_gene) %>% filter(gs_name == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL" | gs_name == "VERHAAK_GLIOBLASTOMA_PRONEURAL")

phillips <- data.frame(gs_name = "PHILLIPS_MESENCHYMAL", entrez_gene = as.numeric(phillips_mes)) %>% 
  rbind(data.frame(gs_name = "PHILLIPS_PRONEURAL", entrez_gene = as.numeric(phillips_pn)))

phillips_sig <-  data.frame(gs_name = "PHILLIPS_PRONEURAL_SIG",
                           entrez_gene = as.numeric(mapIds(org.Hs.eg.db, phillips_pn_sig, 'ENTREZID', 'SYMBOL'))) %>% 
  rbind(data.frame(gs_name = "PHILLIPS_MESENCHYMAL_SIG",
                           entrez_gene = as.numeric(mapIds(org.Hs.eg.db, phillips_mes_sig, 'ENTREZID', 'SYMBOL')))) %>%
  filter(!is.na(entrez_gene))

  
sets.gsea1 <- rbind(sets.gsea1, phillips)
sets.gsea1 <- rbind(sets.gsea1, phillips_sig)

neftel <- data.frame(gs_name = "NEFTEL_MES1", entrez_gene = as.numeric(neftel.mes1)) %>% 
  rbind(data.frame(gs_name = "NEFTEL_MES2", entrez_gene = as.numeric(neftel.mes2))) %>% 
  rbind(data.frame(gs_name = "NEFTEL_OPC", entrez_gene = as.numeric(neftel.opc)))
sets.gsea1 <- rbind(sets.gsea1, neftel)

TT.ranks <- TT.degenes$logFC
names(TT.ranks) <- TT.degenes$ID
TT.ranks <- TT.ranks %>% sort(decreasing = T)

# edo2 <- gseDO(TT.ranks)

# fgseaRes <- fgsea(pathways = sets.gsea, stats = TT.ranks, minSize=5, maxSize = 500, nperm=1000)

em <- GSEA(TT.ranks, TERM2GENE = sets.gsea1, pvalueCutoff = 1)
head(em)


for(i in 1:length(em$Description)){
  print(
    gseaplot(em, geneSetID = i, by = "runningScore", 
             title = bquote(paste(.(em$Description[i]), ": ", "TGF", beta, "+TNF", alpha)))+
      annotation_custom(grobTree(textGrob(bquote(atop(paste("NES: ", .(em$NES[i] %>% round(2))), 
                                                      paste("p-value: ", .(em$pvalue[i] %>% round(4))))), 
                                          x=0.7,  y=0.8, hjust=0,
                                          gp=gpar(col="red", fontsize=8))))+
      plot_theme
  )
  save_loc <- paste0("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_", i, "_TT")
  ggsave(paste0(save_loc, ".svg"))
  ggsave(paste0(save_loc, ".png"), width = 4, height = 2, units = "in")
}


```

```{r unused code, eval=FALSE, include=FALSE}

###  1
gseaplot(em, geneSetID = 1, by = "runningScore", 
         title = bquote(paste(.(em$Description[1]), ": ", "TGF", beta, "+TNF", alpha)))+
  annotation_custom(grobTree(textGrob(bquote(atop(paste("NES: ", .(em$NES[1] %>% round(2))), 
                                       paste("p-value: ", .(em$pvalue[1] %>% round(4))))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=8))))+
  plot_theme
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_1_TT.svg")
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_1_TT.png", width = 4, height = 2, units = "in")

###  2
gseaplot(em, geneSetID = 2, by = "runningScore", 
         title = bquote(paste(.(em$Description[2]), ": ", "TGF", beta, "+TNF", alpha)))+
  annotation_custom(grobTree(textGrob(bquote(atop(paste("NES: ", .(em$NES[2] %>% round(2))), 
                                       paste("p-value: ", .(em$pvalue[2] %>% round(4))))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=8))))+
  plot_theme
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_2_TT.svg")
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_2_TT.png", width = 4, height = 2, units = "in")

###  3
gseaplot(em, geneSetID = 3, by = "runningScore", 
         title = bquote(paste(.(em$Description[3]), ": ", "TGF", beta, "+TNF", alpha)))+
  annotation_custom(grobTree(textGrob(bquote(atop(paste("NES: ", .(em$NES[3] %>% round(2))), 
                                       paste("p-value: ", .(em$pvalue[3] %>% round(4))))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=8))))+
  plot_theme
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_3_TT.svg")
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_3_TT.png", width = 4, height = 2, units = "in")

###  4
gseaplot(em, geneSetID = 4, by = "runningScore", 
         title = bquote(paste(.(em$Description[4]), ": ", "TGF", beta, "+TNF", alpha)))+
  annotation_custom(grobTree(textGrob(bquote(atop(paste("NES: ", .(em$NES[4] %>% round(2))), 
                                       paste("p-value: ", .(em$pvalue[4] %>% round(4))))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=8))))+
  plot_theme
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_4_TT.svg")
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_4_TT.png", width = 4, height = 2, units = "in")



```

```{r} 
# extra NES plots for EMT signatures

TT.degenes <- read_csv("ordered_DEgenes_TGF+TNF.csv")
# NEFTEL_MES <- append(neftel.mes1, neftel.mes2) %>% as.integer()
# NEFTEL_PN <- neftel.opc %>% as.integer()

gene.sets2 <- hall %>% dplyr::select(gs_name, entrez_gene) %>% filter(gs_name == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION")

pc_emt_sig <- data.frame(gs_name = "pcEMT-E", entrez_gene = pcEMT_e) %>% rbind(
  data.frame(gs_name = "pcEMT-M", entrez_gene = pcEMT_m)
)

gene.sets2 <- gene.sets2 %>% rbind(pc_emt_sig) %>% 
  rbind(cp.r %>% dplyr::select(gs_name, entrez_gene) %>% filter(gs_name == "REACTOME_TGF_BETA_RECEPTOR_SIGNALING_IN_EMT_EPITHELIAL_TO_MESENCHYMAL_TRANSITION_")) %>% 
  rbind(go.bp %>% dplyr::select(gs_name, entrez_gene) %>% filter(gs_name == "GO_EPITHELIAL_TO_MESENCHYMAL_TRANSITION")) %>% 
  rbind(go.bp %>% dplyr::select(gs_name, entrez_gene) %>% filter(gs_name == "GO_NEGATIVE_REGULATION_OF_EPITHELIAL_TO_MESENCHYMAL_TRANSITION"))

TT.ranks <- TT.degenes$logFC
names(TT.ranks) <- TT.degenes$ID
TT.ranks <- TT.ranks %>% sort(decreasing = T)

# edo2 <- gseDO(TT.ranks)

# fgseaRes <- fgsea(pathways = sets.gsea, stats = TT.ranks, minSize=5, maxSize = 500, nperm=1000)

em <- GSEA(TT.ranks, TERM2GENE = gene.sets2, pvalueCutoff = 1)
head(em)


for(i in 1:length(em$Description)){
print(
  gseaplot(em, geneSetID = i, by = "runningScore", 
         title = bquote(paste(.(em$Description[i]), ": ", "TGF", beta, "+TNF", alpha)))+
  annotation_custom(grobTree(textGrob(bquote(atop(paste("NES: ", .(em$NES[i] %>% round(2))), 
                                       paste("p-value: ", .(em$pvalue[i] %>% round(4))))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=8))))+
  plot_theme
)
}

```


###TMZ
```{r}
TMZ.degenes <- read_csv("ordered_DEgenes_TMZ.csv")
# NEFTEL_MES <- append(neftel.mes1, neftel.mes2) %>% as.integer()
# NEFTEL_PN <- neftel.opc %>% as.integer()

sets.gsea1 <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CGP") %>% 
  dplyr::select(gs_name, entrez_gene) %>% filter(gs_name == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL" | gs_name == "VERHAAK_GLIOBLASTOMA_PRONEURAL")

phillips <- data.frame(gs_name = "PHILLIPS_MESENCHYMAL", entrez_gene = as.numeric(phillips_mes)) %>%   rbind(data.frame(gs_name = "PHILLIPS_PRONEURAL", entrez_gene = as.numeric(phillips_pn)))
sets.gsea1 <- rbind(sets.gsea1, phillips)

phillips_sig <-  data.frame(gs_name = "PHILLIPS_PRONEURAL_SIG",
                           entrez_gene = as.numeric(mapIds(org.Hs.eg.db, phillips_pn_sig, 'ENTREZID', 'SYMBOL'))) %>% 
  rbind(data.frame(gs_name = "PHILLIPS_MESENCHYMAL_SIG",
                           entrez_gene = as.numeric(mapIds(org.Hs.eg.db, phillips_mes_sig, 'ENTREZID', 'SYMBOL')))) %>%
  filter(!is.na(entrez_gene))

sets.gsea1 <- rbind(sets.gsea1, phillips_sig)

neftel <- data.frame(gs_name = "NEFTEL_MES1", entrez_gene = as.numeric(neftel.mes1)) %>% 
  rbind(data.frame(gs_name = "NEFTEL_MES2", entrez_gene = as.numeric(neftel.mes2))) %>% 
  rbind(data.frame(gs_name = "NEFTEL_OPC", entrez_gene = as.numeric(neftel.opc)))
sets.gsea1 <- rbind(sets.gsea1, neftel)

TMZ.ranks <- TMZ.degenes$logFC
names(TMZ.ranks) <- TMZ.degenes$ID
TMZ.ranks <- TMZ.ranks %>% sort(decreasing = T)

# edo2 <- gseDO(TT.ranks)

# fgseaRes <- fgsea(pathways = sets.gsea, stats = TT.ranks, minSize=5, maxSize = 500, nperm=1000)

em <- GSEA(TMZ.ranks, TERM2GENE = sets.gsea1, pvalueCutoff = 1)
head(em)


for(i in 1:length(em$Description)){
  print(
    gseaplot(em, geneSetID = i, by = "runningScore", 
             title = bquote(paste(.(em$Description[i]), ": TMZ")))+
      annotation_custom(grobTree(textGrob(bquote(atop(paste("NES: ", .(em$NES[i] %>% round(2))), 
                                                      paste("p-value: ", .(em$pvalue[i] %>% round(4))))), 
                                          x=0.7,  y=0.8, hjust=0,
                                          gp=gpar(col="red", fontsize=8))))+
      plot_theme
  )
  save_loc <- paste0("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/NES_", i, "_TMZ")
  ggsave(paste0(save_loc, ".svg"))
  ggsave(paste0(save_loc, ".png"), width = 4, height = 2, units = "in")
}

```

```{r}
## TMZ - EMT signatures

em <- GSEA(TMZ.ranks, TERM2GENE = gene.sets2, pvalueCutoff = 1)
head(em)


for(i in 1:length(em$Description)){
print(
  gseaplot(em, geneSetID = i, by = "runningScore", 
         title = bquote(paste(.(em$Description[i]), ": ", "TMZ")))+
  annotation_custom(grobTree(textGrob(bquote(atop(paste("NES: ", .(em$NES[i] %>% round(2))), 
                                       paste("p-value: ", .(em$pvalue[i] %>% round(4))))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=8))))+
  plot_theme
)
}
```


##generating PLSR model
```{r}
library(plsVarSel)
genes.mes <- phillips_mes
genes.pn <- phillips_pn 

#genes.mes <- NEFTEL_MES
#genes.pn <- NEFTEL_PN 

#genes.mes <- verhaak.mes
#genes.pn <- verhaak.pn

y.mat <- logcpm  %>% data.frame() %>%
  rownames_to_column(var = "gene") %>%
  filter(gene %in% genes.mes | gene %in% genes.pn) %>% #change which genes to include
  column_to_rownames(var = "gene")
y.mat.var <- rowVars(data.matrix(y.mat))
y.mat.mean <- rowMeans(data.matrix(y.mat))
y.mat %<>% mutate(rowVar = y.mat.var/y.mat.mean)
y.mat %<>% slice_max(n=40, order_by = rowVar)
y.mat %<>% dplyr::select(-rowVar)

y.mat <- y.mat %>%  t() %>% data.frame() %>%
  rownames_to_column(var="treat") %>% group_by(treat)
y.mat[,1] <- samples
y.mat <- y.mat %>% group_by(treat) %>%
  summarise_all(funs(mean(.))) %>% arrange(treat)
rownames(y.mat) <- NULL
y.mat %<>% column_to_rownames(var = "treat") %>% data.matrix() %>% scale(scale=T, center = T)
colnames(y.mat) <- gsub("X", "", colnames(y.mat))

x.mat <- read_excel("G816_PMT_pY_blot.xlsx", sheet = "Sheet1") %>% arrange(Treat) %>%
  column_to_rownames(var = "Treat") %>%
  data.matrix()


mydata <- data.frame(yy=I(y.mat),xx=I(x.mat))
PLSR_matrix <- plsr(yy~xx, scale=T, data=mydata, method="kernelpls")

score.min <- 1
vip.scores <- VIP(PLSR_matrix, 2) %>% data.frame()
colnames(vip.scores)[1] <- "VIP.score"
vip.scores <- rownames_to_column(vip.scores, var="protein")
vip.scores <- vip.scores %>% filter(VIP.score>score.min)
proteins.VIP <- vip.scores$protein
#running PLSR again using only proteins w/ high VIP scores

x.mat <- data.matrix(dplyr::select(data.frame(x.mat), proteins.VIP))
mydata <- data.frame(yy=I(y.mat),xx=I(x.mat))
PLSR_matrix <- plsr(yy~xx, scale=T, data=mydata, method="kernelpls")

sc <- scores(PLSR_matrix)
ld <- loadings(PLSR_matrix)
Ysc <-Yscores(PLSR_matrix)
Yld <- Yloadings(PLSR_matrix)
Yld <- data.frame(matrix(as.numeric(Yld), attributes(Yld)$dim, dimnames=attributes(Yld)$dimnames)) %>% rownames_to_column(var="gene")
Yld.pn <- Yld %>% filter(gene %in% genes.pn)
Yld.mes <- Yld %>% filter(gene %in% genes.mes)

ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
#  geom_point(aes(x=ld[,1], y=ld[,2], size = (ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)])), alpha=0.5)+
#  geom_point(aes(x=Yld[,1], y=Yld[,2]), color="blue")+
  labs(title = "Phillips MES & PN markers",
       x=paste0("PC1", " (",explvar(PLSR_matrix)[1]%>%round(1), "%)"), 
       y=paste0("PC2", " (",explvar(PLSR_matrix)[2]%>%round(1), "%)"))+
  geom_text(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#555555",
                  size=3, alpha = 0.8)+
  geom_text(data = Yld.pn,
            aes(x=Yld.pn[,2], y=Yld.pn[,3]), 
            label=Yld.pn[,1] %>% match(genes$ensembl) %>% genes$gene[.], 
            size=3, color="darkred")+
  geom_text(data = Yld.mes,
            aes(x=Yld.mes[,2], y=Yld.mes[,3]), 
            label=Yld.mes[,1] %>% match(genes$ensembl) %>% genes$gene[.], 
            size=3, color="darkblue")+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "right"
  )
```

##PLSDA
```{r eval=FALSE, include=FALSE}
library(DiscriMiner)
#genes.mes <- phillips_mes
#genes.pn <- phillips_pn 

#genes.mes <- NEFTEL_MES %>% as.character()
#genes.pn <- NEFTEL_PN %>% as.character()

#genes.mes <- verhaak.mes
#genes.pn <- verhaak.pn

genes.mes <- c(phillips_mes, 
#               NEFTEL_MES %>% as.character(), 
               verhaak.mes)
genes.pn <- c(phillips_pn, 
#              NEFTEL_PN %>% as.character(), 
              verhaak.pn)

genes.sub <- c(genes.mes, genes.pn)

X <- logcpm %>% t() %>% scale(center = T, scale = T) %>% data.frame
colnames(X) <- gsub("X", "", colnames(X))
X <- X %>% dplyr::select(one_of(genes.sub))
colnames(X) %<>% match(genes$ensembl) %>% genes$gene[.]


Y <- samples
#Y <- data.frame()
#Y <-cbind(c("DMSO", "DMSO", "DMSO", "DMSO", "DMSO", "DMSO", "GP", "GP", "GP", "GP", "GP", "GP", "TMZ", "TMZ", "TMZ", "TMZ", "TMZ", "TMZ"), c("Ctrl", "Ctrl", "Ctrl", "SPRY2", "SPRY2", "SPRY2", "Ctrl", "Ctrl", "Ctrl", "SPRY2", "SPRY2", "SPRY2", "Ctrl", "Ctrl", "Ctrl", "SPRY2", "SPRY2", "SPRY2"))
#colnames(Y) <- c("Treat", "KD")

#Y <- factor(Y[,2])
Y <- factor(Y)

plsda.res <- plsDA(X, Y)
plot(plsda.res)
vip.scores <- plsda.res$VIP %>% data.frame()
genes.VIP <- slice_max(vip.scores, order_by = Model.VIP, prop = 0.5) %>% rownames()

#genes.VIP <- c("CHI3L1", "SERPINE1", "TIMP1", "DLL3", "KLRC3", "SCG3", "C20orf42", "NCAM1")
X <-  X %>% data.frame() %>% dplyr::select(one_of(genes.VIP))

plsda.res <- plsDA(X, Y)
#plot(plsda.res)

plsda.ld <- plsda.res$loadings %>% data.frame() %>% rownames_to_column(var = "gene") %>% mutate(subtype = case_when(
  gene %>% match(genes$gene) %>% genes$ensembl[.] %in% genes.mes  ~ "MES",
  gene %>% match(genes$gene) %>% genes$ensembl[.] %in% genes.pn  ~ "PN",
  T ~ "??"
))

ggplot()+
  geom_point(data = plsda.ld, aes(x = w.1, y = w.2, color = subtype))+
  geom_text_repel(data = subset(plsda.ld, gene %in% genes.VIP),
                  aes(x = w.1, y = w.2, label = gene, color = subtype), max.overlaps = 30, size=3)+
  geom_text(data = plsda.res$y.loadings, 
            aes (x = c1, y = c2, 
                 label = rownames(plsda.res$y.loadings)), 
            color = "black", size=4)+
  theme_cowplot()+
  theme(
    panel.background = element_rect(color = "black", fill = "white"),
    legend.position = "none"
    ) +
  geom_hline(yintercept = 0, alpha = 0.75, linetype = "dashed")+
  geom_vline(xintercept = 0, alpha = 0.75, linetype = "dashed")+
  labs(x = "PC1",
       y = "PC2", 
       title = "Phillips MES & PN genes - PLSDA")


```

##SF188 data
```{r SF188 TMZ sample analysis, eval=FALSE, include=FALSE}
sf188.logcpm <- read.csv("SF188_RNAseq_logcpm.csv")
dat.tmz <- sf188.logcpm %>% data.frame() %>% dplyr::select(-contains(c("spry", "GP"))) %>% 
  column_to_rownames(var="X")

## Statistical thresholds:
adjPvalueCutoff1 <- 0.01 # Threshold for adjusted p-values (FDR) for GENE DEA
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)

## Gene expression data:
data_for_lm <- dat.tmz
colnames(data_for_lm) <- c("DMSO", "DMSO", "DMSO", "TMZ", "TMZ", "TMZ")

### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- data.frame(DMSO = c(1,1,1,0,0,0), TMZ = c(0,0,0,1,1,1))
fit1 <- lmFit(data_for_lm, design1) 
contr1 <- makeContrasts(TMZvsDMSO = TMZ - DMSO, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes

DEgenes$logFC %>% order() %>% DEgenes[.,] %>% 
  mutate(gene = ID %>% match(genes$ensembl) %>% genes$gene[.]) %>%
  write.csv("ordered_DEgenes_SF188_TMZ.csv")


txtsz <- 14
linesz <- 1
allGenes <- mutate(allGenes, gene = allGenes$ID %>% match(genes$ensembl) %>% genes$gene[.])
DEAgenes.TMZ <- allGenes

ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = gene)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1 %>% as.numeric()
        ),
  # color = "red",
    alpha = 0,
    cex = 2.5,
  ) +
   geom_text(
     data = function(x){subset(x, ID %in% phillips_mes)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
     geom_text(
     data = function(x){subset(x, ID %in% phillips_pn)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkred"
   ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA genes (Phillips MES & PN): TMZ vs. DMSO", x = "Expression log2FC", y = "-log10(adj. p-value)"
  ) +
  guides(
    color = FALSE
  )
#ggsave("DEA-genes_TGF+TNF.png", width = 12, height = 8)
```

```{r SF188 NES, eval=FALSE, include=FALSE}

TMZ.degenes <- read_csv("ordered_DEgenes_SF188_TMZ.csv")

TMZ.ranks <- TMZ.degenes$logFC
names(TMZ.ranks) <- TMZ.degenes$ID

##GSEA for SF188 TMZ samples
fgseaRes <- fgsea(pathways = sets.gsea, stats = TMZ.ranks, minSize=5, maxSize = 500, nperm=1000)

##VERHAAK MES
nes.score <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$VERHAAK_GLIOBLASTOMA_MESENCHYMAL, TMZ.ranks) +
  labs(title = "Verhaak MES - TMZ") +
  annotation_custom(annot)

##VERHAAK PN
nes.score <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_PRONEURAL") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_PRONEURAL") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$VERHAAK_GLIOBLASTOMA_PRONEURAL, TMZ.ranks) +
  labs(title = "Verhaak PN - TMZ")+
  annotation_custom(annot)

#HALLMARK EMT
nes.score <- fgseaRes %>% filter(pathway == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION, TMZ.ranks) +
  labs(title = "Hallmark EMT - TMZ")+
  annotation_custom(annot)

##NEFTEL MES
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_MES") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "NEFTEL_MES") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$NEFTEL_MES, TMZ.ranks) +
  labs(title = "Neftel MES - TMZ")+
  annotation_custom(annot)

##NEFTEL PN
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% .$NES %>% round(2)
plotEnrichment(sets.gsea$NEFTEL_PN, TMZ.ranks) +
  labs(title = "Neftel PN - TMZ") +
  annotation_custom(annot)

##PHILLIPS MES
nes.score <- fgseaRes %>% filter(pathway == "PHILLIPS_MES") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PHILLIPS_MES") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PHILLIPS_MES, TMZ.ranks) +
  labs(title = "Phillips MES - TMZ") +
  annotation_custom(annot)

##PHILLIPS PN
nes.score <- fgseaRes %>% filter(pathway == "PHILLIPS_PN") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PHILLIPS_PN") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PHILLIPS_PN, TMZ.ranks) +
  labs(title = "Phillips PN - TMZ") +
  annotation_custom(annot)

##SEGERMAN PMT
nes.score <- fgseaRes %>% filter(pathway == "SEGERMAN_PMT") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "SEGERMAN_PMT") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$SEGERMAN_PMT, TMZ.ranks) +
  labs(title = "Segerman PMT - TMZ") +
  annotation_custom(annot)

##pcEMT - M
nes.score <- fgseaRes %>% filter(pathway == "PCEMT_M") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PCEMT_M") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PCEMT_M, TMZ.ranks) +
  labs(title = "pcEMT-M - TMZ") +
  annotation_custom(annot)

##pcEMT - E
nes.score <- fgseaRes %>% filter(pathway == "PCEMT_E") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PCEMT_E") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PCEMT_E, TMZ.ranks) +
  labs(title = "pcEMT-E - TMZ") +
  annotation_custom(annot)
```
##Tang et al data
```{r eval=FALSE, include=FALSE}
###data comes from https://doi.org/10.15252/embr.201948170

#ctr_dat1 <- read_excel("./GSE134595_RAW/GSM3957939_1123-vehicle-1.xlsx", sheet = "Sheet1")
#ctr_dat1$FPKM %<>% as.numeric(.)
#ctr_dat2 <- read_excel("./GSE134595_RAW/GSM3957940_1123-vehicle-2.xlsx", sheet = "Sheet1")
#tgf_dat1 <- read_excel("./GSE134595_RAW/GSM3957941_1123+TGF-beta-1.xlsx", sheet = "Sheet1")
#tgf_dat2 <- read_excel("./GSE134595_RAW/GSM3957942_1123+TGF-beta-2.xlsx", sheet = "Sheet1")

ctr_dat1 <- read_excel("./GSE134595_RAW/GSM3957943_528-vehicle-1.xlsx", sheet = "Sheet1")
ctr_dat2 <- read_excel("./GSE134595_RAW/GSM3957944_528-vehicle-2.xlsx", sheet = "Sheet1")
tgf_dat1 <- read_excel("./GSE134595_RAW/GSM3957945_528+TGF-beta-1.xlsx", sheet = "Sheet1")
tgf_dat2 <- read_excel("./GSE134595_RAW/GSM3957946_528+TGF-beta-2.xlsx", sheet = "Sheet1")

colnames(ctr_dat1) <- c("ID", "gene", "ctr1")
colnames(ctr_dat2) <- c("ID", "gene", "ctr2")
colnames(tgf_dat1) <- c("ID", "gene", "tgf1")
colnames(tgf_dat2) <- c("ID", "gene", "tgf2")

g528_dat <- full_join(ctr_dat1, ctr_dat2, by = c("gene", "ID")) %>% 
  full_join(tgf_dat1, by = c("gene", "ID")) %>% full_join(tgf_dat2, by = c("gene", "ID"))

g528_dat$ID <-  sub("ENSG[0]*", "", g528_dat$ID)
```

```{r eval=FALSE, include=FALSE}
## Statistical thresholds:
adjPvalueCutoff1 <- 0.05 # Threshold for adjusted p-values (FDR) for GENE DEA
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)

## Gene expression data:
data_for_lm <- g528_dat %>% dplyr::select(-ID) %>% distinct(gene, .keep_all = T) %>% 
  column_to_rownames(var = "gene") %>% filter(is.numeric(ctr1))

### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- data.frame(Ctr = c(1,1,0,0), TGF = c(0,0,1,1))
fit1 <- lmFit(data_for_lm, design1) 
contr1 <- makeContrasts(CTRvsTGF = Ctr - TGF, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes

DEgenes$logFC %>% order() %>% DEgenes[.,] %>% 
  write.csv("ordered_DEgenes_G528_TGFb.csv")


txtsz <- 14
linesz <- 1
allGenes <- mutate(allGenes, gene = allGenes$ID %>% match(genes$gene) %>% genes$ensembl[.])
DEAgenes.TGF <- allGenes

#ggplot(data = allGenes, 
ggplot(data = subset(allGenes, gene %in% c(verhaak.mes, verhaak.pn)),
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1 %>% as.numeric()
        ),
  # color = "red",
    alpha = 0,
    cex = 2.5,
  ) +
   geom_text(
     data = function(x){subset(x, gene %in% verhaak.mes)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkblue"
   ) +
     geom_text(
     data = function(x){subset(x, gene %in% verhaak.pn)},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 25, max.iter = 5000, color = "darkred"
   ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + cust.theme(ticks = T, lnsz = linesz) +
  labs(
    title = "DEA genes (Verhaak MES & PN): Ctrl vs. TGFb", x = "Expression log2FC", y = "-log10(adj. p-value)"
  ) +
  guides(
    color = FALSE
  )+
  #scale_y_continuous(limits = c(1,2.5))+
  scale_x_continuous(limits = c(-100,500))


```

```{r G528 NES, eval=FALSE, include=FALSE}

TGF.degenes <- read_csv("ordered_DEgenes_G528_TGFb.csv")

TGF.ranks <- TGF.degenes$logFC
names(TGF.ranks) <- TGF.degenes$ID

##GSEA for G528 TGF samples
fgseaRes <- fgsea(pathways = sets.gsea, stats = TGF.ranks, minSize=5, maxSize = 500, nperm=1000)

##VERHAAK MES
nes.score <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$VERHAAK_GLIOBLASTOMA_MESENCHYMAL, TGF.ranks) +
  labs(title = "Verhaak MES - TGF") +
  annotation_custom(annot)

##VERHAAK PN
nes.score <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_PRONEURAL") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "VERHAAK_GLIOBLASTOMA_PRONEURAL") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$VERHAAK_GLIOBLASTOMA_PRONEURAL, TGF.ranks) +
  labs(title = "Verhaak PN - TGF")+
  annotation_custom(annot)

#HALLMARK EMT
nes.score <- fgseaRes %>% filter(pathway == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION, TGF.ranks) +
  labs(title = "Hallmark EMT - TGF")+
  annotation_custom(annot)

##NEFTEL MES
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_MES") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "NEFTEL_MES") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$NEFTEL_MES, TGF.ranks) +
  labs(title = "Neftel MES - TGF")+
  annotation_custom(annot)

##NEFTEL PN
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
nes.score <- fgseaRes %>% filter(pathway == "NEFTEL_PN") %>% .$NES %>% round(2)
plotEnrichment(sets.gsea$NEFTEL_PN, TGF.ranks) +
  labs(title = "Neftel PN - TGF") +
  annotation_custom(annot)

##PHILLIPS MES
nes.score <- fgseaRes %>% filter(pathway == "PHILLIPS_MES") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PHILLIPS_MES") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PHILLIPS_MES, TGF.ranks) +
  labs(title = "Phillips MES - TGF") +
  annotation_custom(annot)

##PHILLIPS PN
nes.score <- fgseaRes %>% filter(pathway == "PHILLIPS_PN") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PHILLIPS_PN") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PHILLIPS_PN, TGF.ranks) +
  labs(title = "Phillips PN - TGF") +
  annotation_custom(annot)

##SEGERMAN PMT
nes.score <- fgseaRes %>% filter(pathway == "SEGERMAN_PMT") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "SEGERMAN_PMT") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$SEGERMAN_PMT, TGF.ranks) +
  labs(title = "Segerman PMT - TGF") +
  annotation_custom(annot)

##pcEMT - M
nes.score <- fgseaRes %>% filter(pathway == "PCEMT_M") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PCEMT_M") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PCEMT_M, TGF.ranks) +
  labs(title = "pcEMT-M - TGF") +
  annotation_custom(annot)

##pcEMT - E
nes.score <- fgseaRes %>% filter(pathway == "PCEMT_E") %>% 
  .$NES %>% round(2)
pval <- fgseaRes %>% filter(pathway == "PCEMT_E") %>% 
  .$pval %>% round(4)
annot <- grobTree(textGrob(bquote(atop(paste("NES: ", .(nes.score)), 
                                       paste("p-value: ", .(pval)))), 
                           x=0.7,  y=0.8, hjust=0,
  gp=gpar(col="red", fontsize=13)))
plotEnrichment(sets.gsea$PCEMT_E, TGF.ranks) +
  labs(title = "pcEMT-E - TGF") +
  annotation_custom(annot)

```
###GSVA
```{r , eval=FALSE, include=FALSE}
data_for_gsva <- g528_dat %>% dplyr::select(-gene) %>% distinct(ID, .keep_all = T) %>% 
  column_to_rownames(var = "ID") %>% data.matrix()
#rownames(data_for_gsva) <- exp.symbol$GeneID

force_calculate_gsva <- F
if(!file.exists("g528_gsva_results.rds") | isTRUE(force_calculate_gsva)){
  gsva_res <- gsva(
    data_for_gsva, 
    gene_sets_all, verbose = F, # Change which gene sets are run
    # gene_sets4, verbose = T
    method = "gsva", 
#    kcdf = "Gaussian",
    # abs.ranking = F,
    # mx.diff = T,
    min.sz = 10, # Minimum number of genes required to include a gene set
    parallel.sz=detectCores()#-1
    )
  ## Save GSVA results for speed:
  saveRDS(gsva_res, file = "g528_gsva_results.rds")
  
} else if(file.exists("g528_gsva_results.rds") & !exists("gsva_res")){ 
  ## Load GSVA results, if they haven't been loaded/calculated already:
  gsva_res <- readRDS("g528_gsva_results.rds") 
}
## Build data frame versions of GSVA results:
gsva_res.df <- gsva_res %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
gsva_res.df.t <- gsva_res %>% t() %>% data.frame(ID=rownames(.), .)
rownames(gsva_res.df) <- gsva_res.df$gene.set
```

```{r plot GSVA results for TGF vs Unt samples, eval=FALSE, include=FALSE}
gsva.res <- gsva_res.df %>% dplyr::select(contains("ctr")|contains("tgf"))

### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- data.frame(ctr = c(1,1,0,0), TGF = c(0,0,1,1))
colnames(design1) <- c("ctr","tgf")
fit1 <- lmFit(gsva.res, design1) 
contr1 <- makeContrasts(TGFvsCTR = tgf - ctr, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes

DEgenes$logFC %>% order() %>% DEgenes[.,] %>% write.csv("DEgene_sets_G528_TGF.csv")

logFCcutoff1 <- 0.5 # Threshold for log2(fold-change)
adjPvalueCutoff <- 0.05 # Threshold for adjusted p-values (FDR) - use for GO gene set


#volcano plot for differentially expression gene sets
ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff),
                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
  geom_text_repel(
#    data = function(x){subset(x, adj.P.Val < 10^-4 & logFC<logFCcutoff1)},
    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff)},
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff & ID %in% gene_sets_to_plot)},
    size = 2, alpha = 0.9,
    segment.size = 0.25, segment.alpha = .8, force=30, max.overlaps = 50
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 25),
    plot.title = element_text(size = 23),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "Gene Sets: Untreated vs. TGFb",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits=c(0,5))+  
  scale_x_continuous(limits=c(-1, 0))

#ggsave("GSVA_DE_TGF+TNF-neg.png", width = 16, height = 12)

sets.sub <- append(cp.p$gs_name, cp.b$gs_name) #these gene sets are useful for actual pathways
#sets.sub <- append(go.tft$gs_name, go.tft_leg$gs_name)

allGenes %>% filter(ID %in% sets.sub)
sets.tgf.up <- top_n(allGenes, 20, logFC)
sets.tgf.dn <- top_n(allGenes, -20, logFC)

sets.tgf.up %>%
  arrange(logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Up-regulated Gene Sets: TGF-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
#ggsave("gsva_up_TT_.png", width = 15, height = 6)

sets.tgf.dn %>%
  arrange(-logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y =ID, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  labs(
    title = "Down-regulated Gene Sets: TGF-treated",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = element_blank())+
  theme_cowplot()
#ggsave("gsva_dn_TT.png", width = 15, height = 6)
```

#barplots to look at individual gene expression
```{r}
library(ggstatsplot)
dat <- counts.symbol %>% 
  rownames_to_column(var = "gene") %>% 
  filter(gene %in% c("FN1", "CD44", "CHI3L1", "SERPINE1", "COL1A2", "OLIG2")) %>% 
  pivot_longer(cols = -c(gene), names_to = "treat", values_to = "count") %>% 
  group_by(gene) %>% 
  mutate(FC = count/max(count))

# dat <- logcpm %>% data.frame() %>% 
#   rownames_to_column(var = "entrez") %>% 
#   mutate(gene = .$entrez %>% match(genes$ensembl) %>% genes$gene[.]) %>% 
#   filter(gene %in% c("FN1", "CD44", "CHI3L1", "SERPINE1", "COL1A2", "OLIG2"))
#   
# dat1 <- dat %>% pivot_longer(cols = -c(gene, entrez), names_to = "treat", values_to = "logcpm") %>% 
#   group_by(gene) %>% 
#   mutate(FC = logcpm/max(logcpm))


dat$treat  <- dat$treat %>% gsub(".1|.2", "", .)
dat$treat <- dat$treat %>% factor(c("Unt", "TT", "TMZ"))
dat$gene <- dat$gene %>% factor(c("FN1", "CD44", "SERPINE1", "COL1A2", "OLIG2"))


dat %>% 
  ggbarplot(x = "gene", y = "FC", add = "mean_se", fill = "treat", position = position_dodge(0.8))+
  labs(x = element_blank(), y = "Normalized Gene Expression")+
  theme(legend.title = element_blank(), axis.text.x = element_text(face = "italic"))+
  scale_y_continuous(expand = c(0,0))

```


# Verhaak gene set OA
## mes
```{r fig.width=10, fig.height=6}


alpha <- 0.05
p.cutoff <- 0.05 # P-value cutoff
txt.sz <- 6 # Text size for plots
num_cats <- 20 # Number of gene sets to show in dotplots
######### KEGG:
set1 <- verhaak.mes

kegg_or <- enrichKEGG(gene = set1,
                        universe = as.character(genes$ensembl),
                        organism = "hsa", keyType = "kegg", 
                        pAdjustMethod = "BH", 
                        pvalueCutoff = p.cutoff, 
                        qvalueCutoff = 0.05
                        )
p4 <- dotplot(kegg_or,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "KEGG over-representation\n Verhaak Mesenchymal genes")
p4
######## BIOCARTA:
biocarta_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=cp.b %>% dplyr::select(gs_name, entrez_gene)
                          )
p5 <- dotplot(biocarta_or,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"
  #                    ) +
  labs(size = "Gene count", x = "Gene Ratio", title = "BIOCARTA over-representation\n Verhaak Mesenchymal genes")
p5
######### REACTOME:
reactome_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                          TERM2GENE=cp.r %>% dplyr::select(gs_name, entrez_gene))
p6 <- dotplot(reactome_or, 
        font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"     
  #                    ) + 
  labs(size = "Gene count", x = "Gene Ratio", title = "REACTOME over-representation\n Verhaak Mesenchymal genes")
p6
######### PID:
pid_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                     TERM2GENE=cp.p %>% dplyr::select(gs_name, entrez_gene))
p7 <- dotplot(pid_or, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x" 
  #                    ) +   
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "PID over-representation\n Verhaak Mesenchymal genes")
p7
######### Hallmark:
hallmark_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=hall %>% dplyr::select(gs_name, entrez_gene))
p8 <- dotplot(hallmark_or, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + # USE VIRIDIS COLOR THEME
  # geom_label_repel(aes(label = Description, point.size = Count),  
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"  
  #                    ) +  
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "HALLMARK over-representation\n Verhaak Mesenchymal genes")
p8

pg1 <- plot_grid(p4,p6,p7,p8, align = "h",labels = "AUTO")
pg1



```
##pn
```{r fig.width=10, fig.height=6}

alpha <- 0.05
p.cutoff <- 0.5 # P-value cutoff
txt.sz <- 6 # Text size for plots
num_cats <- 20 # Number of gene sets to show in dotplots
######### KEGG:
set1 <- verhaak.pn

kegg_or <- enrichKEGG(gene = set1,
                        universe = as.character(genes$ensembl),
                        organism = "hsa", keyType = "kegg", 
                        pAdjustMethod = "BH", 
                        pvalueCutoff = p.cutoff, 
                        qvalueCutoff = 0.05
                        )
p4 <- dotplot(kegg_or,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "KEGG over-representation\nVerhaak Proneural genes")
p4
######## BIOCARTA:
biocarta_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=cp.b %>% dplyr::select(gs_name, entrez_gene)
                          )
p5 <- dotplot(biocarta_or,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"
  #                    ) +
  labs(size = "Gene count", x = "Gene Ratio", title = "BIOCARTA over-representation\nVerhaak Proneural genes")
# p5
######### REACTOME:
reactome_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                          TERM2GENE=cp.r %>% dplyr::select(gs_name, entrez_gene))
p6 <- dotplot(reactome_or, 
        font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"     
  #                    ) + 
  labs(size = "Gene count", x = "Gene Ratio", title = "REACTOME over-representation\nVerhaak Proneural genes")
p6
######### PID:
pid_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                     TERM2GENE=cp.p %>% dplyr::select(gs_name, entrez_gene))
p7 <- dotplot(pid_or, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x" 
  #                    ) +   
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "PID over-representation\nVerhaak Proneural genes")
# p7
######### Hallmark:
hallmark_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=hall %>% dplyr::select(gs_name, entrez_gene))
p8 <- dotplot(hallmark_or, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + # USE VIRIDIS COLOR THEME
  # geom_label_repel(aes(label = Description, point.size = Count),  
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"  
  #                    ) +  
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "HALLMARK over-representation\nVerhaak Proneural genes")
p8

pg1 <- plot_grid(p4,p6,p8, align = "h",labels = "AUTO")
pg1



```
#phillips gene set OA
##mes
```{r fig.width=10, fig.height=6}

alpha <- 0.05
p.cutoff <- 0.05 # P-value cutoff
txt.sz <- 6 # Text size for plots
num_cats <- 20 # Number of gene sets to show in dotplots
######### KEGG:
set1 <- phillips_mes

kegg_or <- enrichKEGG(gene = set1,
                        universe = as.character(genes$ensembl),
                        organism = "hsa", keyType = "kegg", 
                        pAdjustMethod = "BH", 
                        pvalueCutoff = p.cutoff, 
                        qvalueCutoff = 0.05
                        )
p4 <- dotplot(kegg_or,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "KEGG over-representation\nPhillips Mesenchymal genes")
p4
######## BIOCARTA:
biocarta_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=cp.b %>% dplyr::select(gs_name, entrez_gene)
                          )
p5 <- dotplot(biocarta_or,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"
  #                    ) +
  labs(size = "Gene count", x = "Gene Ratio", title = "BIOCARTA over-representation\nPhillips Mesenchymal genes")
p5
######### REACTOME:
reactome_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                          TERM2GENE=cp.r %>% dplyr::select(gs_name, entrez_gene))
p6 <- dotplot(reactome_or, 
        font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"     
  #                    ) + 
  labs(size = "Gene count", x = "Gene Ratio", title = "REACTOME over-representation\nPhillips Mesenchymal genes")
p6
######### PID:
pid_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                     TERM2GENE=cp.p %>% dplyr::select(gs_name, entrez_gene))
p7 <- dotplot(pid_or, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x" 
  #                    ) +   
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "PID over-representation\nPhillips Mesenchymal genes")
p7
######### Hallmark:
hallmark_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=hall %>% dplyr::select(gs_name, entrez_gene))
p8 <- dotplot(hallmark_or, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + # USE VIRIDIS COLOR THEME
  # geom_label_repel(aes(label = Description, point.size = Count),  
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"  
  #                    ) +  
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "HALLMARK over-representation\nPhillips Mesenchymal genes")
p8

pg1 <- plot_grid(p4,p5,p6,p7,p8, align = "h",labels = "AUTO")
pg1

```
##pn
```{r fig.width=10, fig.height=6}

alpha <- 0.05
p.cutoff <- 0.5 # P-value cutoff
txt.sz <- 6 # Text size for plots
num_cats <- 20 # Number of gene sets to show in dotplots
######### KEGG:
set1 <- phillips_pn

kegg_or <- enrichKEGG(gene = set1,
                        universe = as.character(genes$ensembl),
                        organism = "hsa", keyType = "kegg", 
                        pAdjustMethod = "BH", 
                        pvalueCutoff = p.cutoff, 
                        qvalueCutoff = 0.05
                        )
p4 <- dotplot(kegg_or,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  labs(size = "Gene count", x = "Gene Ratio", title = "KEGG over-representation\nPhillips Proneural genes")
p4
######## BIOCARTA:
biocarta_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=cp.b %>% dplyr::select(gs_name, entrez_gene)
                          )
p5 <- dotplot(biocarta_or,
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"
  #                    ) +
  labs(size = "Gene count", x = "Gene Ratio", title = "BIOCARTA over-representation\nPhillips Proneural genes")
# p5
######### REACTOME:
reactome_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                          TERM2GENE=cp.r %>% dplyr::select(gs_name, entrez_gene))
p6 <- dotplot(reactome_or, 
        font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"     
  #                    ) + 
  labs(size = "Gene count", x = "Gene Ratio", title = "REACTOME over-representation\nPhillips Proneural genes")
p6
######### PID:
pid_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff,
                     TERM2GENE=cp.p %>% dplyr::select(gs_name, entrez_gene))
p7 <- dotplot(pid_or, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # geom_label_repel(aes(label = Description, point.size = Count),   
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x" 
  #                    ) +   
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "PID over-representation\nPhillips Proneural genes")
# p7
######### Hallmark:
hallmark_or <- enricher(set1, pAdjustMethod = "BH", pvalueCutoff = p.cutoff, 
                          TERM2GENE=hall %>% dplyr::select(gs_name, entrez_gene))
p8 <- dotplot(hallmark_or, 
        # font.size = 8,
        showCategory = num_cats) +
  geom_segment(aes(xend=0, yend = Description), size = 1) +
  # scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + # USE VIRIDIS COLOR THEME
  # geom_label_repel(aes(label = Description, point.size = Count),  
  #                    size = txt.sz/2, color = "black", min.segment.length = 0, direction = "x"  
  #                    ) +  
  labs(size = "Gene count", x = "Gene Ratio", x = "Gene Ratio", title = "HALLMARK over-representation\nPhillips Proneural genes")
p8

pg1 <- plot_grid(p4,p6, align = "h",labels = "AUTO")
pg1



```


#upset plots
##verhaak vs hallmarks
```{r}
#generating upset plots for Verhaak, Phillips, and Hallmark TGFb, TNFa, Hypoxia gene sets
library(ggupset)

upset_paths <- c("VERHAAK_GLIOBLASTOMA_MESENCHYMAL", "VERHAAK_GLIOBLASTOMA_PRONEURAL", "HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_HYPOXIA")

pathway_genes <- msigdbr(species = "Homo sapiens") %>% 
  dplyr::select(gs_name, gene_symbol) %>% filter(gs_name %in% upset_paths)

phillips_mes_sig <- read_excel("mmc4.xls", sheet = "MES") %>% filter(`Signature gene` == "x") %>% distinct(`Gene Symbol`) %>% pull(`Gene Symbol`) 

phillips_pn_sig <- read_excel("mmc4.xls", sheet = "PN") %>% filter(`Signature gene` == "x") %>% distinct(`Gene Symbol`) %>% pull(`Gene Symbol`) %>% gsub("C20orf42", "FERMT1", .) #have to change the C20orf42 to its gene name for correct upset plot

# neftel_mes1 <- gene_sets5 %>% pull(MESlike1)
# neftel_mes2 <- gene_sets5 %>% pull(MESlike2)
# neftel_opc <- gene_sets5 %>% pull(OPClike)

# pathway_genes <- pathway_genes %>% 
#   rbind(data.frame(gs_name = "Phillips MES", gene_symbol = unlist(phillips_mes_sig))) %>% 
#   rbind(data.frame(gs_name = "Phillips PN", gene_symbol = unlist(phillips_pn_sig))) %>% 
  # rbind(data.frame(gs_name = "Neftel MES-like 1", gene_symbol = unlist(neftel_mes1))) %>% 
  # rbind(data.frame(gs_name = "Neftel MES-like 2", gene_symbol = unlist(neftel_mes2))) %>% 
  # rbind(data.frame(gs_name = "Neftel OPC-like", gene_symbol = unlist(neftel_opc))) 

pathway_genes <- pathway_genes %>% group_by(gs_name) %>% distinct(gene_symbol)

pathway_member <- pathway_genes %>% group_by(gene_symbol) %>% summarise(gs_name = list(gs_name))


upset1 <- pathway_member %>% ggplot(aes(x = gs_name))+
  geom_bar()+
  scale_x_upset(order_by = "degree")+
  geom_text(stat = 'count', aes(label = after_stat(count)), vjust = -1)+
  scale_y_continuous(limits = c(0, 210))+
  labs(x = element_blank(), y = "Unique Genes")
upset1
ggsave("upset-hallmark-verhaak.svg", plot = upset1, width = 7, height = 4, units = "in")
ggsave("upset-hallmark-verhaak.png", plot = upset1, width = 7, height = 4, units = "in")
```
##GBM subtyping systems
```{r}
#generating upset plot of Verhaak, Phillips, and Neftel PN/MES gene sets
upset_paths <- c("VERHAAK_GLIOBLASTOMA_MESENCHYMAL", "VERHAAK_GLIOBLASTOMA_PRONEURAL")

pathway_genes <- msigdbr(species = "Homo sapiens") %>% 
  dplyr::select(gs_name, gene_symbol) %>% filter(gs_name %in% upset_paths)

phillips_mes_sig <- read_excel("mmc4.xls", sheet = "MES") %>% filter(`Signature gene` == "x") %>% distinct(`Gene Symbol`) %>% pull(`Gene Symbol`) 

phillips_pn_sig <- read_excel("mmc4.xls", sheet = "PN") %>% filter(`Signature gene` == "x") %>% distinct(`Gene Symbol`) %>% pull(`Gene Symbol`) %>% gsub("C20orf42", "FERMT1", .) #have to change the C20orf42 to its gene name for correct upset plot

neftel_mes1 <- gene_sets5 %>% pull(MESlike1)
neftel_mes2 <- gene_sets5 %>% pull(MESlike2)
neftel_opc <- gene_sets5 %>% pull(OPClike)

pathway_genes <- pathway_genes %>% 
  rbind(data.frame(gs_name = "Phillips MES", gene_symbol = unlist(phillips_mes_sig))) %>% 
  rbind(data.frame(gs_name = "Phillips PN", gene_symbol = unlist(phillips_pn_sig))) %>% 
  rbind(data.frame(gs_name = "Neftel MES-like 1", gene_symbol = unlist(neftel_mes1))) %>% 
  rbind(data.frame(gs_name = "Neftel MES-like 2", gene_symbol = unlist(neftel_mes2))) %>% 
  rbind(data.frame(gs_name = "Neftel OPC-like", gene_symbol = unlist(neftel_opc))) 

pathway_genes <- pathway_genes %>% group_by(gs_name) %>% distinct(gene_symbol)

pathway_member <- pathway_genes %>% group_by(gene_symbol) %>% summarise(gs_name = list(gs_name))


upset2 <- pathway_member %>% ggplot(aes(x = gs_name))+
  geom_bar()+
  scale_x_upset(order_by = "degree")+
  geom_text(stat = 'count', aes(label = after_stat(count)), vjust = -1)+
  scale_y_continuous(limits = c(0, 220))+
  labs(x = element_blank(), y = "Unique Genes")
upset2
ggsave("upset-pmt-sets.svg", width = 7, height = 4, units = "in", plot = upset2)
ggsave("upset-pmt-sets.png", width = 7, height = 4, units = "in", plot = upset2)

semi_join(
  data.frame(dplyr::select(gene_sets5, MESlike1)),
  data.frame(dplyr::select(gene_sets5, MESlike2)),
  by = c("MESlike1" = "MESlike2")
)

```

