---
title: "R Notebook"
output: html_notebook
---

data comes from https://doi.org/10.1016/j.cell.2019.06.024, downloaded from BROAD institute ssRNA-seq portal
#load packages
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(cowplot)
library(plotly)
library(RColorBrewer)
library(colorspace)
library(jcolors)
library(ggsci)
library(Hmisc)
library(stats)
library(NMF)
library(ggfortify)
library(ggrepel)
library(GSVA)
library(GSVAdata)
library(GSEABase)
library(limma)
library(msigdbr)
library(gage)
library(GO.db)
library(heatmaply)
#library(plsVarSel)
#library(scran)
#library(scater)
#library(SC3)
#library(SingleCellExperiment)
#library(singleCellTK)
library(magrittr)
library(BiocManager)
library(paletteer)
library(ggbeeswarm)
library(ggstatsplot)
library(pheatmap)
library(parallel)
library(org.Hs.eg.db)
library(umap)
library(M3C)
library(biomaRt)
library(caret)
library(ConsensusClusterPlus)
```

```{r Clear environment}
### Clear R workspace:
rm(list=ls())
```

```{r Check and set working directory}
### Check current working directory:
cwd <- getwd()
cwd
### Set working directory:
setwd(cwd) # Set current working directory to location of this notebook for easier handling and saving of data associated with the notebook
getwd()
```
#define functions
```{r}
theme_cust <- function(ticks, axis, box, lnsz, color){
  if(missing(color)){color = "black"}
  if(missing(ticks)){ticks <- TRUE}
  if(missing(axis)){axis <- "y"}
  if(missing(box)){box <- TRUE}
  if(missing(lnsz)){lnsz <- 1}
  if(ticks == FALSE & axis == "y" ){
    theme_out <- theme(
      axis.text.y = element_blank(), axis.ticks.y = element_blank(), # Make y axis ticks blank
      axis.line.y = element_blank(), axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = color, fill = "white"),
      )
  } else if(ticks == FALSE & axis == "x"){
    theme_out <- theme(
      axis.text.x = element_blank(), axis.ticks.x = element_blank(), # Make x axis ticks blank
      axis.line.y = element_blank(), axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = color, fill = "white"),
      )
  } else if(ticks == FALSE & axis == "both"){
    theme_out <- theme(
      axis.text.y = element_blank(), axis.ticks.y = element_blank(), # Make y axis ticks blank
      axis.text.x = element_blank(), axis.ticks.x = element_blank(), # Make x axis ticks blank
      axis.line.y = element_blank(), axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = color, fill = "white"),
      )
  } else {
     theme_out <- theme(
       axis.line.y = element_blank(), axis.line.x = element_blank(),
       axis.ticks = element_line(color=color),
       panel.background = element_rect(size = lnsz, color = color, fill = "white"),
       ) 
  }
  
  if(box != TRUE) {
    theme_out <- theme_out + theme(panel.background = element_blank(),
                                   axis.line.y = element_line(size = lnsz, color = color),
                                   axis.line.x = element_line(size = lnsz, color = color)
                                   )
  } else {
    theme_out <- theme_out
  }
}


#not in %ni%
`%ni%` <- Negate(`%in%`)

```
#import data
```{r import data set, message=FALSE, warning=FALSE}
scRNA.import <- read_tsv("IDHwtGBM.processed.SS2.logTPM.txt") #these are log2(TPM/10+1) data
subtypes.data <- read_tsv("IDHwt.GBM.Hierarchy.SS2.txt")[-1,]
cell.info <- read_tsv("IDHwt.GBM.Metadata.SS2.txt")
```
#clean up data
```{r data clean-up and formatting}
sc.data <- scRNA.import %>%
  column_to_rownames(var="GENE")
sc.genes <- pull(scRNA.import, GENE)
sc.samples <- colnames(sc.data)

sc.samples.info <- cell.info[-1,]
sc.samples.info <- filter(sc.samples.info, CellAssignment == "Malignant", GBMType == "Adult")
samples.mal <- pull(sc.samples.info, NAME)
sc.data <- dplyr::select(sc.data, all_of(samples.mal))

sc.samples.info <- sc.samples.info %>% .[order(match(.$NAME, colnames(sc.data))),] %>%
  #select(., c(CellAssignment, NAME)) %>% 
  column_to_rownames(var = "NAME")

## Convert gene names to Entrez IDs:
genes <- rownames(sc.data) %>% 
  mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL', multiVals = "first") %>% 
  data.frame() %>%
  rownames_to_column(var = "GeneID")
colnames(genes)[2] <- "ensemblID"
genes <- genes %>% filter(!is.na(ensemblID))
sc.data <- sc.data %>% rownames_to_column(var = "GeneID")

exp <- sc.data %>% 
  inner_join(genes, by = "GeneID")

exp.symbol <- exp %>% dplyr::select(-ensemblID)
exp.entrez <- exp %>% dplyr::select(-GeneID)

spls <- colnames(
  exp %>% dplyr::select(-c(ensemblID, GeneID))
) %>% 
  data.frame(ID=.)
```
#define gene sets
```{r}
#HIF gene set
hif.sig <- c("IGFBP3", "EDN2", "PFKFB4", "FLT1", "TFR2", "BNIP3L", "TGFA",
             "BNIP3","PGK1","EGLN1","LDHA","EGLN3","CP","TGFB3","PFKFB3",
             "HK1","TFRC","EDN1","CDKN1A","CA9","ADM","HMOX1","SERPINE1",
             "LOX","NDRG1","CA12","PDK1","VEGFA","ERO1L","RORA","P4HA1","MXI1",
             "SLC2A1","STC2","MIF","DDIT4","ENO1","CXCR4","PLOD1","P4HA2","GAPDH","PGAM1","TMEM45A","PIM1") %>% 
  match(genes$GeneID) %>% genes$ensemblID[.]

## HIF signature gene expression:
hif <- match(hif.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(hif) <- NULL
hif <- column_to_rownames(hif, var = "ensemblID")

hif_scores <- hif[,-1] %>% colMeans() %>% data.frame(hif_score=., ID=colnames(hif)[-1])

all_gene_sets <- getGmt("msigdb.v7.0.entrez.gmt")
gene_sets <- all_gene_sets %>% # Load all GSEA/Broad MSigDB gene sets
  .[grep("VERHAAK_GLIOBLASTOMA", names(.))] # Pull curated gene sets by their database/primary identifier

##Mesenchymal gene set
mes.sig <- geneIds(gene_sets)$VERHAAK_GLIOBLASTOMA_MESENCHYMAL
mes <- match(mes.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(mes) <- NULL
mes <- column_to_rownames(mes, var = "ensemblID")
mes_scores <- mes[,-1] %>% colMeans() %>% data.frame(mes_scores=., ID=colnames(mes)[-1])

##proneural gene set
pn.sig <- geneIds(gene_sets)$VERHAAK_GLIOBLASTOMA_PRONEURAL
pn <- match(pn.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(pn) <- NULL
pn <- column_to_rownames(pn, var = "ensemblID")
pn_scores <- pn[,-1] %>% colMeans() %>% data.frame(pn_scores=., ID=colnames(pn)[-1])

##Classical gene set
cl.sig <- geneIds(gene_sets)$VERHAAK_GLIOBLASTOMA_CLASSICAL
cl <- match(cl.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(cl) <- NULL
cl <- column_to_rownames(cl, var = "ensemblID")
cl_scores <- cl[,-1] %>% colMeans() %>% data.frame(cl_scores=., ID=colnames(cl)[-1])

##hallmark hypoxia gene set
hall.hyp <- getGmt("msigdb.v7.0.entrez.gmt") %>% # Load all GSEA/Broad MSigDB gene sets
  .[grep("HALLMARK_HYPOXIA", names(.))]
hyp.sig <- geneIds(hall.hyp)$HALLMARK_HYPOXIA
hyp <- match(hyp.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(hyp) <- NULL
hyp <- column_to_rownames(hyp, var = "ensemblID")
hyp_scores <- hyp[,-1] %>% colMeans() %>% data.frame(hyp_scores=., ID=colnames(hyp)[-1])

#hallmark TGFb gene set
hall.tgfb <- all_gene_sets %>% .[grep("HALLMARK_TGF_BETA_SIGNALING", names(.))]
tgfb.sig <- geneIds(hall.tgfb)$HALLMARK_TGF_BETA_SIGNALING
tgfb <- match(tgfb.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(tgfb) <- NULL
tgfb <- column_to_rownames(tgfb, var = "ensemblID")
tgfb_scores <- tgfb[,-1] %>% colMeans() %>% data.frame(tgfb_scores=., ID=colnames(tgfb)[-1])

#hallmark TNFa gene set
hall.tnfa <- all_gene_sets %>% .[grep("HALLMARK_TNFA_SIGNALING_VIA_NFKB", names(.))]
tnfa.sig <- geneIds(hall.tnfa)$HALLMARK_TNFA_SIGNALING_VIA_NFKB
tnfa <- match(tnfa.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(tnfa) <- NULL
tnfa <- column_to_rownames(tnfa, var = "ensemblID")
tnfa_scores <- tnfa[,-1] %>% colMeans() %>% data.frame(tnfa_scores=., ID=colnames(tnfa)[-1])

#hallmark p53 gene set
hall.p53 <- all_gene_sets %>% .[grep("HALLMARK_P53_PATHWAY", names(.))]
p53.sig <- geneIds(hall.p53)$HALLMARK_P53_PATHWAY
p53 <- match(p53.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(p53) <- NULL
p53 <- column_to_rownames(p53, var = "ensemblID")
p53_scores <- p53[,-1] %>% colMeans() %>% data.frame(p53_scores=., ID=colnames(p53)[-1])

#hallmark MYC gene set
hall.myc <- all_gene_sets %>% .[grep("HALLMARK_MYC_TARGETS_V1", names(.))]
myc.sig <- geneIds(hall.myc)$HALLMARK_MYC_TARGETS_V1
myc <- match(myc.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(myc) <- NULL
myc <- column_to_rownames(myc, var = "ensemblID")
myc_scores <- myc[,-1] %>% colMeans() %>% data.frame(myc_scores=., ID=colnames(myc)[-1])

#PID integrin gene set
pid.integrin <- all_gene_sets %>% .[grep("HALLMARK_MYC_TARGETS_V1", names(.))]
integrin.sig <- geneIds(pid.integrin)$HALLMARK_MYC_TARGETS_V1
integrin <- match(integrin.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(integrin) <- NULL
integrin <- column_to_rownames(integrin, var = "ensemblID")
integrin_scores <- integrin[,-1] %>% colMeans() %>% data.frame(integrin_scores=., ID=colnames(integrin)[-1])

#hallmark STAT3 gene set
hall.stat3 <- all_gene_sets %>% .[grep("HALLMARK_IL6_JAK_STAT3_SIGNALING", names(.))]
stat3.sig <- geneIds(hall.stat3)$HALLMARK_IL6_JAK_STAT3_SIGNALING
stat3 <- match(stat3.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(stat3) <- NULL
stat3 <- column_to_rownames(stat3, var = "ensemblID")
stat3_scores <- stat3[,-1] %>% colMeans() %>% data.frame(stat3_scores=., ID=colnames(stat3)[-1])

#annotate samples with their gene set scores:
spls <- spls %>%
  #mutate(hif.score = scale(hif_scores$hif_score, center = T, scale = T)) %>%
  mutate(mes.score = scale(mes_scores$mes_scores, center = T, scale = T)) %>%
  mutate(pn.score = scale(pn_scores$pn_scores, center = T, scale = T)) %>%
  mutate(hyp.score = scale(hyp_scores$hyp_scores, center = T, scale = T)) %>%
  mutate(tgfb.score = scale(tgfb_scores$tgfb_scores, center = T, scale = T)) %>% 
  mutate(tnfa.score = scale(tnfa_scores$tnfa_scores, center = T, scale = T)) 
#  mutate(p53.score = scale(p53_scores$p53_scores, center = T, scale = T)) %>% 
#  mutate(myc.score = scale(myc_scores$myc_scores, center = T, scale = T)) %>% 
#  mutate(integrin.score = scale(integrin_scores$integrin_scores, center = T, scale = T)) %>% 
#  mutate(stat3.score = scale(stat3_scores$stat3_scores, center = T, scale = T))

```
#UMAP
```{r UMAP plotting - all genes}
## Perform UMAP embedding:
set.seed(123) # For reproducibility
force_run = F

if(!exists("embedding")){
  embedding <- readRDS("umap_embedding.rds")
}


if(force_run){
embedding <- umap::umap(exp.symbol[,-1] %>% t(), 
                      n_neighbors = 30, min_dist = 0.01,
                      n_components = 3
                      )

saveRDS(embedding, file = "umap_embedding.rds")
}

## Plot UMAP embedding:
embedding$layout %>%
  data.frame() %>% mutate(ID = rownames(.)) %>% 
  full_join(spls, by="ID") %>% 
  pivot_longer(-c(ID,X1,X2)) %>% 
  ggplot(aes(X1,X2, label = ID, color = value)) + 
    geom_point() + 
    scale_color_viridis_c(option="B") +
    facet_wrap(~name) +
    theme_cowplot(18) + theme_cust(ticks=F,axis="both",box=T) +
    labs(title="UMAP projection - all genes", x = "UMAP1", y = "UMAP2")

```
##UMAP - PMT
```{r UMAP plotting - mes and pn gene sets}
exp.pmt <- bind_rows(mes, pn, cl)[,-1]
colnames(exp.pmt) <- samples.mal
n_dim <- 3

## Perform UMAP embedding:
set.seed(111) # For reproducibility
if(!exists("embedding2")){
  embedding2 <- readRDS("umap_embedding2.rds")
}
##generate UMAP projection
force_run = F
if(force_run){
embedding2 <- umap::umap(exp.pmt %>% t(), 
                        n_neighbors = 30, min_dist = 0.01,
                        n_components = n_dim
                        )

saveRDS(embedding2, file = "umap_embedding2.rds")
}

## Plot UMAP embedding:
embedding2$layout %>%
  data.frame() %>% mutate(ID = rownames(.)) %>% 
  full_join(spls, by="ID") %>% 
  pivot_longer(-c(ID,X1,X2)) %>% 
  ggplot(aes(X1,X2, label = ID, color = value)) + 
    geom_point() + 
    scale_color_viridis_c(option="C") +
    facet_wrap(~name) +
    theme_cowplot(18) + theme_cust(ticks=F,axis="both",box=T) +
    labs(title="UMAP projection - PMT genes", x = "UMAP1", y = "UMAP2")

umap.proj <- embedding2$layout %>% t() %>% data.matrix()

force.run = F
if(force.run == T){
seed1 <- 12345

cc <- ConsensusClusterPlus(umap.proj, 
                           maxK = 10,
                           reps = 100,
                           pItem = 0.8, 
                           pFeature = 1,
                           clusterAlg = "pam",
                           distance = "euclidean",
                           seed = seed1)

icl <- calcICL(cc)
grps <- icl[["itemConsensus"]]

saveRDS(grps, file = "scRNA_grps.rds")
} else {
  grps <- readRDS("scRNA_grps.rds")
}

numgrps <- 4

grps.sub <- grps %>% 
  filter(k==numgrps) %>% 
  group_by(item) %>% 
  filter(itemConsensus==max(itemConsensus))

#grps$item <- grps$item %>% sub("-", ".", .)

embedding2$layout %>% 
  data.frame() %>% 
  mutate(ID = rownames(.)) %>% 
  full_join(spls, by = "ID") %>% 
  full_join(grps.sub, by = c("ID" = "item")) %>% 
  ggplot(aes(x = X1, y = X2, shape = factor(cluster), color = mes.score))+
  geom_point()+
  scale_color_viridis(option = "B")

embedding2$layout %>% 
  data.frame() %>% 
  mutate(ID = rownames(.)) %>% 
  full_join(spls, by = "ID") %>% 
  full_join(grps.sub, by = c("ID" = "item")) %>% 
  ggplot(aes(x = X1, y = X2, shape = factor(cluster), color = pn.score))+
  geom_point()+
  scale_color_viridis(option = "B")

# grps.sub %>% as.data.frame() %>% count(cluster)
```
##clustering comparison
```{r}
#comparison of my clustering to Neftel's
umap.proj <- embedding2$layout %>%
  data.frame() %>% mutate(ID = rownames(.))

sample.subtypes <- sc.samples.info %>% dplyr::select(MESlike1, MESlike2, OPClike, AClike, NPClike1, NPClike2) %>% 
  sapply(as.numeric) %>%
  as.data.frame() %>% 
  mutate(ID = samples.mal)

umap.proj %>% left_join(sample.subtypes, by = "ID") %>% 
  full_join(grps.sub, by = c("ID" = "item")) %>% 
  na.omit() %>% 
  dplyr::select(-itemConsensus, -k) %>% 
  pivot_longer(-c(ID, X1, X2, X3, cluster)) %>% 
  # filter(cluster==3) %>%
  ggplot(aes(X1,X2, label = ID, color = value, shape = factor(cluster))) + 
    geom_point(alpha = 0.7) + 
    facet_wrap(~name)+
    scale_color_viridis_c(option="C") +
    theme_cowplot(18) + theme_cust(ticks=F,axis="both",box=T) +
    labs(title="UMAP projection - PMT genes", x = "UMAP1", y = "UMAP2")
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/scRNAseq_UMAP.png", width = 18, height = 12)

# umap.proj %>% left_join(sample.subtypes, by = "ID") %>% 
#   pivot_longer(-c(ID, X1, X2, X3)) %>% 
#   ggplot(aes(X1,X3, label = ID, color = value)) + 
#     geom_point() + 
#     facet_wrap(~name)+
#     scale_color_viridis_c(option="C") +
#     theme_cowplot(18) + theme_cust(ticks=F,axis="both",box=T) +
#     labs(title="UMAP projection - PMT genes", x = "UMAP1", y = "UMAP3")

umap.proj %>% left_join(sample.subtypes, by = "ID") %>% 
  full_join(grps.sub, by = c("ID" = "item")) %>% 
  na.omit() %>% 
  dplyr::select(-itemConsensus, -k) %>% 
  pivot_longer(-c(ID, X1, X2, X3, cluster)) %>% 
  filter(cluster==1) %>%
  ggplot(aes(X1,X2, label = ID, color = value, shape = factor(cluster))) + 
    geom_point() + 
    facet_wrap(~name)+
    scale_color_viridis_c(option="C") +
    theme_cowplot(18) + theme_cust(ticks=F,axis="both",box=T) +
    labs(title="UMAP projection - PMT genes", x = "UMAP1", y = "UMAP2")


```

#load gene sets
```{r loading gene sets for GSVA}
## ## Load packages:
# BiocManager::install(c(msigdbr","GSVA","limma"))
library(msigdbr)

### Define MSigDB gene set collection(s) to use --> retrieve with 'msigdbr' package:
species = "Homo sapiens"

## Retrieve Hallmark and canonical pathways collections in the database:
hall = msigdbr(species = species, category = "H")
cp = msigdbr(species = species, category = "C2", subcategory = "CP")
cp.b = msigdbr(species = species, category = "C2", subcategory = "CP:BIOCARTA")
cp.r = msigdbr(species = species, category = "C2", subcategory = "CP:REACTOME")
cp.p = msigdbr(species = species, category = "C2", subcategory = "CP:PID")
cp.k = msigdbr(species = species, category = "C2", subcategory = "CP:KEGG")
cp.w = msigdbr(species = species, category = "C2", subcategory = "CP:WIKIPATHWAYS")
cp.cgp = msigdbr(species = species, category = "C2", subcategory = "CGP")
gene_sets1 <- rbind(hall, cp.b, cp.r, cp.p, cp.k, cp.cgp) %>% split(x = .$gene_symbol, f = .$gs_name)

## Go collections:
go.bp <- msigdbr(species = species, category = "C5", subcategory = "GO:BP")
go.cc <- msigdbr(species = species, category = "C5", subcategory = "GO:CC")
go.mf <- msigdbr(species = species, category = "C5", subcategory = "GO:MF")
gene_sets2 <- rbind(go.mf) %>% split(x = .$gene_symbol, f = .$gs_name)

## Transcription factor target collections:
go.tft <- msigdbr(species = species, category = "C3", subcategory = "TFT:GTRD")
go.tft_leg <- msigdbr(species = species, category = "C3", subcategory = "TFT:TFT_Legacy")
gene_sets3 <- rbind(go.tft, go.tft_leg) %>% split(x = .$gene_symbol, f = .$gs_name)

## 
cgp = msigdbr(species = species, category = "C2", subcategory = "CGP")
gene_sets4 <- rbind(cgp) %>% split(x = .$gene_symbol, f = .$gs_name)

gene_sets_all <- gene_sets1 %>% 
  #append(gene_sets2) %>% 
  #append(gene_sets3) %>% 
  append(gene_sets4)
```
#GSVA
```{r}
data_for_gsva <- exp.symbol[,-1] %>% data.matrix()
rownames(data_for_gsva) <- exp.symbol$GeneID

force_calculate_gsva <- F
if(!file.exists("gbm_gsva_results.rds") | isTRUE(force_calculate_gsva)){
  gsva_res <- gsva(
    data_for_gsva, 
    gene_sets1, verbose = F, # Change which gene sets are run
    # gene_sets4, verbose = T
    method = "zscore", 
    kcdf = "Gaussian",
    # abs.ranking = F,
    # mx.diff = T,
    min.sz = 10, # Minimum number of genes required to include a gene set
    parallel.sz=detectCores()#-1
    )
  ## Save GSVA results for speed:
  saveRDS(gsva_res, file = "gbm_gsva_results.rds")
  
} else if(file.exists("gbm_gsva_results.rds") & !exists("gsva_res")){ 
  ## Load GSVA results, if they haven't been loaded/calculated already:
  gsva_res <- readRDS("gbm_gsva_results.rds") 
}
## Build data frame versions of GSVA results:
gsva_res.df <- gsva_res %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
gsva_res.df.t <- gsva_res %>% t() %>% data.frame(ID=rownames(.), .)
rownames(gsva_res.df) <- gsva_res.df$gene.set
```

```{r plot GSVA data in a heatmap}
pmt_sets <- c(
  "HALLMARK_TGF_BETA_SIGNALING",
  "HALLMARK_HYPOXIA",
  "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
  "VERHAAK_GLIOBLASTOMA_MESENCHYMAL",
  "VERHAAK_GLIOBLASTOMA_PRONEURAL"
)

pmt_gsva <- gsva_res.df %>% subset(gene.set %in% pmt_sets) %>% dplyr::select(-gene.set)

### Draw heatmap:
pheatmap(pmt_gsva,
         clustering_method = "ward.D2", clustering_distance_rows = "canberra",
         # color = colorRampPalette(paletteer_d("RColorBrewer::PuOr", direction = -1))(101),
         color = colorRampPalette(paletteer_c("viridis::magma", n=101))(101),
         #cutree_rows = 2, 
         cutree_cols = 2,
         treeheight_row = 1, treeheight_col = 1, border_color = NA,
         # cellwidth = 7,
         cellheight = 25,
         show_colnames = F, angle_col = 90,
         fontsize = 8,
         main = "scGBM: Enrichment scores",
         )
```

```{r gene set correlations}
## Look at correlations between the PMT signature and the other signatures:
#corr_data <- pmt_gsva %>% t() %>% data.frame(ID=rownames(.)) %>% 
#  full_join(mes_scores, by="ID") %>%
#  full_join(hif_scores, by="ID") 
corr_data <- pmt_gsva %>% t() %>% data.frame(ID=rownames(.)) 

## Get all pairwise correlations:
gsva_corrs <- corr_data %>%
  dplyr::select(-ID) %>% 
  data.matrix() %>% 
  rcorr(type="spearman")

gsva_corrs$P

len_out <- 20
pheatmap(gsva_corrs$r,
         clustering_distance_rows = "euclidean", clustering_distance_cols = "euclidean",
         clustering_method = "ward.D2",
         show_colnames = T, show_rownames = T,
         treeheight_row = 0.3, treeheight_col = 0.3,
         color = colorRampPalette(pals::coolwarm(len_out))(len_out), 
         breaks = seq(-1,1,length.out = len_out),
         fontsize = 6, fontsize_number = 8,
         cellwidth = 30, cellheight = 30,
         angle_col = 90, border_color = NA, display_numbers = T,
         main="GBM cells \nGSVA Enrichment Spearman correlations",
         filename ="C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/scRNAseq_heatmap.pdf"
         )

```
#consensus clustering of samples - DO NOT RUN
```{r eval=FALSE, include=FALSE}
seed1 <- 12345
subtype_genes <- c(gene_sets1$VERHAAK_GLIOBLASTOMA_MESENCHYMAL, gene_sets1$VERHAAK_GLIOBLASTOMA_PRONEURAL, gene_sets1$VERHAAK_GLIOBLASTOMA_CLASSICAL)


d <- exp.symbol %>% column_to_rownames(var = "GeneID") %>% na.omit()
d <- d[subtype_genes,] %>% na.omit()

res <- M3C(d, removeplots = F, iters=5, fsize=8, clusteralg = "pam", seed = seed1)

grps <- res$realdataresults[[3]]$assignments %>% data.frame() %>% rownames_to_column(var = "item")
colnames(grps)[2] <- "cluster"
grps$item <- grps$item %>% sub("-", ".", .)

gsva_res.df.t %>% 
  dplyr::select(one_of("ID", "VERHAAK_GLIOBLASTOMA_MESENCHYMAL", "VERHAAK_GLIOBLASTOMA_PRONEURAL")) %>% 
  inner_join(grps, by = c("ID" = "item")) %>% 
  ggplot(aes(x = VERHAAK_GLIOBLASTOMA_MESENCHYMAL, y = VERHAAK_GLIOBLASTOMA_PRONEURAL, color = factor(cluster)))+
  geom_point()

count(grps, vars = "cluster")

saveRDS(res, file = "M3C_clustering.rds")


```


##DEA
```{r}
#split samples into MES/PN based on Neftel classifications
plotdata <- gsva_res.df.t
#rownames(plotdata) <- rownames(gsva_res.df)

subtype.list <- cell.info[-1,] %>% 
  dplyr::select(NAME, MESlike1, MESlike2, AClike, OPClike, NPClike1, NPClike2) %>%
  filter(NAME %in% samples.mal) %>%
  column_to_rownames(var = "NAME") %>%
  mutate(max_col = colnames(.)[max.col(.)]) %>%
  filter(max_col %in% c("MESlike1", "MESlike2", "OPClike"))

samples.pmt <- rownames(subtype.list)

design1 <- subtype.list %>% 
  transmute(
  mes = case_when(
  .$max_col == "OPClike" ~ 0,
  TRUE ~ 1
  ),
  pn = case_when(
    .$max_col == "MESlike1" ~ 0,
    .$max_col == "MESlike2" ~0,
    TRUE ~ 1
  )
) %>%
  rownames_to_column(var = "ID")


plotdata <- plotdata %>% semi_join(design1, by = "ID")
design1 <- design1 %>% semi_join(plotdata, by = "ID") %>%
  column_to_rownames(var = "ID") %>% data.matrix()

plotdata <- plotdata[,-1]
#highlyCor <- findCorrelation(cor(plotdata), cutoff = 0.9)
#Apply correlation filter at 0.70,
#then we remove all the variable correlated with more 0.7.
#plotdata <- plotdata[,-highlyCor]

plotdata <- plotdata %>% t()
```

```{r}
#splits samples into MES/PN based on Verhaak genes UMAP + consensus clustering
plotdata <- gsva_res.df.t %>% arrange(ID)

subtype.list <- grps.sub %>% arrange(item)

design1 <- subtype.list %>% transmute(
  mes = case_when(cluster==1 ~ 1,
                  T ~ 0),
  pn = case_when(cluster==2 ~1, 
                 T ~ 0)
) %>% column_to_rownames(var = "item") %>% 
  data.matrix()

plotdata <- plotdata[,-1]
plotdata <- plotdata %>% t()


```

```{r plot GSVA results}
## Set color scheme for heat map:
mycol <- brewer.pal(n=11, name="RdBu") %>% rev() # Color scheme for heatmap
#mycol <- colorRampPalette(c("#0000cc","#cccccc","#ff0000")) # Custom color scheme
#### Differential expression analysis ####
## Define statistical thresholds:
logFCcutoff1 <- 1 # Threshold for log2(fold-change)
adjPvalueCutoff <- 10^-2 # Threshold for adjusted p-values (FDR) - use for GO gene set
#adjPvalueCutoff <- 0.00000001 # use for other gene sets (KEGG, BIOCARTA, etc.)
## Linear model with limma: GSVA enrichment scores

gene_sets_to_plot <- c(
#  "HALLMARK_HYPOXIA",
#  "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
#  "DAUER_STAT3_TARGETS_UP",
#  "OZANNE_AP1_TARGETS_UP",
#  "WP_TGFB_SIGNALING_IN_THYROID_CELLS_FOR_EPITHELIALMESENCHYMAL_TRANSITION",
#  "VERHAAK_GLIOBLASTOMA_MESENCHYMAL"
)

fit2 <- lmFit(plotdata, design1)
contr2 <- makeContrasts(mes - pn, # Make contrasts --> compare groups of interest
                        levels = colnames(coef(fit2)))
tmp2 <- contrasts.fit(fit2, contr2)
tmp2 <- eBayes(tmp2)
allGeneSets <- topTable(tmp2, sort.by = "P", number=Inf)
allGeneSets$ID <- rownames(allGeneSets)
DEgeneSets <- topTable(tmp2, sort.by = "P", number=Inf,
 p.value=adjPvalueCutoff, adjust="BH")
res2 <- decideTests(tmp2, p.value=adjPvalueCutoff)
summary(res2)

#DEgenes
DEgeneSets

#### Plots ####
### Volcano plots --> MvsP gene sets
gene_sets_sub <- filter(allGeneSets, grepl("TGF|TNFA|_NFKB", ID)) %>% rownames()


#DEgenes
DEgeneSets
DEgeneSets$logFC %>% order() %>% DEgeneSets[.,] %>% write.csv("DEgene_sets_scGBM.csv")

#### Plots ####
### Volcano plots --> MvsP gene sets
#onts <- c(hall$gs_name, )


allGeneSets %>% 
  filter(ID %ni% cp.cgp$gs_name) %>%   #change this to select which gene sets get plotted
#  filter(ID %in% hall$gs_name) %>% 
  ggplot(
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
#  geom_vline(xintercept = logFCcutoff1, col = "black", linetype = "dashed", alpha = 0.6, size = 0.7)+
#  geom_vline(xintercept = -logFCcutoff1, col = "black", linetype = "dashed", alpha = 0.6, size = 0.7)+
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    alpha = 0.5,
#    cex = 2.5
  ) +
#  geom_text(aes(x = -0.5, y = -log10(adjPvalueCutoff),
#                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
  geom_label_repel(
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff)},
    data = function(x){subset(x, ID %in% gene_sets_sub)},
    size = 3, alpha = 0.7,
    segment.size = 0.25, segment.alpha = 1, label.padding = 0.1, max.overlaps = 20, max.iter = 100000
  ) +
  geom_point(
    data = function(x){subset(x, ID %in% gene_sets_sub)},
    alpha = 0.7
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 10),
    axis.title = element_text(size = 10),
    plot.title = element_text(size = 10),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "Gene Sets: Mesenchymal vs. Proneural",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits = c(0,100))+
  scale_x_continuous(limits = c(-2,2))
#ggsave("scRNAseq_mes_vs_pn_full.png", width = 10, height = 8)
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/scRNAseq_mes_pn.svg")



## volcano plot with numbers
geneSetsSub <- allGeneSets %>% filter(ID %ni% cp.cgp$gs_name)
ranked_sets <- geneSetsSub %>% filter(ID %in% gene_sets_sub) %>% 
  mutate(rank = 1:nrow(.))
geneSetsSub <- geneSetsSub %>% full_join(ranked_sets)

geneSetsSub %>% 
  ggplot(
       aes(x = logFC, y = -log10(adj.P.Val), label = rank)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    alpha = 0.5,
#    cex = 2.5
  ) +
 geom_text(aes(x = 0, y = -log10(adjPvalueCutoff)*1.1,
               label = paste("p = ",adjPvalueCutoff)), size = 2, color = "red") +
    geom_point(
    data = function(x){subset(x, ID %in% gene_sets_sub)},
    alpha = 0.7
  ) +
  geom_label_repel(
    data = function(x){subset(x, ID %in% gene_sets_sub)},
    size = 3, alpha = 0.9, max.overlaps = 30) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(10,
    line_size = 0.2
  ) +
  theme(
    axis.text = element_text(size = 10),
    axis.title = element_text(size = 10),
    plot.title = element_text(size = 10),
    panel.background = element_rect(size = 1, color = "black", fill = "white"),
    ) +
  labs(
    title = "",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits = c(0,100))+
  scale_x_continuous(limits = c(-2,2))
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/scRNAseq_mes_pn_numbers.svg", width = 5, height = 4, units = "in")
rownames(ranked_sets)



allGeneSets %>% 
  filter(adj.P.Val < adjPvalueCutoff) %>% 
  arrange(desc(logFC)) %>% 
  filter(ID %ni% c(cp.cgp$gs_name)) %>%   #change this to select which gene sets get plotted
  # filter(ID %in% hall$gs_name) %>% 
  top_n(50, abs(logFC)) %>% 
  ggplot(aes(y = reorder(ID, logFC),  x = logFC, label = ID, fill = adj.P.Val)) +
  geom_bar(stat= "identity")
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/scRNAseq_mes_pn_barchart.png", width = 20, height = 15)

```

```{r eval=FALSE, include=FALSE}
#define gene sets
gfs1 <- read_csv("cytokines_gfs.csv") %>% pull(gene)
pks1 <- read_csv("protein_kinases.csv") %>% pull(gene)

tnfa_genes <- gene_sets_all$HALLMARK_TGF_BETA_SIGNALING

subtype.list <- cell.info[-1,] %>% 
  dplyr::select(NAME, MESlike1, MESlike2, AClike, OPClike, NPClike1, NPClike2) %>%
  filter(NAME %in% samples.mal) %>%
  column_to_rownames(var = "NAME") %>%
  mutate(max_col = colnames(.)[max.col(.)]) %>%
  filter(max_col %in% c("MESlike1", "MESlike2", "OPClike"))

samples.pmt <- rownames(subtype.list)

design1 <- subtype.list %>% 
  transmute(
  mes = case_when(
  .$max_col == "OPClike" ~ 0,
  TRUE ~ 1
  ),
  pn = case_when(
    .$max_col == "MESlike1" ~ 0,
    .$max_col == "MESlike2" ~0,
    TRUE ~ 1
  )
) %>%
  rownames_to_column(var = "ID")

## Statistical thresholds:
adjPvalueCutoff1 <- 0.001 # Threshold for adjusted p-values (FDR) for GENE DEA
logFCcutoff1 <- log2(1) # Threshold for log2(fold-change)

data_for_lm <- exp.symbol %>% dplyr::select(any_of(design1$ID))
rownames(data_for_lm) <- exp.symbol$GeneID

### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
rownames(design1) <- NULL
design1 <- design1 %>% data.frame() %>% column_to_rownames(var = "ID")
fit1 <- lmFit(data_for_lm, design1) 
contr1 <- makeContrasts(MesvsPn = mes - pn, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes


########## Various visualizations ########## 
## Gene LFC vs. average log-cpm:
plotMD(tmp1, column=1, status=res1[,1], main=colnames(tmp1)[1], 
       xlim=c(-8,13))

###################################### Volcano plot --> GENES ######################################
txtsz <- 14
linesz <- 1

DEAgenes <- allGenes
DEAgenes$logFC %>% order() %>% DEAgenes[.,] %>% write.csv("ordered_DEgenes.csv")


ggplot(data = subset(allGenes, ID %in% tnfa_genes), 
#ggplot(data = allGenes,
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1 %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5,
  ) +
#  geom_point(color = "black",
 #   shape = 15,
    # cex = 2.5
#  ) +
   geom_text_repel(
     data = function(x){subset(x, (adj.P.Val < adjPvalueCutoff1) & (abs(logFC) > logFCcutoff1))},
     position = "dodge",
     size = 2, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 10, max.iter = 5000
   ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz)+
  labs(
    title = "DEA genes: Protein kinases", x = "Expression log2FC", y = "-log10(adj. p-value)"
  ) +
  guides(
    color = FALSE
  )#+
  #scale_y_log10(limits = c(1, 150))+
  #scale_x_continuous(limits = c(0,2.5))

```

# Chk2-hypoxia connection
```{r}
gsva.hyp <- gsva_res.df.t %>% dplyr::select(ID, HALLMARK_HYPOXIA, BIOCARTA_HIF_PATHWAY)

chk2.exp <- exp.symbol %>% filter(GeneID == "CHEK2") %>% 
  column_to_rownames(var = "GeneID") %>% 
  t() %>% 
  data.frame() %>% 
  rownames_to_column(var = "ID")

chk2.mat <- full_join(gsva.hyp, chk2.exp) 

chk2.mat <- chk2.mat %>% filter(CHEK2>0)

ggscatterstats(data = chk2.mat,
               x = HALLMARK_HYPOXIA,
               y = CHEK2)

ggscatterstats(data = chk2.mat,
               x = BIOCARTA_HIF_PATHWAY,
               y = CHEK2)

hyp.med <- median(chk2.mat$HALLMARK_HYPOXIA)
hif.med <- median(chk2.mat$BIOCARTA_HIF_PATHWAY)
chk2.mat <- chk2.mat %>% mutate(hypoxia_high = if_else(HALLMARK_HYPOXIA > hyp.med, T, F),
                                hif_high = if_else(BIOCARTA_HIF_PATHWAY > hif.med, T, F))

ggbetweenstats(data = chk2.mat,
               x = hypoxia_high,
               y = CHEK2)

ggbetweenstats(data = chk2.mat,
               x = hif_high,
               y = CHEK2)
```

# pathway-protein connection (general)
```{r}
library(ggstatsplot)

p1 <- c("HALLMARK_TNFA_SIGNALING_VIA_NFKB")
g1 <- c("PPP2R5E")

gsva.sub <- gsva_res.df.t %>% dplyr::select(p1) %>% 
  rownames_to_column(var = "sample")


g1.exp <- exp.symbol %>% filter(GeneID %in% g1) %>%
  column_to_rownames(var = "GeneID") %>% 
  t() %>% 
  data.frame() %>% 
  rownames_to_column(var = "sample")

g1.mat <- full_join(gsva.sub, g1.exp)


for(path in p1){
  for(gene in g1){
    print(
      g1.mat %>% 
        filter(!!sym(gene)>0) %>% 
      ggscatterstats(x = !!sym(path),
                     y = !!sym(gene),
                     ggplot.component = list(scale_x_log10(), scale_y_log10()))
    )
  }
}

```

##boxplots and linear correlations
```{r}
library(ggbeeswarm)
library(ggstatsplot)

samples.mes <- design1 %>% data.frame() %>% mutate(ID = rownames(.)) %>% filter(mes == 1) %>% pull(ID)
samples.pn <- design1 %>% data.frame() %>% mutate(ID = rownames(.)) %>% filter(pn == 1) %>% pull(ID)

gsva.res.subtype <- gsva_res.df.t %>% filter(ID %in% samples.mes | ID %in% samples.pn) %>% 
  pivot_longer(cols = -ID, names_to = "pathway", values_to = "score") %>% 
  mutate(subtype = if_else(ID %in% samples.mes, "MES", "PN"))

path.to.plot <- c("HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_HYPOXIA")

gsva.res.subtype$pathway <- gsva.res.subtype$pathway %>% factor(path.to.plot)
gsva.res.subtype$subtype <- gsva.res.subtype$subtype %>% factor(c("PN", "MES"))

# BOXPLOT
gsva.res.subtype %>% 
  filter(pathway %in% path.to.plot) %>% 
  ggplot(aes(x = subtype, y = score, color = pathway))+
  geom_boxplot()+
  # geom_beeswarm(aes(shape = "."))+
  facet_wrap(vars(pathway), nrow = 1)+
  theme_cowplot()+
  theme(legend.position = "none")+
  scale_y_continuous(limits = c(-8, 35))+
  labs(y = "GSVA Score", x = element_blank())+
  scale_color_manual(values = c("darkred", "darkgreen", "darkblue"))+
  theme(text=element_text(size=7), axis.text.x=element_text(size=7),
          axis.text.y=element_text(size=7))
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/scRNA_GSVA_boxplot.svg", width = 4, height = 3)

gsva.res.subtype %>% 
  filter(pathway %in% path.to.plot) %>% 
  ggplot(aes(x = subtype, y = score, color = pathway))+
  geom_violin(aes(alpha = subtype, fill = pathway))+
  geom_boxplot(width = 0.4, outlier.shape = NA)+
  # geom_beeswarm(aes(shape = "."))+
  facet_wrap(vars(pathway), nrow = 1, scales = "free_y")+
  theme_cowplot()+
  theme(legend.position = "none")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  labs(y = "GSVA Score", x = element_blank())+
  scale_color_manual(values = c("darkred", "darkgreen", "darkblue"))+
  scale_alpha_discrete(range = c(0.5, 1))+
  theme(text=element_text(size=7), axis.text.x=element_text(size=7),
          axis.text.y=element_text(size=7))
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/scRNA_GSVA_violin.svg", width = 4, height = 3)

#STATS PLOTS
for(p in path.to.plot){
  print(
  gsva.res.subtype %>%
    filter(pathway == p) %>%
      ggbetweenstats(x = "subtype", y = "score", type = "np")+
    labs(title = p)
)
}
```

```{r}

path.to.plot2 <- c("HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_HYPOXIA", "VERHAAK_GLIOBLASTOMA_PRONEURAL")
# SCATTER PLOTS
for(p in path.to.plot2){
  g.cor <- gsva_res.df %>% filter(gene.set == p|gene.set == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>%
    dplyr::select(-gene.set) %>%
    t() %>%
    data.matrix()

  corr1 <- cor.test(g.cor[,1], g.cor[,2], method = "pearson")
  corr <- as.character(round(corr1$estimate, digits = 3))
  pval <- as.character(formatC(signif(corr1$p.value, digits=3), digits=2, format="e"))
  statlab = paste0("R = ",corr,"\n p = ", pval)

  max.y <- gsva_res.df.t %>% pull(p) %>% max
  max.x <- gsva_res.df.t %>% pull("VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% max
  min.y <- gsva_res.df.t %>% pull(p) %>% min
  min.x <- gsva_res.df.t %>% pull("VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% min

  print(
    gsva_res.df.t %>%
      dplyr::select(!!sym(p), "VERHAAK_GLIOBLASTOMA_MESENCHYMAL")%>%
      ggplot(aes(x = VERHAAK_GLIOBLASTOMA_MESENCHYMAL, y = !!sym(p)))+
      geom_point(size = 0.5)+
      geom_smooth(method = 'lm')+
      geom_label(aes(x = max.x*(0.8), y = (max.y-min.y)*0.2+min.y, label = statlab), size = 1.5)+
      theme_cowplot()+
      theme(text=element_text(size=7), axis.text.x=element_text(size=7),
          axis.text.y=element_text(size=7))+
      scale_y_continuous(limits = c(min.y, max.y))
  )
  save.loc <- "C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/"
  # ggsave(filename = paste0(save.loc, "scRNA_GSVA_biplot_", p, ".svg", sep=""), height = 3/2, width = 2)
  ggsave(filename = paste0(save.loc, "scRNA_GSVA_biplot_", p, ".png", sep=""), height = 3/2, width = 2, dpi = 600)
}



```

#remove overlapping genes
```{r}
#what I am doing here is removing any genes that appear in Verhaak's mesenchymal gene set from the TNFa, TGFb, and hypoxia gene sets

gene_sets_rel <- c("VERHAAK_GLIOBLASTOMA_MESENCHYMAL", 
                   "VERHAAK_GLIOBLASTOMA_PRONEURAL", 
                   "HALLMARK_TGF_BETA_SIGNALING", 
                   "HALLMARK_TNFA_SIGNALING_VIA_NFKB", 
                   "HALLMARK_HYPOXIA")

gene_sets_ro <- msigdbr(species = "Homo sapiens") %>% 
  # dplyr::select(gs_name, entrez_gene) %>% 
  filter(gs_name %in% gene_sets_rel)

g_mes <- gene_sets_ro %>% filter(gs_name == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL")
g_pn <- gene_sets_ro %>% filter(gs_name == "VERHAAK_GLIOBLASTOMA_PRONEURAL")
g_tgfb <- gene_sets_ro %>% filter(gs_name == "HALLMARK_TGF_BETA_SIGNALING")
g_tnfa <- gene_sets_ro %>% filter(gs_name == "HALLMARK_TNFA_SIGNALING_VIA_NFKB")
g_hyp <- gene_sets_ro %>% filter(gs_name == "HALLMARK_HYPOXIA")

g_tgfb <- g_tgfb %>% filter(entrez_gene %ni% g_mes$entrez_gene)
g_tnfa <- g_tnfa %>% filter(entrez_gene %ni% g_mes$entrez_gene)
g_hyp <- g_hyp %>% filter(entrez_gene %ni% g_mes$entrez_gene)


g1 <- rbind(g_mes, g_pn) %>% rbind(g_tgfb) %>% rbind(g_tnfa) %>% rbind(g_hyp) %>% 
  split(x = .$gene_symbol, f = .$gs_name)


```
##GSVA again
```{r}
data_for_gsva <- exp.symbol[,-1] %>% data.matrix()
rownames(data_for_gsva) <- exp.symbol$GeneID

force_calculate_gsva <- T
if(!file.exists("gsva_results_2.rds") | isTRUE(force_calculate_gsva)){
  gsva_res2 <- gsva(
    data_for_gsva, 
    g1, verbose = F, # Change which gene sets are run
    method = "zscore", 
    kcdf = "Gaussian", 
    min.sz = 10, # Minimum number of genes required to include a gene set
    parallel.sz=detectCores()#-1
    )
  ## Save GSVA results for speed:
  saveRDS(gsva_res2, file = "gsva_results_2.rds")
  
} else if(file.exists("gsva_results_2.rds") & !exists("gsva_res2")){ 
  ## Load GSVA results, if they haven't been loaded/calculated already:
  gsva_res2 <- readRDS("gsva_results_2.rds") 
}
## Build data frame versions of GSVA results:
gsva_res.df <- gsva_res2 %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
gsva_res.df.t <- gsva_res2 %>% t() %>% data.frame(ID=rownames(.), .)
rownames(gsva_res.df) <- gsva_res.df$gene.set

# gsva_res.df.t$ID <- gsva_res.df.t$ID %>% gsub("-", ".", .)


gsva_res_filt <- gsva_res.df.t %>% filter(ID %in% samples.mes | ID %in% samples.pn)
```

##boxplots
```{r}

gsva.res.subtype <- gsva_res.df.t %>% filter(ID %in% samples.mes | ID %in% samples.pn) %>% 
  pivot_longer(cols = -ID, names_to = "pathway", values_to = "score") %>% 
  mutate(subtype = if_else(ID %in% samples.mes, "MES", "PN"))

path.to.plot <- c("HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_HYPOXIA")

gsva.res.subtype$pathway <- gsva.res.subtype$pathway %>% factor(path.to.plot)
gsva.res.subtype$subtype <- gsva.res.subtype$subtype %>% factor(c("PN", "MES"))


gsva.res.subtype %>% 
  filter(pathway %in% path.to.plot) %>% 
  ggplot(aes(x = subtype, y = score, color = pathway))+
  geom_violin(aes(alpha = subtype, fill = pathway))+
  geom_boxplot(width = 0.4, outlier.shape = NA)+
  # geom_beeswarm()+
  facet_wrap(vars(pathway), nrow = 1, scales = "free_y")+
  theme_cowplot()+
  theme(legend.position = "none")+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  labs(y = "GSVA Score", x = element_blank())+
  scale_color_manual(values = c("darkred", "darkgreen", "darkblue"))+
  scale_alpha_discrete(range = c(0.5, 1))+
  theme(text=element_text(size=7), axis.text.x=element_text(size=7),
          axis.text.y=element_text(size=7))
ggsave("C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/extra/scRNA_GSVA_violin.svg", width = 4, height = 3)

for(p in path.to.plot){
  print(
  gsva.res.subtype %>% 
    filter(pathway == p) %>% 
      ggbetweenstats(x = "subtype", y = "score", type = "np")+
    labs(title = p)
)
}
```

##linear correlations
```{r}
path.to.plot2 <- c("HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_HYPOXIA", "VERHAAK_GLIOBLASTOMA_PRONEURAL")
# SCATTER PLOTS
for(p in path.to.plot2){
  g.cor <- gsva_res.df %>% filter(gene.set == p|gene.set == "VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>%
    dplyr::select(-gene.set) %>%
    t() %>%
    data.matrix()

  corr1 <- cor.test(g.cor[,1], g.cor[,2], method = "pearson")
  corr <- as.character(round(corr1$estimate, digits = 3))
  pval <- as.character(formatC(signif(corr1$p.value, digits=3), digits=2, format="e"))
  statlab = paste0("R = ",corr,"\n p = ", pval)

  max.y <- gsva_res.df.t %>% pull(p) %>% max
  max.x <- gsva_res.df.t %>% pull("VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% max
  min.y <- gsva_res.df.t %>% pull(p) %>% min
  min.x <- gsva_res.df.t %>% pull("VERHAAK_GLIOBLASTOMA_MESENCHYMAL") %>% min

  print(
    gsva_res.df.t %>%
      dplyr::select(!!sym(p), "VERHAAK_GLIOBLASTOMA_MESENCHYMAL")%>%
      ggplot(aes(x = VERHAAK_GLIOBLASTOMA_MESENCHYMAL, y = !!sym(p)))+
      geom_point(size = 0.5)+
      geom_smooth(method = 'lm')+
      geom_label(aes(x = max.x*(0.8), y = (max.y-min.y)*0.2+min.y, label = statlab), size = 1.5)+
      theme_cowplot()+
      theme(text=element_text(size=7), axis.text.x=element_text(size=7),
          axis.text.y=element_text(size=7))+
      scale_y_continuous(limits = c(min.y, max.y))
  )
  save.loc <- "C:/Users/willh/Documents/Lab/GBM PMT/PMT manuscript/extra/"
  # ggsave(filename = paste0(save.loc, "scRNA_GSVA_biplot_", p, ".svg", sep=""), height = 3/2, width = 2)
  ggsave(filename = paste0(save.loc, "scRNA_GSVA_biplot_", p, ".png", sep=""), height = 3/2, width = 2, dpi = 600)
}

```