---
title: "R Notebook"
output: html_notebook
---

data comes from https://doi.org/10.1016/j.cell.2019.06.024, downloaded from BROAD institute ssRNA-seq portal

```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(cowplot)
library(plotly)
library(RColorBrewer)
library(colorspace)
library(jcolors)
library(ggsci)
library(cowplot)
library(Hmisc)
library(stats)
library(NMF)
library(ggfortify)
library(ggrepel)
library(GSVA)
library(GSVAdata)
library(GSEABase)
library(limma)
library(msigdbr)
library(gage)
library(GO.db)
library(heatmaply)
#library(plsVarSel)
#library(scran)
#library(scater)
#library(SC3)
#library(SingleCellExperiment)
#library(singleCellTK)
library(magrittr)
library(BiocManager)
library(paletteer)
library(ggbeeswarm)
library(ggstatsplot)
library(pheatmap)
library(parallel)
library(org.Hs.eg.db)
library(umap)
library(M3C)
library(biomaRt)
```

```{r Clear environment}
### Clear R workspace:
rm(list=ls())
```

```{r Check and set working directory}
### Check current working directory:
cwd <- getwd()
cwd
### Set working directory:
setwd(cwd) # Set current working directory to location of this notebook for easier handling and saving of data associated with the notebook
```
```{r}
theme_cust <- function(ticks, axis, box, lnsz, color){
  if(missing(color)){color = "black"}
  if(missing(ticks)){ticks <- TRUE}
  if(missing(axis)){axis <- "y"}
  if(missing(box)){box <- TRUE}
  if(missing(lnsz)){lnsz <- 1}
  if(ticks == FALSE & axis == "y" ){
    theme_out <- theme(
      axis.text.y = element_blank(), axis.ticks.y = element_blank(), # Make y axis ticks blank
      axis.line.y = element_blank(), axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = color, fill = "white"),
      )
  } else if(ticks == FALSE & axis == "x"){
    theme_out <- theme(
      axis.text.x = element_blank(), axis.ticks.x = element_blank(), # Make x axis ticks blank
      axis.line.y = element_blank(), axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = color, fill = "white"),
      )
  } else if(ticks == FALSE & axis == "both"){
    theme_out <- theme(
      axis.text.y = element_blank(), axis.ticks.y = element_blank(), # Make y axis ticks blank
      axis.text.x = element_blank(), axis.ticks.x = element_blank(), # Make x axis ticks blank
      axis.line.y = element_blank(), axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = color, fill = "white"),
      )
  } else {
     theme_out <- theme(
       axis.line.y = element_blank(), axis.line.x = element_blank(),
       axis.ticks = element_line(color=color),
       panel.background = element_rect(size = lnsz, color = color, fill = "white"),
       ) 
  }
  
  if(box != TRUE) {
    theme_out <- theme_out + theme(panel.background = element_blank(),
                                   axis.line.y = element_line(size = lnsz, color = color),
                                   axis.line.x = element_line(size = lnsz, color = color)
                                   )
  } else {
    theme_out <- theme_out
  }
}

```

```{r import data set, message=FALSE, warning=FALSE}
scRNA.import <- read_tsv("IDHwtGBM.processed.SS2.logTPM.txt") #these are log2(TPM/10+1) data
subtypes.data <- read_tsv("IDHwt.GBM.Hierarchy.SS2.txt")
cell.info <- read_tsv("IDHwt.GBM.Metadata.SS2.txt")
```

```{r data clean-up and formatting}
sc.data <- scRNA.import %>%
  column_to_rownames(var="GENE")
sc.genes <- pull(scRNA.import, GENE)
sc.samples <- colnames(sc.data)

sc.samples.info <- cell.info[-1,]
sc.samples.info <- filter(sc.samples.info, CellAssignment == "Malignant", GBMType == "Adult")
samples.mal <- pull(sc.samples.info, NAME)
sc.data <- dplyr::select(sc.data, all_of(samples.mal))

sc.samples.info <- sc.samples.info %>% .[order(match(.$NAME, colnames(sc.data))),] %>%
  #select(., c(CellAssignment, NAME)) %>% 
  column_to_rownames(var = "NAME")

## Convert gene names to Entrez IDs:
genes <- rownames(sc.data) %>% 
  mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL', multiVals = "first") %>% 
  data.frame() %>%
  rownames_to_column(var = "GeneID")
colnames(genes)[2] <- "ensemblID"
genes <- genes %>% filter(!is.na(ensemblID))
sc.data <- sc.data %>% rownames_to_column(var = "GeneID")

exp <- sc.data %>% 
  inner_join(genes, by = "GeneID")

exp.symbol <- exp %>% dplyr::select(-ensemblID)
exp.entrez <- exp %>% dplyr::select(-GeneID)

spls <- colnames(
  exp %>% dplyr::select(-c(ensemblID, GeneID))
) %>% 
  data.frame(ID=.)
```

```{r}
#HIF gene set
hif.sig <- c("IGFBP3", "EDN2", "PFKFB4", "FLT1", "TFR2", "BNIP3L", "TGFA",
             "BNIP3","PGK1","EGLN1","LDHA","EGLN3","CP","TGFB3","PFKFB3",
             "HK1","TFRC","EDN1","CDKN1A","CA9","ADM","HMOX1","SERPINE1",
             "LOX","NDRG1","CA12","PDK1","VEGFA","ERO1L","RORA","P4HA1","MXI1",
             "SLC2A1","STC2","MIF","DDIT4","ENO1","CXCR4","PLOD1","P4HA2","GAPDH","PGAM1","TMEM45A","PIM1") %>% 
  match(genes$GeneID) %>% genes$ensemblID[.]

## HIF signature gene expression:
hif <- match(hif.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(hif) <- NULL
hif <- column_to_rownames(hif, var = "ensemblID")

hif_scores <- hif[,-1] %>% colMeans() %>% data.frame(hif_score=., ID=colnames(hif)[-1])


gene_sets <- getGmt("msigdb.v7.0.entrez.gmt") %>% # Load all GSEA/Broad MSigDB gene sets
  .[grep("VERHAAK_GLIOBLASTOMA", names(.))] # Pull curated gene sets by their database/primary identifier
##Mesenchymal gene set
mes.sig <- geneIds(gene_sets)$VERHAAK_GLIOBLASTOMA_MESENCHYMAL
mes <- match(mes.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(mes) <- NULL
mes <- column_to_rownames(mes, var = "ensemblID")
mes_scores <- mes[,-1] %>% colMeans() %>% data.frame(mes_scores=., ID=colnames(mes)[-1])

#proneural gene set
pn.sig <- geneIds(gene_sets)$VERHAAK_GLIOBLASTOMA_PRONEURAL
pn <- match(pn.sig, exp$ensemblID) %>% na.omit() %>% exp[.,] %>% data.frame()
rownames(pn) <- NULL
pn <- column_to_rownames(pn, var = "ensemblID")
pn_scores <- pn[,-1] %>% colMeans() %>% data.frame(pn_scores=., ID=colnames(pn)[-1])

#annotate samples with their gene set scores:
spls <- spls %>%
  mutate(hif.score = scale(hif_scores$hif_score, center = T, scale = T)) %>%
  mutate(mes.score = scale(mes_scores$mes_scores, center = T, scale = T)) %>%
  mutate(pn.score = scale(pn_scores$pn_scores, center = T, scale = T))
```

```{r UMAP plotting of gene sets}
## Perform UMAP embedding:
set.seed(123) # For reproducibility
embedding <- umap::umap(exp.symbol[,-1] %>% t(), 
                        n_neighbors = 30, min_dist = 0.01,
                        n_comp = 2
                        )


## Plot UMAP embedding:
embedding$layout %>%
  data.frame() %>% mutate(ID = rownames(.)) %>% 
  full_join(spls, by="ID") %>% 
  pivot_longer(-c(ID,X1,X2)) %>% 
  ggplot(aes(X1,X2, label = ID, color = value)) + 
    geom_point() + 
    scale_color_viridis_c(option="D") +
    facet_wrap(~name) +
    theme_cowplot(18) + theme_cust(ticks=F,axis="both",box=T) +
    labs(title="GBM cells", x = "UMAP1", y = "UMAP2")

## Mesenchymal subsignature:
len_out_m <- 12
breaks_m <- c(seq(min(mes_scores, na.rm=T), 0, length.out=ceiling(len_out_m/2) + 1), 
              seq(max(mes_scores, na.rm=T)/len_out_m,
                  max(mes_scores, na.rm=T),
                  length.out=floor(len_out_m/2)))

pheatmap(mes_scores,
         # scale = "column",
         clustering_distance_rows = "euclidean", clustering_distance_cols = "canberra",
         clustering_method = "ward.D2",
         color = colorRampPalette(paletteer_c("viridis::plasma", n=len_out_m))(len_out_m),
         # color = colorRampPalette(c("blue","grey80","red"))(len_out_m), breaks = breaks_m,
         annotation_col = mes,
         angle_col = 90, show_rownames = F, treeheight_row = 2, treeheight_col = 1,
         main = "pcEMT signature: Mesenchymal genes",
         # file = 
         )

```



```{r setting up data with SingleCellExperiment}
sc.data <- scRNA.import %>%
  column_to_rownames(var="GENE")
sc.genes <- pull(scRNA.import, GENE)
sc.samples <- colnames(sc.data)

sc.samples.info <- cell.info[-1,]
sc.samples.info <- filter(sc.samples.info, CellAssignment == "Malignant", GBMType == "Adult")
samples.mal <- pull(sc.samples.info, NAME)
sc.data <- dplyr::select(sc.data, all_of(samples.mal))

sc.samples.info <- sc.samples.info %>% .[order(match(.$NAME, colnames(sc.data))),] %>%
  #select(., c(CellAssignment, NAME)) %>% 
  column_to_rownames(var = "NAME")

sc.data.mat <- data.matrix(sc.data)

sce <- SingleCellExperiment(
  assays = list(
    counts = sc.data.mat,
    logcounts = log2(sc.data.mat+1)
  ),
  colData = sc.samples.info
)


rowData(sce)$feature_symbol <- rownames(sce)
sce <- sce[!duplicated(rowData(sce)$feature_symbol),]

#sce <- sce[rowData(sce)$feature_symbol=="Malignant",]

sce <- runPCA(sce)
plotPCA(sce)

```
```{r SC GSVA}
sce.gsva <- sce %>% convertGeneIDs(inSymbol = "SYMBOL", outSymbol = "ENSEMBL", database = "org.Hs.eg.db")

importGeneSetsFromMSigDB(
  inSCE,
  categoryIDs,
  species = "Homo sapiens",
  mapping = c("gene_symbol", "human_gene_symbol", "entrez_gene"),
  by = "rownames",
  verbose = TRUE
)

sce.gsva <- gsvaSCE(sce.gsva, 
               useAssay = "logcounts", 
               pathwaySource = "MSigDB c2 (Human, Entrez ID only)", 
               pathwayNames = "ALL")

```

```{r sc3 analysis}

sce <- sc3(sce, ks=3:4, biology = TRUE)
#sc3_interactive(sc.con)
sc3_plot_consensus(sce, k=3)
sc3_plot_markers(sce, k=4)

```

```{r prep data for GSVA}

scRNA.subtypes <- subtypes.data[-1,]  # remove first row of matrix
scRNA.subtypes <- scRNA.subtypes %>% # assign subtypes based on the paper this data comes from 
  transmute(
    subtype = case_when(
      X > 0 & Y > 0 ~ "NPC",
      X > 0 & Y < 0 ~ "MES",
      X < 0 & Y > 0 ~ "OPC",
      TRUE ~ "AC"
    ),
    NAME=NAME
)
scRNA.subtypes <- scRNA.subtypes %>% filter(subtype =="MES" | subtype == "OPC")

scRNA.data <- scRNA.import 

scRNA.data <- column_to_rownames(scRNA.data, var="GENE")
scRNA.data <- scRNA.data %>% 
  t() %>% 
  as.data.frame() %>%
  .[, -(which(colSums(.)==0))] %>%
  rownames_to_column(var="NAME")
scRNA.data <- scRNA.subtypes %>%
  inner_join(scRNA.data, by="NAME") %>%
  sample_frac(size=0.5, replace=F) %>%
  column_to_rownames(var="NAME")

```

```{r scran analysis}
sce <- scRNA.import %>%
  column_to_rownames(var="GENE") %>% 
  as.matrix()
dec2 <- modelGeneVarWithSpikes(sce)
sced <- denoisePCA(sce, dec2, subset.row=getTopHVGs(dec2, prop=0.1))
ncol(reducedDim(sced, "PCA"))

```

```{r Running GSVA}
data_for_gsva <- scRNA.data %>% 
  .[,-1] %>%
 # scale(center = T, scale = T) %>%
  as.matrix() %>% 
  t()


## Convert gene names to Entrez IDs:
entrez_ids <- rownames(data_for_gsva) %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL')
rownames(data_for_gsva) <- entrez_ids

### Define MSigDB gene set collection to use:
gene_sets <- getGmt("msigdb.v7.0.entrez.gmt") %>% # Load all GSEA/Broad MSigDB gene sets
  .[grep("^HALLMARK|^KEGG|^REACTOME_SIGNALING_BY|^PID_|^BIOCARTA_", names(.))] # Pull curated gene sets by their database/primary identifier

gene_set_choice <- "GO" # Define this as either "GO" or something completely different (like "MSigDB") --> if "GO" is not used, then key words for MSigDB gene sets will be chosen automatically for the GSVA analysis below

### Load desired gene set(s) and perform GSVA:
if(gene_set_choice == "GO"){
  ### Calculate GSVA enrichment scores:
  tn.gsva <- gsva(
    data_for_gsva, 
    go.gsets()[["go.sets"]], 
    method="zscore", 
    parallel.sz=detectCores()-1
    )
  tn.gsva.df <- tn.gsva %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
  tn.gsva.df$gene.set <- gsub("GO.|[0-9]{7}", "", tn.gsva.df$gene.set) 
  rownames(tn.gsva.df) <- tn.gsva.df$gene.set
  rm(tn.gsva) # Remove matrix form from environment
  
} else {
  ## Select key words to further subset the selected MSigDB gene set collection:
gs.kw <- c(
  "REACTOME",
  "PID",
  "BIOCARTA",
  "KEGG",
  "HALLMARK"
  )

  store <- 0
  for (i in 1:length(gs.kw)) {
    temp <- str_subset(names(gene_sets), gs.kw[i])
    store <- c(store, temp)
  }

  store <- str_subset(names(gene_sets), paste(gs.kw, collapse = "|"))
  
  ## Pull the gene sets we want:
  gene.sets <- store[-1] %>% 
    na.omit() %>% 
    match(names(gene_sets)) %>% 
    na.omit() %>% 
    gene_sets[.]
    
  ### Calculate GSVA enrichment scores:
  tn.gsva <- gsva(
    data_for_gsva, 
    gene.sets, 
    method="zscore", 
    parallel.sz=detectCores()-1
    )
  
  tn.gsva.df <- tn.gsva %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
#  tn.gsva.df$gene.set <- gsub(".*HALLMARK_|.*REACTOME_|.*PID_|BIOCARTA_|.*KEGG_", "", tn.gsva.df$gene.set) #Unused: .*KEGG_|
  rownames(tn.gsva.df) <- tn.gsva.df$gene.set
  rm(tn.gsva) # Remove matrix form from environment
}

# Get data to plot:
plotdata <- tn.gsva.df[,-1]
plotdata <- plotdata %>% t() %>% data.frame() %>%
  rownames_to_column(var="sample") %>%
  arrange(sample) %>%
  column_to_rownames(var="sample") %>% t() %>% data.matrix()
rownames(plotdata) <- rownames(tn.gsva.df)

subtype.list <- scRNA.data$subtype %>% data.frame()
rownames(subtype.list) <- rownames(scRNA.data)
design1 <- subtype.list %>% 
  transmute(
  mes = case_when(
  .=="OPC" ~ 0,
  TRUE ~ 1
  ),
  pn = case_when(
    .=="MES" ~ 0,
    TRUE ~ 1
  )
) %>% data.matrix()


```

```{r Differental expression analysis, message=FALSE, warning=FALSE}
#### Differential expression analysis ####
## Define statistical thresholds:
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)
adjPvalueCutoff <- 10^-6 # Threshold for adjusted p-values (FDR) - use for GO gene set
#adjPvalueCutoff <- 0.00000001 # use for other gene sets (KEGG, BIOCARTA, etc.)
## Linear model with limma: GSVA enrichment scores

fit2 <- lmFit(plotdata, design1)
contr2 <- makeContrasts(mes - pn, # Make contrasts --> compare groups of interest
                        levels = colnames(coef(fit2)))
tmp2 <- contrasts.fit(fit2, contr2)
tmp2 <- eBayes(tmp2)
allGeneSets <- topTable(tmp2, sort.by = "P", number=Inf)
allGeneSets$ID <- rownames(allGeneSets)
DEgeneSets <- topTable(tmp2, sort.by = "P", number=Inf,
 p.value=adjPvalueCutoff, adjust="BH")
res2 <- decideTests(tmp2, p.value=adjPvalueCutoff)
summary(res2)

#DEgenes
DEgeneSets

#### Plots ####
### Volcano plots --> MvsP gene sets
ggplot(data = allGeneSets, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff),
                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
#  geom_text_repel(
#    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff)},
#    size = 4, alpha = 0.9,
#    segment.size = 0.25, segment.alpha = .8, label.padding = 5, force=50
#  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 25),
    plot.title = element_text(size = 23),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "GO Gene Sets: Mesenchymal vs. Proneural",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits=c(0,16))
#ggsave("TCGA_mes_vs_pn_GO_no_labels.png", width = 8, height = 6)

#subset of volcano plot
ggplot(data = allGeneSets, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff),
                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
  geom_text_repel(
    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff)},
    size = 4, alpha = 0.9,
    segment.size = 0.25, segment.alpha = .8, max.iter = 10000, label.padding = 1, force=50
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 25),
    plot.title = element_text(size = 23),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "GO Gene Sets: Mesenchymal vs. Proneural",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits=c(10.5,12))+
  scale_x_continuous(limits=c(0,1))
#ggsave("TCGA_mes_vs_pn_GO_subset.svg", width = 8, height = 6)
```

```{r plot GSVA results, eval=FALSE, include=FALSE}
## Set color scheme for heat map:
mycol <- brewer.pal(n=11, name="RdBu") %>% rev() # Color scheme for heatmap
#mycol <- colorRampPalette(c("#0000cc","#cccccc","#ff0000")) # Custom color scheme

plotdata.1 <- data.frame(pathway=rownames(plotdata.mes), mean=rowMedians(plotdata.mes))
pws <- top_n(plotdata.1, 10)


### Generate heatmap:
nk_col <- 4
hm <- 
heatmaply(
  # plotdata[1:40,],
  plotdata,
  k_col = nk_col, # How many column dendrogram clusters to color
  k_row = 8, # How many row dendrogram clusters to color
  limits = c(-max(abs(plotdata)),max(abs(plotdata))),
  #distfun_row = function(x) as.dist(1-cor(t(x), method = "spearman")),
  #distfun_col = function(x) as.dist(1-cor(x, method = "spearman")),
  # dist_method = "manhattan", # If NULL, Euclidean is used
  hclust_method = NA,
  # show_dendrogram = c(TRUE, FALSE),
  # show_dendrogram = c(FALSE, TRUE),
  show_dendrogram = c(TRUE, TRUE),
  branches_lwd = 0.3,
  colors=mycol,
  # scale = "row",
  # scale = "col",
  key.title="GSVA \nEnrichment Score",
  # showticklabels = c(FALSE, TRUE),
  # showticklabels = c(TRUE, FALSE),
  showticklabels = c(TRUE, TRUE),
  cexCol = 0.5, # Scale size of column tick/label text
  cexRow = 1, # Scale size of row tick/label text
  column_text_angle = 90,
  # ylab = "Sample",
  # xlab = "Sample",
  xlab = "Sample",
  main = "GSVA of RNA-seq Data",
  # file = "GSVA_CCLE_RNAseq_TNBC.html", # Save the heatmap to html
  )
#hm
```