Notebook to analyze TCGA data regarding EGFRviii and potential stress relief

```{r}
library(tidyverse)
library(magrittr)
library(cowplot)
library(plotly)
library(readxl)
library(pls)
library(plsVarSel)
library(ggfortify)
library(ggrepel)
library(pheatmap)
library(ggbeeswarm)
library(org.Hs.eg.db)
library(GSVAdata)
library(GSEABase)
library(limma)
library(msigdbr)
library(gage)
library(plyr)
library(GO.db)
library(GSVA)
library(factoextra)
library(ggstatsplot)
```

```{r Clear environment}
### Clear R workspace:
rm(list=ls())
```

```{r Check and set working directory}
### Check current working directory:
cwd <- getwd()
cwd
### Set working directory:
setwd(cwd) # Set current working directory to location of this notebook for easier handling and saving of data associated with the notebook
```
##Define functions
```{r define functions}
rq_func <- function(model){
  ### Get X and Y matrices:
  xmat <- model$model$X
  ymat <- model$model$Y
  ### Performance metrics:
  R2X <- explvar(model)
  names(R2X) <- seq(1:length(R2X)) %>% paste("PC",.,sep="")
  R2X.cum <- cumsum(R2X)
  PRESS <- model$validation$PRESS
  ## Calculate R2Y values for PLSR model:
  model.TSS <- sum((ymat-mean(ymat))^2)
  R2Y <- 0
  R2Y.cum <- 0
  for(i in 1:model$ncomp){
    R2Y.cum[i] <- (1-sum(model$residuals[,,i]^2)/model.TSS)*100
    if(i == 1){
     R2Y[i] <- R2Y.cum[i] 
    } 
    else{
      R2Y[i] <- R2Y.cum[i]-R2Y.cum[i-1]
    }
  }
  ## Calculate Q2Y values for PLSR model:
  Q2Y <- 0
  Q2Y.cum <- 0
  for(i in 1:model$ncomp){
    Q2Y.cum[i] <- (1-sum(PRESS[,i])/model.TSS)*100
    if(i == 1){
      Q2Y[i] <- Q2Y.cum[i] 
      } 
    else{
      Q2Y[i] <- Q2Y.cum[i]-Q2Y.cum[i-1]
      }
  }
  ### Define output list:
  output <- list(R2X=R2X, R2X.cum=R2X.cum, R2Y=R2Y, R2Y.cum=R2Y.cum, Q2Y=Q2Y, Q2Y.cum=Q2Y.cum)
  return(output)
}


`%!in%` <- Negate(`%in%`)

```
##load data
```{r load data}
HiSeqV2_exon <- read.delim("HiSeqV2_exon", header=T)
GBM_mc3_gene_level <- read.delim("GBM_mc3_gene_level.txt")
HiSeqV2 <- read.delim("HiSeqV2.txt") #log2+1 transformed RSEM values
mmc5 <- read_excel("mmc5.xlsx", range = "A4:G168")
```

```{r}
dat1 <- HiSeqV2_exon %>% filter(grepl("chr7", .$sample), grepl(":+", .$sample, fixed = T))
dat1 %<>% mutate(exon.start = str_match(.$sample, "chr7:\\s*(.*?)\\s*-"))
dat1 %<>% mutate(exon.end = str_match(.$sample, "-\\s*(.*?)\\s*:+"))
dat1$exon.start <- dat1$exon.start[,2]
dat1$exon.end <- dat1$exon.end[,2]
dat1 %<>% select(exon.start, exon.end, everything(), -sample) 

dat1$exon.start <- as.numeric(dat1$exon.start)
dat1$exon.end <- as.numeric(dat1$exon.end)
dat1 %<>% mutate(exon.length = 1+ exon.end - exon.start)
dat1 %<>% select(exon.start, exon.end, exon.length, everything()) 

dat1.egfr <- dat1 %>% filter(.$exon.start > 55086710 & exon.end < 55279321)
dat1.egfr %<>% arrange(exon.start)
dat1.egfr %<>% slice(1,3:16, 19:31)

exons.egfr <- dat1.egfr %>% select(-exon.start, -exon.end, -exon.length) %>% 
  rownames_to_column("exon")

#exons.egfr.t <- t(exons.egfr)[-1,]
exons.egfr$exon %<>% as.numeric()

samples <- colnames(exons.egfr)[-1]
```
##PCA
```{r}
dat2 <- exons.egfr

dat3 <- dat2[,1:2]
colnames(dat3) <- c("exon", "expression")


for (i in 3:ncol(dat2)) {
  s <- colnames(dat2)[i]
  e <- pull(dat2, exon)
  v <- pull(dat2, s)
  ev <- bind_cols(e,v)
  colnames(ev) <- c("exon", "expression")
  dat3 <- dat3 %>% bind_rows(ev)
}

dat3$exon %<>% as.character()
exons <- 1:nrow(dat2) %>% as.character()
dat3$exon <- factor(dat3$exon, exons)

ggplot(dat3, aes(x = exon, y = expression))+
  geom_boxplot()
  #geom_dotplot(binaxis = "y", binwidth = 0.1)
  #geom_beeswarm()

dat2.1 <- dat2 %>%  column_to_rownames("exon")
pheatmap(scale(dat2.1, T, T), cluster_rows = F)
pheatmap(t(dat2.1))

dat2.t <- t(dat2.1) %>% data.frame()
colnames(dat2.t) <- exons
dat2.mean <- dat2.t %>% mutate(mean.total = rowMeans(.), mean.viii = rowMeans(select(., -2:7)))
dat2.mean %<>% mutate(frac.viii = mean.viii/mean.total)


res.pca <- prcomp(dat2.t, scale = T)
fviz_eig(res.pca)

fviz_pca_biplot(res.pca, repel=T)

dat2.egfrviii <- top_n(dat2.mean, 40, frac.viii)
```
##GSVA
```{r}
tcga.rna.data <- HiSeqV2
tcga.rna.data <- column_to_rownames(tcga.rna.data, var="sample")
tcga.rna.data <- t(tcga.rna.data)


tcga.rna <- as.data.frame(tcga.rna.data)
tcga.rna <- tcga.rna %>% data.frame()
#tcga.rna <- tcga.rna %>% select_if(~ !any(is.na(.)))
tcga.rna <- tcga.rna %>% .[, colSums(is.na(.))==0]

data_for_gsva <- tcga.rna %>% scale(center=T, scale=T) %>% data.frame()
#rownames(data_for_gsva) <- exp.symbol$GeneID
colnames(data_for_gsva) <- colnames(data_for_gsva) %>% 
  mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL')
genes <- colnames(data_for_gsva) %>% na.omit()
data_for_gsva <- data_for_gsva %>% dplyr::select(one_of(genes)) %>% t()

#gene_sets <- getGmt("msigdb.v7.0.entrez.gmt") %>% # Load all GSEA/Broad MSigDB gene sets
#  .[grep("^HALLMARK|^KEGG|^REACTOME_SIGNALING_BY|^PID_|^BIOCARTA_", names(.))] # Pull curated gene sets by their database/primary identifier

gene_sets <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "GO:BP") %>% split(x = .$entrez_gene, f = .$gs_name)

force_calculate_gsva <- F
if(!file.exists("gsva_results.rds") | isTRUE(force_calculate_gsva)){
  gsva_res <- gsva(
    data_for_gsva, 
    gene_sets, verbose = F, # Change which gene sets are run
    # gene_sets4, verbose = T
    method = "gsva", 
#    kcdf = "Gaussian",
    # abs.ranking = F,
    # mx.diff = T,
    min.sz = 10, # Minimum number of genes required to include a gene set
    parallel.sz=detectCores()#-1
    )
  ## Save GSVA results for speed:
  saveRDS(gsva_res, file = "gsva_results.rds")
  
} else if(file.exists("gsva_results.rds") & !exists("gsva_res")){ 
  ## Load GSVA results, if they haven't been loaded/calculated already:
  gsva_res <- readRDS("gsva_results.rds") 
}
## Build data frame versions of GSVA results:
gsva_res.df <- gsva_res %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
gsva_res.df.t <- gsva_res %>% t() %>% data.frame()
rownames(gsva_res.df) <- gsva_res.df$gene.set


```
##EGFRvIII expression
```{r}
v3.status <- mmc5 %>% select(-`EGFR CNA`)
v3.status$Sample %<>% gsub("TCGA.", "", .)

tcga.rna.sub <- tcga.rna
tcga.rna.sub %<>% rownames_to_column("Sample")
tcga.rna.sub$Sample %<>% gsub("TCGA\\.[0-9]{2}\\.", "", .)
tcga.rna.sub$Sample %<>% gsub("\\.[0-9]{2}", "", .)

v3.status %<>% semi_join(tcga.rna.sub, by = "Sample") %>% arrange(Sample)
tcga.rna.sub %<>% semi_join(v3.status, by= "Sample") %>% distinct(Sample, .keep_all = T) %>% arrange(Sample)

colnames(v3.status)[3] <- "viii"
v3.status %<>% select(Sample, viii)
v3.status$viii %<>% gsub("<0.01", "0", .) %>% as.numeric()

gsva_res.sub <- gsva_res.df.t
gsva_res.sub %<>% rownames_to_column("Sample")
gsva_res.sub$Sample %<>% gsub("TCGA\\.[0-9]{2}\\.", "", .)
gsva_res.sub$Sample %<>% gsub("\\.[0-9]{2}", "", .)
gsva_res.sub %<>% semi_join(v3.status, by= "Sample") %>% distinct(Sample, .keep_all = T) %>% arrange(Sample)


v3.status %<>% column_to_rownames("Sample")
gsva_res.sub %<>% column_to_rownames("Sample")

samples.viii <- filter(v3.status, viii>0) %>% rownames()
tcga.rna.sub %<>% mutate(viii.status = case_when(
  Sample %in% samples.viii ~ "Positive",
  Sample %!in% samples.viii ~ "Negative"))


```
##plotting gene expression based on EGFRvIII status
```{r}
ggbetweenstats(data = tcga.rna.sub,
               x = viii.status,
               y = PTEN)

ggbetweenstats(data = tcga.rna.sub,
               x = viii.status,
               y = PTPN11)

ggbetweenstats(data = tcga.rna.sub,
               x = viii.status,
               y = EGFR)


ggbetweenstats(data = tcga.rna.sub,
               x = viii.status,
               y = LAMP1)

tcga.rna.sub %>% filter(Sample %in% samples.viii) %>% 
  cbind(filter(v3.status, viii>0)) %>% 
  ggplot(aes(x = viii, y = SOCS2))+
  geom_point()

```
##DEA based on EGFRvIII status
```{r}
## Statistical thresholds:
adjPvalueCutoff1 <- 0.01 # Threshold for adjusted p-values (FDR) for GENE DEA
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)

## Gene expression data:
data_for_lm <- data.frame(tcga.rna.sub) %>% select(-viii.status) %>% column_to_rownames("Sample") %>% t()


### Fit linear model with limma: Normalized gene count data #########################################################
## Define design matrix of sample groups:
design1 <- cbind(Sample = colnames(data_for_lm)) %>% data.frame()
design1 %<>% mutate(viii.pos = case_when(
  Sample %in% samples.viii ~ 1,
  T ~ 0
)) %>% mutate(viii.neg = case_when(
  Sample %in% samples.viii ~ 0,
  T ~ 1
)) %>% 
  column_to_rownames("Sample")

fit1 <- lmFit(data_for_lm, design1) 
contr1 <- makeContrasts(PosvsNeg = viii.pos - viii.neg, # Make contrasts --> compare groups of interest
                        levels = colnames(design1))
## Perform DEA:
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- rownames(allGenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- rownames(DEgenes) #%>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)
DEgenes

###################################### Volcano plot --> GENES ######################################
txtsz <- 14
linesz <- 1
## Save list of ordered, differentially expressed genes:
#allGenes <- mutate(allGenes, gene = allGenes$ID %>% match(genes$ensembl) %>% genes$gene[.])
DEAgenes.TT <- allGenes
#DEAgenes.TT$logFC %>% order() %>% DEAgenes.TT[.,] %>% write.csv("ordered_DEgenes_TGF+TNF.csv")


##PLOTTING
ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(P.Value), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    color = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = P.Value < adjPvalueCutoff1 & abs(logFC) > logFCcutoff1 %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), size = 3, color = "red") +
   geom_text_repel(
     data = function(x){subset(x, (P.Value < adjPvalueCutoff1) & (abs(logFC) > logFCcutoff1))},
     position = "dodge",
     size = 3, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
     max.overlaps = 20, max.iter = 500000
   ) +
#  geom_text_repel(data = function(x){subset(x, P.Value)}) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(txtsz, line_size = linesz) + 
  labs(
#    title = "DEA: TGFb+TNFa vs. Untreated", 
    x = expression("Expression log"[2]*"FC"), 
    y = expression ("-log"[10]*"(p-value)")
  ) +
  guides(
    color = FALSE
  )


ggplot(data = arrange(top_n(allGenes, 30, -P.Value), desc(logFC)),
       aes(y = reorder(ID, logFC),
           x = logFC,
       fill = P.Value
       ))+
  geom_bar(stat="identity")+
  labs(
#    title = "DEA: TGFb+TNFa vs. Untreated", 
    x = expression("Expression log"[2]*"FC (EGFRvIII(+) - EGFRvIII(-)"), 
    y = element_blank())

topGenes <- allGenes %>% filter(P.Value < 0.01) 
topGenes %<>% top_n(15, logFC) %>% 
  rbind(top_n(topGenes, 15, -logFC)) %>% 
  arrange(logFC)

ggplot(data = arrange(topGenes, desc(logFC)),
       aes(y = reorder(ID, logFC),
           x = logFC,
       fill = P.Value
       ))+
  geom_bar(stat="identity")+
  labs(
#    title = "DEA: TGFb+TNFa vs. Untreated", 
    x = expression("Expression log"[2]*"FC (EGFRvIII(+) - EGFRvIII(-)"), 
    y = element_blank())
```

```{r PLSR}
y.mat <- v3.status %>% as.matrix() %>% scale(center = T, scale =T)
#x.mat <- gsva_res.sub %>% select(contains("stress")) %>% as.matrix() %>% scale(center = T, scale =T)

x.mat <- tcga.rna.sub %>% column_to_rownames("Sample") %>% as.matrix()# %>% scale (T, T)
#y.mat <- dat2.t %>% rownames_to_column("sample") %>% arrange(sample) %>% column_to_rownames("sample") %>% 
#  as.matrix() %>% scale(center = T, scale =T)
#x.mat <- gsva_res.df.t %>% rownames_to_column("sample") %>% arrange(sample) %>% column_to_rownames("sample") %>% 
#  select(contains("endocytosis")) %>% 
#  as.matrix() %>% scale(center = T, scale =T)


mydata <- data.frame(Y=I(y.mat),X=I(x.mat))
pls1 <- plsr(Y~X, scale=F, data=mydata, validation = "LOO", method="kernelpls") # generate PLSR model

sc <- scores(pls1)
ld <- loadings(pls1)
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)
Yld <- data.frame(matrix(as.numeric(Yld), attributes(Yld)$dim, dimnames=attributes(Yld)$dimnames)) 

#plot biplot of scores and loadings - full model
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], alpha=0.5))+
  geom_text(aes(x=Yld[,1], y=Yld[,2]), color="blue", label = rownames(Yld), size = 3)+
  labs(title="", 
       x="PC1", 
       y="PC2")+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#222222",
                  size=3)+
#  geom_text_repel(aes(x=Yld[,1], y=Yld[,2]), 
#            label=rownames(Yld), size=3, color="blue")+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.1))

#plot R2X, R2Y, Q2Y
pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
#    geom_vline(xintercept = comp, linetype="dashed") + 
    geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  theme(panel.background = element_rect(color = "#000000"))+
  scale_y_continuous(limits = c(0,100))
#ggsave("TCGA_PLSR_var-expl.png", width = 5, height = 4)


#VIP score calculations
vip.scores <- VIP(pls1, opt.comp = comp) %>% as.data.frame()
colnames(vip.scores) <- "score"
vip.scores %<>% rownames_to_column("analyte")

#plot VIP scores
ggplot(data=arrange(vip.scores, desc(score)), 
       aes(x=reorder(analyte, -score), y=score))+
  geom_bar(stat="identity")+
    theme_cowplot()+
  theme(axis.text.x=element_text(angle=90, vjust=.5, hjust=1),
        panel.background = element_rect(color = "#000000")
        )+
  scale_fill_gradient2( # Custom color gradient for bars
    midpoint=0,
    low="red",
    mid="#555555",
    high="blue",
    aesthetics="fill"
  )+
  geom_hline(yintercept = 1, linetype = "dashed")
```

```{r filtered PLSR model}
#filter out analytes with low VIP scores and regenerate the model
keep.scores <- vip.scores %>% top_n(50, score) %>% .[,1]
x.mat <- data.frame(x.mat) %>% select(keep.scores) %>% data.matrix()
mydata <- data.frame(Y=I(y.mat),X=I(x.mat))
pls2 <- plsr(Y~X, scale=F, data=mydata, validation = "LOO", method="kernelpls") # generate PLSR model

sc <- scores(pls2)
ld <- loadings(pls2)
Ysc <-Yscores(pls2)
Yld <- Yloadings(pls2)

ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], alpha=0.5))+
  geom_text(aes(x=Yld[,1], y=Yld[,2]), color="blue", label = rownames(Yld))+
  labs(title="", 
       x="PC1", 
       y="PC2")+
#       x=paste0("PC1", " (",explvar(pls2)[1]%>%round(1), "%)"), 
#       y=paste0("PC2", " (",explvar(pls2)[2]%>%round(1), "%)"))+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#222222",
                  size=3)+
#  geom_text_repel(aes(x=Yld[,1], y=Yld[,2]), 
#            label=rownames(Yld), size=3, color="blue")+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.1))

#plot R2X, R2Y, Q2Y
pls2res <- rq_func(pls2)
R2X <- pls2res$R2X; R2X_cum <- pls2res$R2X.cum
R2Y <- pls2res$R2Y; R2Y_cum <- pls2res$R2Y.cum
Q2Y_cum <- pls2res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

plot(pls2, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls2res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls2res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls2res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
#    geom_vline(xintercept = comp, linetype="dashed") + 
    geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  theme(panel.background = element_rect(color = "#000000"))+
  scale_y_continuous(limits = c(0,100))
```


