---
title: "GSVA Analysis of CCLE RNA-seq Data"
output: 
  html_document: 
    fig_height: 8
    fig_width: 18
    theme: spacelab
    toc: yes
Author: Paul J. Myers
Date of creation: 1/17/2020
R version: 4.0.0
---


# Housekeeping Code
We load the necessary libraries for this script and clear the R workspace.

```{r  Load packages,message=FALSE}
### Load packages; not all of these are necessary for the calculations done here, but I can't remember which packages I am/am not using throughout, so I've left them all here:
library(tidyverse)
library(magrittr)
library(reshape2)
library(ggthemes)
library(ggpubr)
library(paletteer)
library(cowplot)
library(plotly)
library(RColorBrewer)
library(colorspace)
library(ggrepel)
library(Hmisc)
```

```{r Clear environment}
### Clear R workspace:
rm(list=ls())
```

```{r Check and set working directory}
### Check current working directory:
cwd <- getwd()
cwd
### Set working directory:
setwd(cwd)
```

# Load CCLE RNA-seq Data
We now load RNA-seq data from the Cancer Cell Line Encyclopedia (CCLE). (See https://depmap.org/portal/download/ and download the file "CCLE_expression.csv".) We are only interested in using the RNA-seq data for protein coding genes. Other data from the CCLE, including the CCLE's RPPA data and information on all of the cell lines in the CCLE (required for the code below) are also available for download from this portal. These data are supplied as log2-transformed, 1-shifted RSEM TPM values, denoted below as $x_{i,j}^{log}$:

$$x_{i,j}^{log}=log_2(TPM_{i,j}+1)$$
We'll start by first loading the information on all triple-negative breast cancer (TNBC) cell lines that are in the CCLE and then loading the TNBC RNA-seq data itself.

```{r Load CCLE cell line info, warning=FALSE}
##### Load cell line info:
## Path for cell line info:
fn.cells <- "CCLE_sample_info.csv"

## Load info for TNBC cell lines:
cl.sum <- read.csv(fn.cells)   #Cell line summaries
cl.id <- cl.sum$DepMap_ID %>% as.character()  #DepMap IDs for cell lines
cl.name <- cl.sum$stripped_cell_line_name %>% as.character()  #Cell line names
ccle.names <- cl.sum$CCLE_Name %>% as.character() #CCLE cell line names

## Desired TNBC cell lines:
subtype <- c("ERneg_HER2neg")
tcl.names <- which(cl.sum$lineage_sub_subtype %in% subtype) %>% cl.sum[.,] %>% .$stripped_cell_line_name %>% as.character %>% c("HS274T") # Info of cell lines of interest
# tcl.names <- c("BT20","BT549","CAL120","CAL148","CAL51","CAL851","DU4475","HCC1143","HCC1187",
#                "HCC1395","HCC1599","HCC1806","HCC1937","HCC2157","HCC38","HCC70","HDQP1",
#                "HMEL","HS274T","HS578T","MDAMB157","MDAMB231","MDAMB436","MDAMB453","MDAMB468","SUM159PT")


tcl.ind <- which(cl.name %in% tcl.names) #Indices of TNBC lines in full list of cell lines

## Extract TNBC cell line names:
tcl.info <- cl.sum[tcl.ind,]
tcl.id <- cl.id[tcl.ind] #TNBC cell line DepMap IDs...
tcl.subtype <- cl.sum[tcl.ind,c("stripped_cell_line_name","lineage_molecular_subtype")]
tcl.name <- cl.name[tcl.ind] %>% as.character() #...and the stripped cell line names in the same order.
tcl.ccle.names <- ccle.names[tcl.ind] # Full CCLE names for TNBC lines
```

And now load the CCLE RNA-seq data.

```{r Load CCLE RNA-seq data, warning=FALSE}
##### Load CCLE RNA-seq data: log2(TPM+1) values
## File path for CCLE RNA-seq data:
fn4 = "CCLE_RNAseq_gene_expression_TPM_log2.csv"

## Load CCLE RNA-seq data:
ccle.rna <- read_csv(fn4, col_names=T)
ccle.genes <- colnames(ccle.rna)[-1] %>%
  gsub("\\ .*","",.) # Remove extraneous suffixes on gene names
colnames(ccle.rna) <- c("Cell.line", ccle.genes)

## Get TNBC cell line data:
ccle.id.rna <- ccle.rna$Cell.line %>% as.character() # Get DepMap IDs from CCLE data

# Pull out the TNBC IDs, names, and data:
tcl.ind.ccle.rna <- which(ccle.id.rna %in% tcl.id) # Match TNBC cell line DepMap IDs to DepMap IDs in data set
tcl.id.ccle.rna <- ccle.id.rna[tcl.ind.ccle.rna] # DepMap IDs in data set
tcl.name.ccle.rna <- match(tcl.id.ccle.rna, tcl.id) %>%
  tcl.name[.] # Names

# Pull out the TNBC data:
tcl.ccle.rna <- ccle.rna[tcl.ind.ccle.rna,] # mRNA expression

tcl.ccle.rna$Cell.line <- tcl.name.ccle.rna # Replace DepMap IDs with stripped cell line names
```

# GSVA: CCLE RNA-seq Data
### mRNA Expression Data
Here, we'll use the technique known as gene set variation analysis (GSVA) to analyze the molecular/cellular signaling pathways that are up- and down-regulated in each of the TNBC cell lines of interest. GSVA is closely related to single-sample gene set enrichment analysis (ssGSEA). The GSVA package we'll use below actually includes an option for running ssGSEA in the main command for the analysis, if it's desired.

The description for the GSVA R package can be found [here](https://www.bioconductor.org/packages/release/bioc/html/GSVA.html). 
```{r GSVA: CCLE RNA-seq data, warning=FALSE, message=FALSE}
### Install packages if not done already:
# BiocManager::install(c("limma","GSVA","GSVAdata","org.Hs.eg.db","GSEABase","snow","rlecuyer","edgeR")) # Comment out this line if you've already installed these packages to your current R distribution.

### Load packages:
library(GSVA) # Gene set variation analysis package from Bioconductor --> This package contains a command that can perform four types of gene set enrichment analyses: GSVA, ssGSEA, PLAGE, and z-score.
library(GSVAdata)
library(org.Hs.eg.db)
library(GSEABase)
library(limma)
library(msigdbr)

### Start by putting data in .GCT format:
data_for_gsva <- tcl.ccle.rna %>% as.matrix() %>% t() %>% .[-1,]
colnames(data_for_gsva) <- tcl.ccle.rna$Cell.line

## Convert gene names to Entrez IDs:
entrez_ids <- rownames(data_for_gsva) %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL')
rownames(data_for_gsva) <- entrez_ids

### Define MSigDB gene set collection(s) to use:
gene_sets <- getGmt("msigdb.v7.0.entrez.gmt") %>% # Load all GSEA/Broad MSigDB gene sets using Entrez IDs --> download this file from the MSigDB website; change this file name to match the file of the gene set collection you want to use from MSigDB
  .[grep("^HALLMARK_|^KEGG_|^REACTOME_|^PID_|^BIOCARTA_",
         names(.))] # Pull curated gene sets by their database/primary identifier
```

## GSVA calculation
```{r warning=F,message=F}
## Select key words to further subset the selected MSigDB gene set collection:
gs.kw <- c(
  "REACTOME",
  "PID",
  "BIOCARTA",
  "KEGG",
  "HALLMARK"
  )

## Now get the gene sets matching the above key words from the MSigDB list:
store <- str_subset(names(gene_sets), paste(gs.kw, collapse = "|"))

## Pull the gene sets we want:
gene.sets <- store %>% 
  na.omit() %>% 
  match(names(gene_sets)) %>% 
  na.omit() %>% unique() %>% 
  gene_sets[.]

#### Expression data prep ####
### Match BCAR3 expression data to cell lines and format correctly to append to GSVA data above, if we want to:
b3.data_for_heatmap <- match(colnames(data_for_gsva), tcl.ccle.rna$Cell.line) %>%
  tcl.ccle.rna$BCAR3[.] %>% scale %>% data.frame(BCAR3.mRNA=.) %>% 
  mutate(Cell.Subtype = match(colnames(data_for_gsva),tcl.info$stripped_cell_line_name) %>% tcl.info$lineage_molecular_subtype[.])
rownames(b3.data_for_heatmap) <- colnames(data_for_gsva)
b3.data_for_heatmap$Cell.Subtype[b3.data_for_heatmap$Cell.Subtype==""] <- "basal_B"

### If desired, define phenotypic/morphological subtypes to perform GSVA on. Comment/uncomment the lines of code below depending on whether you would like to do use this option:
subtypes_to_use <- c("basal_A","basal_B") %>% paste(collapse = "|")
cell.subtype.inds <- str_detect(b3.data_for_heatmap$Cell.Subtype,
                                 subtypes_to_use
                                 )
cell.subtype <- b3.data_for_heatmap$Cell.Subtype[cell.subtype.inds] %>% factor
data_for_gsva <- rownames(b3.data_for_heatmap)[cell.subtype.inds] %>% 
  match(tcl.ccle.rna$Cell.line) %>% 
  tcl.ccle.rna[.,-1] %>% t()
colnames(data_for_gsva) <- rownames(b3.data_for_heatmap)[cell.subtype.inds] # Col names = cell lines, matching subtype vector
rownames(data_for_gsva) <- colnames(tcl.ccle.rna)[-1] %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') # Row names = Entrez gene IDs

  
#### Calculate GSVA enrichment scores ####
tn.gsva <- gsva(
  data_for_gsva, 
  gene.sets, 
  method="gsva", 
  parallel.sz=detectCores()-1
  )
tn.gsva.df <- tn.gsva %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
# tn.gsva.df$gene.set <- gsub(".*HALLMARK_|.*REACTOME_|.*PID_|.*KEGG_", "", tn.gsva.df$gene.set) #Unused:
rownames(tn.gsva.df) <- tn.gsva.df$gene.set
# rm(tn.gsva) # Remove matrix form from environment (for memory)

# Get data to plot for heatmap:
plotdata <- tn.gsva.df[,-1] %>% data.matrix()
rownames(plotdata) <- rownames(tn.gsva.df)
```

## Heatmap visualization of enrichment scores
Now, let's visualize the enrichment score results from GSVA via a heatmap.

```{r Visualize GSVA Results, message=FALSE, warning=FALSE}
#### Visualize with heatmaply:
## Load packages:
# library(heatmaply)
library(pheatmap)

## Set color scheme for heat map:
ncols <- 11
# mycol <- brewer.pal(n=11, name="RdBu") %>% rev() # Color scheme for heatmap
mycol <- colorRampPalette(c("#0000cc","#cccccc","#ff0000"))(ncols) # Custom color scheme
mycol <- pals::coolwarm(ncols) # Red-white-blue color scheme
mycol <- colorRampPalette(c("red","black","green"))(ncols) # Custom color scheme

## Color mapping for side bars on heatmaps of GSVA ESs:
ann_colors <- list(
  Cell.Subtype = c(basal = "#61D04F", basal_A = "#2297E6", basal_B = "#DF536B", HER2_amp = "#28E2E5", luminal = "#CD0BBC", luminal_HER2_amp = "#F5C710"),
  BCAR3.mRNA = pals::coolwarm(n=20)
)

#### Generate heatmaps ####
distmeth <- "euclidean"
pheatmap(
  plotdata,
  clustering_distance_cols = distmeth,
  clustering_distance_rows = distmeth,
  clustering_method = "ward.D2",
  color = mycol,
  # border_color = "black",
  treeheight_row = 0.3,
  treeheight_col = 0.3,
  show_rownames = F,
  # annotation_row = annotations_row, 
  annotation_col = b3.data_for_heatmap,
  fontsize_row = 6,
  fontsize_col = 9,
  annotation_colors = ann_colors,
  # cutree_rows = 2, # Divide heatmap rows into sections based on dendrogram groups
  cutree_cols = 4, # Divide heatmap columns into sections based on dendrogram groups
  cellwidth = 10,
  angle_col = 90,
  main = "CCLE RNA-seq GSVA ESs",#, Maximum Difference Method",
  # file = "GSVA1_CCLE_RNAseq_TNBC.png", # Save the heatmap
  )

```

## DEA

For this analysis, we just want to look at the basal-like (Basal A) and claudin-low (Basal B) cell lines, so we just pull out those cell lines for this analysis.
```{r Differental expression analysis, message=F,warning=FALSE}
#### Differential expression analysis ####
## Define statistical thresholds:
adjPvalueCutoff1 <- 0.01 # Threshold for adjusted p-values (FDR)
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)

### Define phenotypic/morphological subtypes: Basal-like (BL/basal_A) or claudin-low/mesenchymal-like (M/basal_B)
## Get gene expression data for desired subtypes (defined in cell w/ GSVA calculations):
data_for_lm <- rownames(b3.data_for_heatmap)[cell.subtype.inds] %>% 
  match(tcl.ccle.rna$Cell.line) %>% 
  tcl.ccle.rna[.,-1] %>% t()
colnames(data_for_lm) <- rownames(b3.data_for_heatmap)[cell.subtype.inds] # Col names = cell lines, matching subtype vector
rownames(data_for_lm) <- colnames(tcl.ccle.rna)[-1] %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') # Row names = Entrez gene IDs

## Fit linear model with limma: CCLE RNA-seq gene expression data
design1 <- model.matrix(~ 0 + cell.subtype)
colnames(design1) <- c("basal_A","basal_B")
fit1 <- lmFit(data_for_lm, design1) 
contr1 <- makeContrasts(basal_B - basal_A, # Make contrasts --> compare groups of interest; here, want to know what is differentially expressed in basal_B versus basal_A (mesenchymal-like versus epithelial-like)
                        levels = colnames(coef(fit1))) 
tmp1 <- contrasts.fit(fit1, contr1)
tmp1 <- eBayes(tmp1)
allGenes <- topTable(tmp1, sort.by = "P", number=Inf)
allGenes$ID <- allGenes$ID %>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
DEgenes <- topTable(tmp1, number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
DEgenes$ID <- DEgenes$ID %>% mapIds(org.Hs.eg.db, ., "SYMBOL", "ENTREZID")
res1 <- decideTests(tmp1, p.value=adjPvalueCutoff1)
summary(res1)

## Linear model with limma: GSVA enrichment scores
adjPvalueCutoff2 <- 0.05 # Threshold for adjusted p-values (FDR)
design2 <- model.matrix(~ 0 + cell.subtype)
colnames(design2) <- c("basal_A","basal_B")
fit2 <- lmFit(plotdata, design2)
contr2 <- makeContrasts(basal_B - basal_A, # Make contrasts --> compare groups of interest
                        levels = colnames(coef(fit2)))
tmp2 <- contrasts.fit(fit2, contr2)
tmp2 <- eBayes(tmp2)
allGeneSets <- topTable(tmp2, sort.by = "P", number=Inf)
allGeneSets$ID <- rownames(allGeneSets)
DEgeneSets <- topTable(tmp2, sort.by = "P", number=Inf,
 p.value=adjPvalueCutoff1, adjust="BH")
res2 <- decideTests(tmp2, p.value=adjPvalueCutoff2)
summary(res2)

DEgenes
DEgeneSets


#### Plots ####
### Volcano plot --> MvsBL genes
ggplot(data = allGenes, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff1),
    col = "black", linetype = "dashed", alpha = 0.75, size = 0.75
  ) +  
  geom_vline(xintercept = -logFCcutoff1, col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) +  
  geom_vline(xintercept = logFCcutoff1,col = "black",linetype = "dashed",alpha = 0.75,size = 0.75) + 
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff1 & abs(allGenes$logFC) > logFCcutoff1 %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5,
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff1),
                label = paste("p = ",adjPvalueCutoff1)), nudge_y = 0.06, size = 3, color = "red") +
  geom_text_repel(
    data = function(x){subset(x, (adj.P.Val < adjPvalueCutoff1) & (abs(logFC) > logFCcutoff1))},
    # position = "dodge",
    size = 2, alpha = 0.9, segment.size = 0.25, segment.alpha = .8,
    max.overlaps = 10, max.iter = 5000
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 18),
    axis.title = element_text(size = 18),
    plot.title = element_text(size = 18),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "DEA Genes: Mesenchymal vs. basal-like",
    x = "Expression log2FC",
    y = "-log10(adj. p-value)"
  ) +
  guides(
    color = FALSE
  )
#ggsave("TNBC_MvsBL_DEA-genes.png", width = 12, height = 8) # To most easily view the text labels, save the plot as an image and view outside the notebook



### Volcano plot --> MvsBL gene sets
ggplot(data = allGeneSets, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff2),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff2 %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff2),
                label = paste("p = ",adjPvalueCutoff2)), size = 3, color = "red") +
  geom_text_repel(
    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff2)},
    position = "dodge",
    size = 2, alpha = 0.9,
    max.overlaps = 10, segment.size = 0.25, segment.alpha = .8,
    force = 1
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 18),
    axis.title = element_text(size = 18),
    plot.title = element_text(size = 18),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "DEA Gene Sets: Mesenchymal vs. basal-like",
    x = "GSVA enrichment score log2FC",
    y = "-log10(adj. p-value)"
  ) +
  guides(
    color = FALSE
  )
#ggsave("TNBC_MvsBL_DEA-genesets.png", width = 12, height = 8)
```
