---
title: "TCGA_CCLE_GBM"
output: html_notebook
---
All TCGA data downloaded from UCSC Xena portal, with phenotypic info from NCI GDC portal
All CCLE data downloaded from DepMap portal

```{r  Load packages, message=FALSE, warning=FALSE}
# BiocManager::install(c("limma","GSVA","GSVAdata","org.Hs.eg.db","GSEABase","snow","rlecuyer","edgeR", "gage", "GO.db")) # Comment out this line if you've already installed these packages to your current R distribution.
### Load packages:
library(tidyverse)
library(reshape2)
library(ggthemes)
library(ggpubr)
library(pls)
library(cowplot)
library(plotly)
library(RColorBrewer)
library(colorspace)
library(jcolors)
library(ggsci)
library(cowplot)
library(Hmisc)
library(stats)
library(NMF)
library(ggfortify)
library(ggrepel)
library(GSVA) # Gene set variation analysis package from Bioconductor --> This package contains a command that can perform four types of gene set enrichment analyses: GSVA, ssGSEA, PLAGE, and z-score.
library(org.Hs.eg.db)
library(GSVAdata)
library(GSEABase)
library(limma)
library(msigdbr)
library(gage)
library(plyr)
library(GO.db)
library(heatmaply)
library(plsVarSel)
library(paletteer)

```

```{r Clear environment}
### Clear R workspace:
rm(list=ls())
```

```{r Check and set working directory}
### Check current working directory:
cwd <- getwd()
cwd
### Set working directory:
setwd(cwd) # Set current working directory to location of this notebook for easier handling and saving of data associated with the notebook
```

#load data sets
```{r Load TCGA RPPA GBM data}
## Load RPPA data for TCGA GBM samples
tcga.rppa.path <- "RPPA_RBN.txt"
tcga.rppa <- read_tsv(tcga.rppa.path)
tcga.rppa <- column_to_rownames(tcga.rppa, var="Sample_description")
tcga.rppa <- t(tcga.rppa)
#rppa <- rownames_to_column(as.data.frame(rppa), var="sample")

```
```{r Load TCGA RNA-seeq data}
tcga.rna.path <- "HiSeqV2.txt"
tcga.rna.data <- read_tsv(tcga.rna.path)
tcga.rna.data <- column_to_rownames(tcga.rna.data, var="sample")
tcga.rna.data <- t(tcga.rna.data)
```

```{r splitting TCGA RNA-seq data into subtypes}
subtype.info <- "gbm_tcga_pub2013_clinical_data.tsv"
subtype <- read_tsv(subtype.info)
subtype <- dplyr::rename(subtype, `sample`=`Sample ID`)
tcga.sub <- filter(subtype, `Gene Expression Subtype`=="Mesenchymal"|`Gene Expression Subtype`=="Proneural")

tcga.rna.sub <- tcga.rna.data %>% as.data.frame()
tcga.rna.sub <- rownames_to_column(tcga.rna.sub, var="sample")
tcga.rna.sub <- semi_join(tcga.rna.sub, tcga.sub, by="sample") 
tcga.sub <- semi_join(tcga.sub, tcga.rna.sub, by="sample") %>% column_to_rownames(var="sample")
tcga.rna.sub <- tcga.rna.sub %>% column_to_rownames(var="sample")
```

```{r Load CCLE cell line info, warning=FALSE}
##### Load cell line info:
## Path for cell line info:
fn.cells <- "CCLE_sample_info.csv" 

### Load info for cell lines:
cl.sum <- read.csv(fn.cells)   #Cell line summaries

## Desired cell line lineages:
glio.lin <- "central_nervous_system"
cell.info <- which(cl.sum$lineage %in% glio.lin) %>% cl.sum[.,] # Info of cell lines of interest

a <- split(cell.info, with(cell.info, lineage), drop = TRUE)

## Pull out each specific cancer:

glio <- a$central_nervous_system %>% split(., with(.,lineage_sub_subtype)) %>% .$glioblastoma

rm(a)
```

Load CCLE RNA-seq data
```{r Load CCLE RNA-seq data, warning=FALSE, message=F}
##### Load CCLE RNA-seq data: log2(TPM+1) values
## File path for CCLE RNA-seq data:
fn4 = "CCLE_RNAseq_gene_expression_TPM_log2.csv" 

### Load CCLE RNA-seq data:
ccle.rna <- read_csv(fn4, col_names=T)
ccle.genes <- colnames(ccle.rna)[-1] %>%
  gsub("\\ .*","",.) # Remove extraneous suffixes on gene names
colnames(ccle.rna) <- c("Cell.line", ccle.genes)
```

```{r pull out GBM data from CCLE RNA-seq data,message = F,warning=F}
## Get desired cell line data:
ccle.id.rna <- ccle.rna$Cell.line %>% as.character() # Get DepMap IDs from CCLE data

#### GLIOMA ####
## Pull out glioma cell line IDs, names, and data:
glio.ind.ccle.rna <- which(ccle.id.rna %in% glio$DepMap_ID) # Match glioma cell line DepMap IDs to DepMap IDs in data set
glio.id.ccle.rna <- ccle.id.rna[glio.ind.ccle.rna] # DepMap IDs in data set
glio.name.ccle.rna <- match(glio.id.ccle.rna, glio$DepMap_ID) %>%
  glio$stripped_cell_line_name[.] # Names

# Pull out the glioma data:
glio.ccle.rna <- ccle.rna[glio.ind.ccle.rna,] # mRNA expression
glio.ccle.rna$Cell.line <- glio.name.ccle.rna # Replace DepMap IDs with stripped cell line names
```

## CCLE RPPA Data
Reverse-phase protein array (RPPA) data from the CCLE can also be obtained from either the primary CCLE website or from the DepMap Portal.

```{r Load CCLE RPPA data, warning = F, message = F}
##### Load CCLE RPPA data: 
## File path for CCLE RPPA data:
fn5 = "CCLE_RPPA_20181003.csv" # Note, as written, this file has been stored one folder up from the current directory in which this notebook file resides -- remove the ../ if the data are in the same directory as this notebook

## Load CCLE RPPA data:
ccle.rppa <- read_csv(fn5, col_names=T)
ccle.ans <- colnames(ccle.rppa)[-1] %>% 
  gsub("\\_Caution.*","",.) # Remove extraneous "_Caution" suffixes on analyte names
colnames(ccle.rppa) <- c("Cell.line", ccle.ans)

## Get cell line names in RPPA data:
ccle.names.rppa <- ccle.rppa$Cell.line %>% as.character() # Get CCLE names from CCLE RPPA data

## Pull out the glio IDs, names, and data:
glio.ind.ccle.rppa <- which(ccle.names.rppa %in% glio$CCLE_Name) # Match glio cell line DepMap IDs to DepMap IDs in data set
glio.names.ccle.rppa <- ccle.names.rppa[glio.ind.ccle.rppa] %>% 
  gsub("\\_.*","",.) # Cleaned (stripped) CCLE cell line names

## Pull out the glio data:
glio.ccle.rppa <- ccle.rppa[glio.ind.ccle.rppa,] # Data
glio.ccle.rppa$Cell.line <- glio.names.ccle.rppa # Replace DepMap IDs with stripped cell line names
```
```{r PLSR metrics function}
library(pls)
### R2, Q2, and VIP helper function for extracting PLSR model performance metrics:
rq_func <- function(model){
  ### Get X and Y matrices:
  xmat <- model$model$X
  ymat <- model$model$Y
  ### Performance metrics:
  R2X <- explvar(model)
  names(R2X) <- seq(1:length(R2X)) %>% paste("PC",.,sep="")
  R2X.cum <- cumsum(R2X)
  PRESS <- model$validation$PRESS
  ## Calculate R2Y values for PLSR model:
  model.TSS <- sum((ymat-mean(ymat))^2)
  R2Y <- 0
  R2Y.cum <- 0
  for(i in 1:model$ncomp){
    R2Y.cum[i] <- (1-sum(model$residuals[,,i]^2)/model.TSS)*100
    if(i == 1){
     R2Y[i] <- R2Y.cum[i] 
    } 
    else{
      R2Y[i] <- R2Y.cum[i]-R2Y.cum[i-1]
    }
  }
  ## Calculate Q2Y values for PLSR model:
  Q2Y <- 0
  Q2Y.cum <- 0
  for(i in 1:model$ncomp){
    Q2Y.cum[i] <- (1-sum(PRESS[,i])/model.TSS)*100
    if(i == 1){
      Q2Y[i] <- Q2Y.cum[i] 
      } 
    else{
      Q2Y[i] <- Q2Y.cum[i]-Q2Y.cum[i-1]
      }
  }
  ### Define output list:
  output <- list(R2X=R2X, R2X.cum=R2X.cum, R2Y=R2Y, R2Y.cum=R2Y.cum, Q2Y=Q2Y, Q2Y.cum=Q2Y.cum)
  return(output)
}


```

**This is where the analysis of the data actually starts**

```{r TCGA RPPA PLSR}

rppa.prot <- tcga.rppa
rppa.prot <- rppa.prot %>% scale(center=T, scale=T) %>% as.data.frame() #mean-center and variance scaling the data
#pull out proteins to be used as phenotypic markers
proteins.y <- c( "FIBRONECTIN", "PAI1", "COLLAGENVI")
y.mat <- data.matrix(dplyr::select(rppa.prot, proteins.y))
x.mat <- data.matrix(dplyr::select(rppa.prot, -proteins.y))

#run PLSR
mydata <- data.frame(Y=I(y.mat),X=I(x.mat))
pls1 <- plsr(Y~X, scale=F, data=mydata, validation = "LOO", method="kernelpls")
#summary(pls1)

#filtering points based on VIP scores
score.min <- 1
vip.scores <- VIP(pls1, 3) %>% data.frame()
colnames(vip.scores)[1] <- "VIP.score"
vip.scores <- rownames_to_column(vip.scores, var="protein")
vip.scores <- vip.scores %>% filter(VIP.score>score.min)
proteins.VIP <- vip.scores$protein

#running PLSR again using only proteins w/ high VIP scores
x.mat <- data.matrix(dplyr::select(rppa.prot, proteins.VIP))

mydata <- data.frame(Y=I(y.mat),X=I(x.mat))
pls1 <- plsr(Y~X, scale=F, data=mydata, validation = "LOO", method="kernelpls")

plot(pls1, plottype = "scores")
sc <- scores(pls1)
ld <- loadings(pls1)
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)
vip.scores <- vip.scores %>% cbind(ld[,1])
colnames(vip.scores)[ncol(vip.scores)] <- "PC1"
ld <- cbind(ld, vip.scores$VIP.score)

#plotting proteins with VIP scores over threshold
ggplot(data=arrange(vip.scores, desc(VIP.score)), 
       aes(x=reorder(protein, -VIP.score), y=VIP.score, fill=PC1))+
  geom_bar(stat="identity")+
    theme_cowplot()+
  theme(axis.text.x=element_text(angle=90, vjust=.5, hjust=1),
        panel.background = element_rect(color = "#000000")
        )+
  scale_fill_gradient2( # Custom color gradient for bars
    midpoint=0,
    low="red",
    mid="#555555",
    high="blue",
    aesthetics="fill"
  ) +
  labs(title = paste0("VIP Scores > ", score.min),
       x = element_blank(),
       y = "VIP score")
#ggsave("TCGA_RPPA_VIP.png", width=8, height=6)

#loadings plot
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], size = (ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)])), alpha=0.5)+
  geom_point(aes(x=Yld[,1], y=Yld[,2]), color="blue")+
  labs(title="GBM TCGA RPPA PLSR", 
       x="PC1", 
       y="PC2")+
#       x=paste0("PC1", " (",explvar(pls1)[1]%>%round(1), "%)"), 
#       y=paste0("PC2", " (",explvar(pls1)[2]%>%round(1), "%)"))+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#222222",
                  size=3)+
  geom_text_repel(aes(x=Yld[,1], y=Yld[,2]), 
            label=rownames(Yld), size=3, color="blue")+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.1))
#ggsave("TCGA_RPPA_PLSR.png", width=6, height=4)
#ggsave("TCGA_RPPA_PLSR.svg", width=4, height=4)


pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
    geom_vline(xintercept = comp, linetype="dashed") + geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  theme(panel.background = element_rect(color = "#000000"))+
  scale_x_continuous(limits=c(0,20))+
  scale_y_continuous(limits = c(0,100))
#ggsave("TCGA_PLSR_var-expl.png", width = 5, height = 4)
```

```{r CCLE RPPA PLSR}
#get and format CCLE RPPA data
rppa.prot <- glio.ccle.rppa
rppa.prot <- rppa.prot %>% column_to_rownames(var="Cell.line")
rppa.prot <- rppa.prot %>% scale(center=T, scale=T) %>% as.data.frame()
rppa.prot <- rppa.prot %>% .[, colSums(is.na(.))==0]

#pull out proteins to be used as phenotypic markers
proteins.y <- c( "Fibronectin", "PAI-1", "Collagen_VI")
Y <- data.matrix(dplyr::select(rppa.prot, proteins.y))
X <- dplyr::select(rppa.prot, -proteins.y) %>% as.matrix()

#run PLSR
mydata <- data.frame(yy=I(Y),xx=I(X))
PLSR_matrix <- plsr(yy~xx, scale=F, data=mydata, method="kernelpls")
#summary(PLSR_matrix)

#filtering points based on VIP scores
score.min <- 1.2
vip.scores <- VIP(PLSR_matrix, 5) %>% data.frame()
colnames(vip.scores)[1] <- "VIP.score"
vip.scores <- rownames_to_column(vip.scores, var="protein")
vip.scores <- vip.scores %>% filter(VIP.score>score.min)
proteins.VIP <- vip.scores$protein
#running PLSR again using only proteins w/ high VIP scores

X <- data.matrix(dplyr::select(rppa.prot, proteins.VIP))
mydata <- data.frame(yy=I(Y),xx=I(X))
PLSR_matrix <- plsr(yy~xx, scale=F, data=mydata, method="kernelpls")
#plot(PLSR_matrix, plottype = "scores")
sc <- scores(PLSR_matrix)
ld <- loadings(PLSR_matrix)
Ysc <-Yscores(PLSR_matrix)
Yld <- Yloadings(PLSR_matrix)
ld <- cbind(ld, vip.scores$VIP.score)
vip.scores <- vip.scores %>% cbind(ld[,1])
colnames(vip.scores)[ncol(vip.scores)] <- "PC1"

#plotting proteins by VIP score
ggplot(data=arrange(vip.scores, desc(VIP.score)), aes(x=reorder(protein, -VIP.score), y=VIP.score, fill=PC1))+
  geom_bar(stat="identity")+
  theme_cowplot()+
  theme(axis.text.x=element_text(angle=90, vjust=.5, hjust=1),
        panel.background = element_rect(color = "#000000")
        )+
  scale_fill_gradient2( # Custom color gradient for bars
    midpoint=0,
    low="red",
    mid="#555555",
    high="blue",
    aesthetics="fill"
  ) +
  labs(title = paste0("VIP scores > ", score.min),
       x = element_blank(),
       y = "VIP score")
#ggsave("CCLE_RPPA_VIP.png", width=8, height=6)

#loadings plot
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], size = (ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)])), alpha=0.5)+
  geom_point(aes(x=Yld[,1], y=Yld[,2]), color="blue")+
  labs(title="GBM CCLE RPPA PLSR", 
       x=paste0("PC1", " (",explvar(PLSR_matrix)[1]%>%round(1), "%)"), 
       y=paste0("PC2", " (",explvar(PLSR_matrix)[2]%>%round(1), "%)"))+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#222222",
                  size=3)+
  geom_text_repel(aes(x=Yld[,1], y=Yld[,2]), 
            label=rownames(Yld), 
            size=3, 
            color="blue")+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.1))
#ggsave("CCLE_RPPA_PLSR.png", width=6, height=4)

pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
    geom_vline(xintercept = comp, linetype="dashed") + geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")

```

```{r TCGA RNA-seq PLSR}
#assigning genes to either proneural (+1) or mesenchymal (-1) - this just makes things work later on
genes.y <- rbind(
  c("FN1", 1),
  c("DLL3", -1),
  c("SOX2", -1),
  c("ERBB3", -1),
  c("OLIG2", -1),
  c("CHI3L1", 1),
  c("TRADD", 1), 
  c("TLR2", 1),
  c("RELB", 1)
#  c("FGFR3", 0),
#  c("PDGFA", 0),
#  c("EGFR", 0),
#  c("AKT2", 0),
#  c("NES", 0)
)
colnames(genes.y) <- c("gene", "survival")
genes.y.surv <- genes.y[,2] %>% as.numeric()

#run PLSR
tcga.rna <- as.data.frame(tcga.rna.data)
tcga.rna <- tcga.rna %>% scale(center=T, scale=T) %>% data.frame()
#tcga.rna <- tcga.rna %>% select_if(~ !any(is.na(.)))
tcga.rna <- tcga.rna %>% .[, colSums(is.na(.))==0]
tcga.rna.samples <- rownames(tcga.rna)

Y <- data.matrix(dplyr::select(tcga.rna, genes.y[,1]))
X <- dplyr::select(tcga.rna, -genes.y[,1]) %>% data.matrix()

mydata <- data.frame(yy=I(Y),xx=I(X))
PLSR_matrix <- plsr(yy~xx, scale=F, data=mydata, method="kernelpls")
#summary(PLSR_matrix)

#filtering points based on VIP scores
score.min <- 1.8
vip.scores <- VIP(PLSR_matrix, 5) %>% data.frame()
colnames(vip.scores)[1] <- "VIP.score"
vip.scores <- rownames_to_column(vip.scores, var="protein")
vip.scores <- vip.scores %>% filter(VIP.score>score.min)
proteins.VIP <- vip.scores$protein
#running PLSR again using only proteins w/ high VIP scores

X <- data.matrix(dplyr::select(tcga.rna, proteins.VIP))
mydata <- data.frame(yy=I(Y),xx=I(X))
PLSR_matrix <- plsr(yy~xx, scale=F, data=mydata, method="kernelpls")
plot(PLSR_matrix, plottype = "scores")
sc <- scores(PLSR_matrix)
ld <- loadings(PLSR_matrix)
Ysc <-Yscores(PLSR_matrix)
Yld <- Yloadings(PLSR_matrix)

Yld <- cbind(Yld, genes.y.surv)
ld <- cbind(ld, vip.scores$VIP.score)
colnames(ld)[ncol(ld)] <- "VIP.score"

vip.scores <- vip.scores %>% cbind(ld[,1])
colnames(vip.scores)[ncol(vip.scores)] <- "PC1"

ggplot(data=arrange(vip.scores, desc(VIP.score)), 
       aes(x=reorder(protein, -VIP.score), y=VIP.score, fill=PC1))+
  geom_bar(stat="identity")+
  theme_cowplot()+
    theme(axis.text.x=element_text(angle=90, vjust=.5, hjust=1),
        panel.background = element_rect(color = "#000000")
        )+
  scale_fill_gradient2( # Custom color gradient for bars
    midpoint=0,
    low="red",
    mid="#555555",
    high="blue",
    aesthetics="fill"
  ) +
  labs(title = paste0("VIP Scores > ", score.min),
       x = element_blank(),
       y = "VIP score")
#ggsave("TCGA_RNA_VIP.png", width=10, height=6)

#loadings plot
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], size=((ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)]))), alpha=0.5)+
  geom_point(aes(x=Yld[,1], y=Yld[,2], color=genes.y.surv))+
  labs(title="GBM TCGA RNA-seq PLSR", 
       x=paste0("PC1", " (",explvar(PLSR_matrix)[1]%>%round(1), "%)"), 
       y=paste0("PC2", " (",explvar(PLSR_matrix)[2]%>%round(1), "%)"))+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  size=3)+
  geom_text_repel(aes(x=Yld[,1], y=Yld[,2], color=genes.y.surv), 
            label=rownames(Yld), 
            size=3)+
  scale_color_gradient2( # Custom color gradient for bars
    limits = c(-1,1),
    midpoint=0,
    high="#0033ff",
    mid="#8f00ff",
    low="#ff3300"
  ) +
  scale_size(
    limits=c(0,1),
  )+
  theme_cowplot() +
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.05))
#ggsave("TCGA_RNA_PLSR.png", width=6, height=4)

tcga.rna.ld <- ld

```
##survival PLSR


```{r PLSR using survival vs a subset of genes}
library(readxl)
verhaak_genes <- read_csv("verhaak_subtypes.csv")
verhaak_m <- verhaak_genes %>% dplyr::filter(subtype == "M") %>% pull(var = "gene")
verhaak_p <- verhaak_genes %>% dplyr::filter(subtype == "P") %>% pull(var = "gene")
verhaak_all <- verhaak_genes %>% pull(var = "gene")

phillips_mes <- read_excel("mmc4.xls", sheet = "MES") %>% pull(`Gene Symbol`)
phillips_pn <- read_excel("mmc4.xls", sheet = "PN") %>% pull(`Gene Symbol`)

mes_genes <- c(verhaak_m, phillips_mes)
pn_genes <- c(verhaak_p, phillips_pn)

#genes.x <- c(verhaak_all, phillips_mes, phillips_pn)
#genes.x <- c(phillips_mes, phillips_pn)
#genes.x <- c ("FN1", "SERPINE1", "CHI3L1", "CD44", "TIMP1", "OLIG2", "MYB", "CNTN1", "NCAM1", "COL1A2")
genes.x <- c("CHI3L1", "SERPINE1", "TIMP1", "OLIG2", "MYB", "CNTN1", "DLL3")
#genes.x <- c("OSMR", "LOC151300", "EPHB1", "TIMP4", "BMP2", "GPR27", "NTN4")
#genes.x <- genes1

survival.mat <- subtype %>% dplyr::select(sample, `Overall Survival (Months)`) %>% filter(`Overall Survival (Months)`>1)

#run PLSR
tcga.rna <- as.data.frame(tcga.rna.data)
tcga.rna <- tcga.rna %>% 
  scale(center=T, scale=T) %>% 
  data.frame()
#tcga.rna <- tcga.rna %>% select_if(~ !any(is.na(.)))
tcga.rna <- tcga.rna %>% .[, colSums(is.na(.))==0]
tcga.rna.samples <- rownames(tcga.rna)
tcga.rna <- tcga.rna %>% rownames_to_column(var = "sample")
tcga.rna <- tcga.rna %>% semi_join(survival.mat, by = "sample") %>% arrange(sample)

survival.mat <- survival.mat %>% semi_join(tcga.rna, by = "sample") %>% 
  arrange(sample) %>% 
  column_to_rownames(var = "sample") 
tcga.rna <- tcga.rna %>% column_to_rownames(var = "sample") 

x.mat <- dplyr::select(tcga.rna, any_of(genes.x)) %>% data.matrix()
y.mat <- survival.mat %>% data.matrix() %>% scale()


mydata <- data.frame(Y=I(y.mat),X=I(x.mat))
pls1 <- plsr(Y~X, scale=F, data=mydata, validation = "LOO", method="kernelpls")
#summary(pls1)

#filtering points based on VIP scores
score.min <- 0
vip.scores <- VIP(pls1, 5) %>% data.frame()
colnames(vip.scores)[1] <- "VIP.score"
vip.scores <- rownames_to_column(vip.scores, var="gene")
vip.scores <- vip.scores %>% filter(VIP.score>score.min)
genes.VIP <- vip.scores$gene
#running PLSR again using only genes w/ high VIP scores

x.mat <- data.matrix(dplyr::select(tcga.rna, genes.VIP))
mydata <- data.frame(Y=I(y.mat),X=I(x.mat))
pls1 <- plsr(Y~X, scale=F, data=mydata, validation = "LOO", method="kernelpls")
plot(pls1, plottype = "scores")
sc <- scores(pls1)
ld <- loadings(pls1)
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)

Yld <- cbind(Yld, genes.y.surv)
ld <- cbind(ld, vip.scores$VIP.score)
colnames(ld)[ncol(ld)] <- "VIP.score"

vip.scores <- vip.scores %>% cbind(ld[,1])
colnames(vip.scores)[ncol(vip.scores)] <- "PC1"

ggplot(data=arrange(vip.scores, desc(VIP.score)), 
       aes(x=reorder(gene, -VIP.score), y=VIP.score, fill=PC1))+
  geom_bar(stat="identity")+
  theme_cowplot()+
    theme(axis.text.x=element_text(angle=90, vjust=.5, hjust=1),
        panel.background = element_rect(color = "#000000")
        )+
  scale_fill_gradient2( # Custom color gradient for bars
    midpoint=0,
    low="red",
    mid="#555555",
    high="blue",
    aesthetics="fill"
  ) +
  labs(title = paste0("VIP Scores > ", score.min),
       x = element_blank(),
       y = "VIP score")
#ggsave("TCGA_RNA_VIP.png", width=10, height=6)

#loadings plot
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], size=((ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)]))), alpha=0.5)+
  geom_point(aes(x=Yld[,1], y=Yld[,2]))+
  labs(x="PC1", 
       y="PC2")+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  size=3)+
  geom_text_repel(aes(x=Yld[,1], y=Yld[,2]), 
            label="Survival", 
            size=3)+
  scale_size(
    limits=c(0,1),
  )+
  theme_cowplot() +
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.05))
#ggsave("TCGA_RNA_PLSR.png", width=6, height=4)

## Get performance metrics: R2X, R2Y, Q2Y
pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
    geom_vline(xintercept = comp, linetype="dashed") + geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")
plot(RMSEP(pls1), legendpos="topleft")
```

```{r linear modeling of survival data}
xy <- data.frame(x.mat,survival.mat)
colnames(xy)[ncol(xy)] <- "survival"

ols1 <- lm(log(survival) ~ .+1, data = xy)
summary(ols1)
plot(ols1)
plot(y.mat, ols1$fitted.values)
var1 <- step(ols1)
summary(var1)
genes1 <- names(var1$coefficients)[-1]
genes1.subtype <- genes1 %>% data.frame() %>% mutate(subtype = case_when(
  . %in% mes_genes ~ "MES",
  . %in% pn_genes ~ "PN",
  T ~ "??"
))
```

```{r Cox modeling}
library(glmnet)
genes.x <- c(verhaak_m, verhaak_p, phillips_mes, phillips_pn)
#genes.x <- c("CHI3L1", "SERPINE1", "TIMP1", "OLIG2", "MYB", "CNTN1", "DLL3")

x.mat <- dplyr::select(tcga.rna, any_of(genes.x)) %>% rownames_to_column(var = "sample")
#y.mat <- survival.mat %>% data.matrix() %>% scale()

surv.cox <- subtype %>% dplyr::select(sample, `Overall Survival (Months)`, `Overall Survival Status`) %>% 
  dplyr::filter(`Overall Survival (Months)` > 0)
colnames(surv.cox) <- c("sample","time", "survival")

surv.cox <- mutate(surv.cox, status = case_when(
                                       survival == "LIVING" ~ 0,
                                       survival == "DECEASED" ~ 1
                                     )) %>% 
  dplyr::select(-survival)



surv.cox <- surv.cox %>% semi_join(x.mat, by = "sample") %>% 
  arrange(sample) %>% 
  column_to_rownames(var = "sample") %>% data.matrix()
x.mat <- x.mat %>% column_to_rownames(var = "sample") %>% data.matrix()

fit1 <- glmnet(x.mat, surv.cox, family = "cox")
plot(fit1, label = T, xvar = "dev")
coef(fit1)
```

```{r CCLE RNA-seq PLSR}
ccle.rna <- glio.ccle.rna 
ccle.rna <- ccle.rna %>% column_to_rownames(var="Cell.line")
ccle.rna <- ccle.rna %>% scale(center=T, scale=T) %>% data.frame()
#ccle.rna <- ccle.rna %>% select_if(~ !any(is.na(.)))
ccle.rna <- ccle.rna %>% .[, colSums(is.na(.))==0]

genes.y <- rbind(
  c("FN1", 1),
  c("DLL3", -1),
  c("SOX2", -1),
  c("ERBB3", -1),
  c("OLIG2", -1),
  c("CHI3L1", 1),
  c("TRADD", 1), 
  c("TLR2", 1),
  c("RELB", 1)
#  c("FGFR3", 0),
#  c("PDGFA", 0),
#  c("EGFR", 0),
#  c("AKT2", 0),
#  c("NES", 0)
)

colnames(genes.y) <- c("gene", "survival")
genes.y.surv <- genes.y[,2] %>% as.numeric()

#genes.y <- c( "FN1", "SERPINE1", "COL6A1", "COL1A1", "COL1A2", "CHI3L1", "PDGFRA", "OLIG2", "SOX2")
Y <- data.matrix(dplyr::select(ccle.rna, genes.y[,1]))
X <- dplyr::select(ccle.rna, -genes.y[,1]) %>% data.matrix()

#run PLSR
mydata <- data.frame(yy=I(Y),xx=I(X))
pls1 <- plsr(yy~xx, scale=F, data=mydata, method="kernelpls")
#summary(pls1)

#filtering points based on VIP scores
score.min <- 1.5
vip.scores <- VIP(pls1, 40) %>% data.frame()
colnames(vip.scores)[1] <- "VIP.score"
vip.scores <- rownames_to_column(vip.scores, var="gene")
vip.scores <- vip.scores %>% filter(VIP.score>score.min)
genes.VIP <- vip.scores$gene
#running PLSR again using only genes w/ high VIP scores

X <- data.matrix(dplyr::select(ccle.rna, genes.VIP))
mydata <- data.frame(yy=I(Y),xx=I(X))
pls1 <- plsr(yy~xx, scale=F, data=mydata, method="kernelpls")
#plot(pls1, plottype = "scores")
sc <- scores(pls1)
ld <- loadings(pls1)
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)

Yld <- cbind(Yld, genes.y.surv)
ld <- cbind(ld, vip.scores$VIP.score)
colnames(ld)[ncol(ld)] <- "VIP.score"

vip.scores <- vip.scores %>% cbind(ld[,1])
colnames(vip.scores)[ncol(vip.scores)] <- "PC1"

ggplot(data=arrange(vip.scores, desc(VIP.score)), 
       aes(x=reorder(gene, -VIP.score), y=VIP.score, fill=PC1))+
  geom_bar(stat="identity")+
  theme_cowplot()+
  theme(axis.text.x=element_text(angle=90, vjust=.5, hjust=1),
        panel.background = element_rect(color = "#000000"),)+
  scale_fill_gradient2( # Custom color gradient for bars
    midpoint=0,
    high="red",
    mid="#555555",
    low="blue",
    aesthetics="fill"
  ) +
  labs(title = paste0("VIP scores > ", score.min),
       x = element_blank(),
       y = "VIP score")
#ggsave("CCLE_RNA_VIP.png", width=10, height=6)


ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=-ld[,1], y=ld[,2], size=((ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)]))), alpha=0.5)+
  geom_point(aes(x=-Yld[,1], y=Yld[,2], color=genes.y.surv))+
  labs(title="GBM CCLE RNA-seq PLSR", 
       x=paste0("PC1", " (",explvar(pls1)[1]%>%round(1), "%)"), 
       y=paste0("PC2", " (",explvar(pls1)[2]%>%round(1), "%)"))+
  geom_text_repel(aes(x=-ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  size=3)+
  geom_text_repel(aes(x=-Yld[,1], y=Yld[,2], color=genes.y.surv), 
            label=rownames(Yld), 
            size=3)+
  scale_color_gradient2( # Custom color gradient for bars
    limits = c(-1,1),
    midpoint=0,
    high="#0033ff",
    mid="#8f00ff",
    low="#ff3300"
  ) +
    scale_size(
    limits=c(0,1),
  )+
   theme_cowplot() +
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.05))
#ggsave("CCLE_RNA_PLSR.png", width=6, height=4)


ccle.rna.ld <- ld

ld.both <- intersect(tcga.rna.ld, ccle.rna.ld)
```

```{r Running GSVA on TCGA mes RNA-seq data}
data_for_gsva <- tcga.rna.sub %>% as.matrix() %>% t() %>% .[-1,]

## Convert gene names to Entrez IDs:
entrez_ids <- rownames(data_for_gsva) %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL')
rownames(data_for_gsva) <- entrez_ids

### Define MSigDB gene set collection to use:
gene_sets <- getGmt("msigdb.v7.0.entrez.gmt") %>% # Load all GSEA/Broad MSigDB gene sets
  .[grep("^HALLMARK|^KEGG|^REACTOME_SIGNALING_BY|^PID_|^BIOCARTA_", names(.))] # Pull curated gene sets by their database/primary identifier

gene_set_choice <- "GO" # Define this as either "GO" or something completely different (like "MSigDB") --> if "GO" is not used, then key words for MSigDB gene sets will be chosen automatically for the GSVA analysis below

go.list <- go.gsets()
go.sets <- go.list[["go.sets"]]
go.bp <- go.sets[1:16227]
### Load desired gene set(s) and perform GSVA:
if(gene_set_choice == "msigdb"){
  ### Calculate GSVA enrichment scores:
  tn.gsva <- gsva(
    data_for_gsva, 
    go.bp, 
    method="gsva", 
    min.sz = 10,
    parallel.sz=detectCores()-1
    )
  tn.gsva.df <- tn.gsva %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
  tn.gsva.df$gene.set <- gsub("GO.|[0-9]{7}", "", tn.gsva.df$gene.set) 
  rownames(tn.gsva.df) <- tn.gsva.df$gene.set
  rm(tn.gsva) # Remove matrix form from environment
  
} else {
  ## Select key words to further subset the selected MSigDB gene set collection:
gs.kw <- c(
  "REACTOME",
  "PID",
  "BIOCARTA",
  "KEGG",
  "HALLMARK"
  )

  store <- 0
  for (i in 1:length(gs.kw)) {
    temp <- str_subset(names(gene_sets), gs.kw[i])
    store <- c(store, temp)
  }

  store <- str_subset(names(gene_sets), paste(gs.kw, collapse = "|"))
  
  ## Pull the gene sets we want:
  gene.sets <- store[-1] %>% 
    na.omit() %>% 
    match(names(gene_sets)) %>% 
    na.omit() %>% 
    gene_sets[.]
    
  ### Calculate GSVA enrichment scores:
  tn.gsva <- gsva(
    data_for_gsva, 
    gene.sets, 
    method="gsva", 
    min.sz = 10,
    parallel.sz=detectCores()-1
    )
  
  tn.gsva.df <- tn.gsva %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
  #tn.gsva.df$gene.set <- gsub(".*HALLMARK_|.*REACTOME_|.*PID_|BIOCARTA_|.*KEGG_", "", tn.gsva.df$gene.set) #Unused: .*KEGG_|
  rownames(tn.gsva.df) <- tn.gsva.df$gene.set
  rm(tn.gsva) # Remove matrix form from environment
}

# Get data to plot:
plotdata <- tn.gsva.df[,-1]
plotdata <- plotdata %>% t() %>% data.frame() %>%
  rownames_to_column(var="sample") %>%
  arrange(sample) %>%
  column_to_rownames(var="sample") %>% t() %>% data.matrix()
rownames(plotdata) <- rownames(tn.gsva.df)

subtype.list <- tcga.sub$`Gene Expression Subtype` %>% data.frame()
rownames(subtype.list) <- rownames(tcga.sub)
design1 <- subtype.list %>% 
  transmute(
  mes = case_when(
  .=="Proneural" ~ 0,
  TRUE ~ 1
  ),
  pn = case_when(
    .=="Mesenchymal" ~ 0,
    TRUE ~ 1
  )
) %>% data.matrix()


```

```{r plot GSVA results, eval=FALSE}
## Set color scheme for heat map:
mycol <- brewer.pal(n=11, name="RdBu") %>% rev() # Color scheme for heatmap
#mycol <- colorRampPalette(c("#0000cc","#cccccc","#ff0000")) # Custom color scheme

plotdata.1 <- data.frame(pathway=rownames(plotdata.mes), mean=rowMedians(plotdata.mes))
pws <- top_n(plotdata.1, 10)


### Generate heatmap:
nk_col <- 4
hm <- 
heatmaply(
  # plotdata[1:40,],
  plotdata,
  k_col = nk_col, # How many column dendrogram clusters to color
  k_row = 8, # How many row dendrogram clusters to color
  limits = c(-max(abs(plotdata)),max(abs(plotdata))),
  #distfun_row = function(x) as.dist(1-cor(t(x), method = "spearman")),
  #distfun_col = function(x) as.dist(1-cor(x, method = "spearman")),
  # dist_method = "manhattan", # If NULL, Euclidean is used
  hclust_method = NA,
  # show_dendrogram = c(TRUE, FALSE),
  # show_dendrogram = c(FALSE, TRUE),
  show_dendrogram = c(TRUE, TRUE),
  branches_lwd = 0.3,
  colors=mycol,
  # scale = "row",
  # scale = "col",
  key.title="GSVA \nEnrichment Score",
  # showticklabels = c(FALSE, TRUE),
  # showticklabels = c(TRUE, FALSE),
  showticklabels = c(TRUE, TRUE),
  cexCol = 0.5, # Scale size of column tick/label text
  cexRow = 1, # Scale size of row tick/label text
  column_text_angle = 90,
  # ylab = "Sample",
  # xlab = "Sample",
  xlab = "TCGA Sample",
  main = "GSVA of RNA-seq Data",
  # file = "GSVA_CCLE_RNAseq_TNBC.html", # Save the heatmap to html
  )
hm
```

```{r Differental expression analysis, message=F,warning=FALSE}
#### Differential expression analysis ####
## Define statistical thresholds:
logFCcutoff1 <- log2(2) # Threshold for log2(fold-change)
adjPvalueCutoff <- 10^-2 # Threshold for adjusted p-values (FDR) - use for GO gene set
#adjPvalueCutoff <- 0.00000001 # use for other gene sets (KEGG, BIOCARTA, etc.)
## Linear model with limma: GSVA enrichment scores

fit2 <- lmFit(plotdata, design1)
contr2 <- makeContrasts(mes - pn, # Make contrasts --> compare groups of interest
                        levels = colnames(coef(fit2)))
tmp2 <- contrasts.fit(fit2, contr2)
tmp2 <- eBayes(tmp2)
allGeneSets <- topTable(tmp2, sort.by = "P", number=Inf)
allGeneSets$ID <- rownames(allGeneSets)
DEgeneSets <- topTable(tmp2, sort.by = "P", number=Inf,
 p.value=adjPvalueCutoff, adjust="BH")
res2 <- decideTests(tmp2, p.value=adjPvalueCutoff)
summary(res2)

#DEgenes
DEgeneSets
DEgeneSets$logFC %>% order() %>% DEgeneSets[.,] %>% write.csv("DEgene_sets.csv")

gene_sets_sub <- filter(allGeneSets, grepl("TGF|TNFA|_NFKB|HYPOXIA|HIF", ID)) %>% rownames()

#### Plots ####
### Volcano plots --> MvsP gene sets
ggplot(data = allGeneSets, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
#  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff),
#                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
  geom_text_repel(
    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff)},
    size = 4, alpha = 0.9,
    segment.size = 0.25, segment.alpha = .8, label.padding = 5, force=60, max.overlaps = 50, max.iter = 100000
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 25),
    plot.title = element_text(size = 23),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "Gene Sets: Mesenchymal vs. Proneural",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits=c(0,16))+
  scale_x_continuous(limits = c(-0.5, 1.2))
#ggsave("TCGA_mes_vs_pn_full.png", width = 10, height = 8)

#subset of volcano plot
ggplot(data = allGeneSets, 
       aes(x = logFC, y = -log10(adj.P.Val), label = ID)) +
  geom_hline(
    yintercept = -log10(adjPvalueCutoff),
    col = "black", linetype = "dashed", alpha = 0.6, size = 0.7
  ) +
  geom_point(
    aes(
        color = adj.P.Val < adjPvalueCutoff %>% as.numeric()
        ),
    # color = "red",
    alpha = 0.5,
    cex = 2.5
  ) +
#  geom_text(aes(x = 0, y = -log10(adjPvalueCutoff),
#                label = paste("p = ",adjPvalueCutoff)), size = 3, color = "red") +
  geom_text_repel(
    data = function(x){subset(x, adj.P.Val < adjPvalueCutoff)},
    size = 4, alpha = 0.9,
    segment.size = 0.25, segment.alpha = .8, max.iter = 10000, label.padding = 1, force=50, max.overlaps = 50
  ) +
  scale_color_manual(values = c("#999999","red")) +
  theme_cowplot(18,
    line_size = 0
  ) +
  theme(
    axis.text = element_text(size = 20),
    axis.title = element_text(size = 25),
    plot.title = element_text(size = 23),
    panel.background = element_rect(size = 2, color = "black", fill = "white"),
    ) +
  labs(
    title = "Gene Sets: Mesenchymal vs. Proneural",
    x = expression("GSVA enrichment score log"[2]*"(fold change)"),
    y = expression("-log"[10]*"(adj. p-value)")
  ) +
  guides(
    color = FALSE
  )+
  scale_y_continuous(limits=c(8,12))+
  scale_x_continuous(limits=c(0,1))
#ggsave("TCGA_mes_vs_pn_msigdb_subset.svg", width = 10, height = 8)

allGeneSets %>% data.frame() %>%  rownames_to_column(var = "pathway") %>% 
  top_n(50, logFC) %>% 
  arrange(logFC) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(pathway = factor(pathway, levels=pathway)) %>%   # This trick update the factor levels
  ggplot(aes(x = logFC, y = pathway, fill = adj.P.Val))+
  geom_bar(stat = "identity") + 
  scale_fill_gradient(low = "blue", high = "red")+
  theme_cowplot()+
#ggsave("TCGA_mes_vs_pn_GO_ranked.png", width = 15, height=15)
```

##GSVA
###load gene sets
```{r Load packages and collections, message=F, warning=F}
## Load packages:
library(clusterProfiler)
library(org.Hs.eg.db)
library(msigdbr)

### Define MSigDB gene set collection(s) to use --> retrieve with 'msigdbr' package:
species = "Homo sapiens"

## Retrieve Hallmark and canonical pathways collections in the database:
hall = msigdbr(species = species, category = "H")
cp = msigdbr(species = species, category = "C2", subcategory = "CP")
cp.b = msigdbr(species = species, category = "C2", subcategory = "CP:BIOCARTA")
cp.r = msigdbr(species = species, category = "C2", subcategory = "CP:REACTOME")
cp.p = msigdbr(species = species, category = "C2", subcategory = "CP:PID")
cp.k = msigdbr(species = species, category = "C2", subcategory = "CP:KEGG")
cp.cgp = msigdbr(species = species, category = "C2", subcategory = "CGP")
gene_sets1 <- rbind(hall, cp, cp.b, cp.r, cp.p, cp.k, cp.cgp) %>% split(x = .$entrez_gene, f = .$gs_name)
gene_sets3 <- rbind(hall, cp, cp.b, cp.r, cp.p, cp.k) %>% split(x = .$entrez_gene, f = .$gs_name)

## Go collections:
go.bp <- msigdbr(species = species, category = "C5", subcategory = "GO:BP")
go.cc <- msigdbr(species = species, category = "C5", subcategory = "GO:CC")
go.mf <- msigdbr(species = species, category = "C5", subcategory = "GO:MF")
gene_sets2 <- rbind(go.mf, go.cc, go.bp) %>% split(x = .$entrez_gene, f = .$gs_name)
gene_sets1 <- gene_sets1 %>% append(gene_sets2)

ph.m <- phillips_mes %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') %>% 
  na.omit() %>% as.integer()

ph.p <- phillips_pn %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') %>% 
  na.omit() %>% as.integer()

neftel_sets <- read_tsv("IDHwt.GBM.MetaModules.tsv")
neftel_mes <- append(pull(neftel_sets, MESlike1), pull(neftel_sets, MESlike2)) %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') %>% 
  na.omit() %>% as.integer()

neftel_pn <- neftel_sets %>% pull(OPClike) %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL') %>% 
  na.omit() %>% as.integer() 

neftel_m <- neftel_mes %>% as.character() %>% mapIds(org.Hs.eg.db, ., 'SYMBOL', 'ENTREZID') %>% as.character()
neftel_p <- neftel_pn %>% as.character() %>% mapIds(org.Hs.eg.db, ., 'SYMBOL', 'ENTREZID') %>% as.character()

gene_sets1[["PHILLIPS_MES"]] <- ph.m
gene_sets1[["PHILLIPS_PN"]] <- ph.p
gene_sets1[["NEFTEL_MES"]] <- neftel_mes
gene_sets1[["NEFTEL_PN"]] <- neftel_pn

```
###gsva calc
```{r}
tcga.rna <- as.data.frame(tcga.rna.data)
tcga.rna <- tcga.rna %>% 
  scale(center=T, scale=T) %>% 
  data.frame()
#tcga.rna <- tcga.rna %>% select_if(~ !any(is.na(.)))
tcga.rna <- tcga.rna %>% .[, colSums(is.na(.))==0]

data_for_gsva <- tcga.rna
#rownames(data_for_gsva) <- exp.symbol$GeneID
colnames(data_for_gsva) <- colnames(data_for_gsva) %>% 
  mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL')
genes <- colnames(data_for_gsva) %>% na.omit()
data_for_gsva <- data_for_gsva %>% dplyr::select(one_of(genes)) %>% t()

force_calculate_gsva <- T
if(!file.exists("gsva_results.rds") | isTRUE(force_calculate_gsva)){
  gsva_res <- gsva(
    data_for_gsva, 
    gene_sets1, verbose = F, # Change which gene sets are run
    # gene_sets4, verbose = T
    method = "gsva", 
#    kcdf = "Gaussian",
    # abs.ranking = F,
    # mx.diff = T,
    min.sz = 10, # Minimum number of genes required to include a gene set
    parallel.sz=detectCores()#-1
    )
  ## Save GSVA results for speed:
  saveRDS(gsva_res, file = "gsva_results.rds")
  
} else if(file.exists("gsva_results.rds") & !exists("gsva_res")){ 
  ## Load GSVA results, if they haven't been loaded/calculated already:
  gsva_res <- readRDS("gsva_results.rds") 
}
## Build data frame versions of GSVA results:
gsva_res.df <- gsva_res %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
gsva_res.df.t <- gsva_res %>% t() %>% data.frame(ID=rownames(.), .)
rownames(gsva_res.df) <- gsva_res.df$gene.set

```
###GSVA corr heatmap
```{r}
pmt_sets <- c(
  "HALLMARK_TGF_BETA_SIGNALING",
  "HALLMARK_HYPOXIA",
  "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
  "VERHAAK_GLIOBLASTOMA_MESENCHYMAL",
  "VERHAAK_GLIOBLASTOMA_PRONEURAL"
)

pmt_gsva <- gsva_res.df %>% subset(gene.set %in% pmt_sets) %>% dplyr::select(-gene.set)

### Draw heatmap:
pheatmap(pmt_gsva,
         clustering_method = "ward.D2", clustering_distance_rows = "canberra",
         # color = colorRampPalette(paletteer_d("RColorBrewer::PuOr", direction = -1))(101),
         color = colorRampPalette(paletteer_c("viridis::magma", n=101))(101),
         #cutree_rows = 2, 
         cutree_cols = 2,
         treeheight_row = 1, treeheight_col = 1, border_color = NA,
         # cellwidth = 7,
         cellheight = 25,
         show_colnames = F, angle_col = 90,
         fontsize = 8,
         main = "CPTAC GSVA: Enrichment scores",
         )

corr_data <- pmt_gsva %>% t() %>% data.frame(ID=rownames(.)) 

## Get all pairwise correlations:
gsva_corrs <- corr_data %>%
  dplyr::select(-ID) %>% 
  data.matrix() %>% 
  rcorr(type="spearman")

len_out <- 20
pheatmap(gsva_corrs$r,
         clustering_distance_rows = "euclidean", clustering_distance_cols = "euclidean",
         clustering_method = "ward.D2",
         show_colnames = T, show_rownames = T,
         treeheight_row = 0.3, treeheight_col = 0.3,
         color = colorRampPalette(pals::coolwarm(len_out))(len_out), 
         breaks = seq(-1,1,length.out = len_out),
         fontsize = 6, fontsize_number = 8,
         cellwidth = 30, cellheight = 30,
         angle_col = 90, border_color = NA, display_numbers = T,
         main="TCGA \nGSVA Enrichment Spearman correlations",
         #file="GSVA_corrs_heatmap.png"
         )


```

###GSVA-PLSR
```{r}
gsva_res.all <- gsva_res.df.t %>% select(-ID) %>% rownames_to_column(var = "sample") %>% arrange(sample)
survival.data <- survival.mat %>% rownames_to_column(var = "sample") %>% semi_join(gsva_res.all, by = "sample")
gsva_res.all <- gsva_res.all %>% semi_join(survival.data, by = "sample")
gsva_res.all %<>% column_to_rownames(var = "sample")
survival.data %<>% column_to_rownames(var = "sample")
colnames(survival.data) <- "Overall Survival"


pathways.y <- c("TGFb", "TNFa", "NFKB", "TGF_BETA")

x.mat <- gsva_res.all %>% select(-contains(pathways.y)) %>% as.matrix() %>% scale(center = T, scale =T)
#y.mat <- dat2 %>% as.matrix() %>% scale(center = T, scale =T)
y.mat <- gsva_res.all %>% select(contains(pathways.y)) %>% cbind(survival.data) %>% 
  as.matrix() %>% scale(center = T, scale = T)

mydata <- data.frame(Y=I(y.mat),X=I(x.mat))
pls1 <- plsr(Y~X, scale=F, data=mydata, validation = "LOO", method="kernelpls") # generate PLSR model

sc <- scores(pls1)
ld <- loadings(pls1)
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)

#VIP score calculations
vip.scores <- VIP(pls1, opt.comp = 3) %>% as.data.frame()
colnames(vip.scores) <- "score"
vip.scores %<>% rownames_to_column("analyte")

#plot VIP scores
ggplot(data=arrange(vip.scores, desc(score)), 
       aes(x=reorder(analyte, -score), y=score))+
  geom_bar(stat="identity")+
    theme_cowplot()+
  theme(axis.text.x=element_text(angle=90, vjust=.5, hjust=1),
        panel.background = element_rect(color = "#000000")
        )+
  scale_fill_gradient2( # Custom color gradient for bars
    midpoint=0,
    low="red",
    mid="#555555",
    high="blue",
    aesthetics="fill"
  )+
  geom_hline(yintercept = 1, linetype = "dashed")

#plot biplot loadings - full model
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], alpha=0.5))+
  geom_text(aes(x=Yld[,1], y=Yld[,2]), color="blue", label = rownames(Yld))+
  labs(title="", 
       x="PC1", 
       y="PC2")+
#       x=paste0("PC1", " (",explvar(pls1)[1]%>%round(1), "%)"), 
#       y=paste0("PC2", " (",explvar(pls1)[2]%>%round(1), "%)"))+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#222222",
                  size=3)+
#  geom_text_repel(aes(x=Yld[,1], y=Yld[,2]), 
#            label=rownames(Yld), size=3, color="blue")+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.1))

##scores biplot
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
#  geom_point(aes(x=sc[,1], y=sc[,2], alpha=0.5))+
  geom_text(aes(x=sc[,1], y=sc[,2]), label = rownames(sc))+
  labs(title="", 
       x="PC1", 
       y="PC2")+
#       x=paste0("PC1", " (",explvar(pls1)[1]%>%round(1), "%)"), 
#       y=paste0("PC2", " (",explvar(pls1)[2]%>%round(1), "%)"))+
  geom_text(aes(x=ld[,1]*10, y=ld[,2]*10), 
                  label=rownames(ld),
                  color="#222222",
                  size=3)+
#  geom_text_repel(aes(x=Yld[,1], y=Yld[,2]), 
#            label=rownames(Yld), size=3, color="blue")+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.1))

#plot R2X, R2Y, Q2Y
pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

#plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
#plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
#plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
#plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
#matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
#    geom_vline(xintercept = comp, linetype="dashed") + 
    geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  theme(panel.background = element_rect(color = "#000000"))+
  scale_y_continuous(limits = c(0,100))
#ggsave("TCGA_PLSR_var-expl.png", width = 5, height = 4)


```

```{r}
#GSVA + VIP enrichment

vip.scores1 <- vip.scores
x.mat1 <- x.mat
y.mat1 <- y.mat

keep.scores <- vip.scores1 %>% 
  filter(score>1.5) %>% 
  .[,1]
x.mat2 <- data.frame(x.mat1) %>% select(keep.scores) %>% data.matrix()

mydata <- data.frame(Y=I(y.mat1),X=I(x.mat1))
pls2 <- plsr(Y~X, scale=F, data=mydata, validation = "LOO", method="kernelpls") # generate PLSR model

sc <- scores(pls2)
ld <- loadings(pls2)
Ysc <-Yscores(pls2)
Yld <- Yloadings(pls2)

#VIP score calculations
vip.scores2 <- VIP(pls2, opt.comp = 3) %>% as.data.frame()
colnames(vip.scores2) <- "score"
vip.scores2 %<>% rownames_to_column("analyte")

#plot VIP scores
ggplot(data=arrange(vip.scores2, desc(score)), 
       aes(y=reorder(analyte, -score), x=score))+
  geom_bar(stat="identity")+
    theme_cowplot()+
  theme(axis.text.x=element_text(angle=90, vjust=.5, hjust=1),
        panel.background = element_rect(color = "#000000")
        )+
  scale_fill_gradient2( # Custom color gradient for bars
    midpoint=0,
    low="red",
    mid="#555555",
    high="blue",
    aesthetics="fill"
  )+
  geom_hline(yintercept = 1, linetype = "dashed")

#plot loadings 
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], alpha=0.5))+
  geom_text(aes(x=Yld[,1], y=Yld[,2]), 
            color="blue", 
            label = rownames(Yld),
            size = 3)+
  labs(title="", 
       x="PC1", 
       y="PC2")+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#222222",
                  size=3, max.iter = 100000, max.overlaps = 50)+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.1))

##scores biplot
ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
#  geom_point(aes(x=sc[,1], y=sc[,2], alpha=0.5))+
  geom_text(aes(x=sc[,1], y=sc[,2]), label = rownames(sc))+
  labs(title="", 
       x="PC1", 
       y="PC2")+
#       x=paste0("PC1", " (",explvar(pls1)[1]%>%round(1), "%)"), 
#       y=paste0("PC2", " (",explvar(pls1)[2]%>%round(1), "%)"))+
  geom_text(aes(x=ld[,1]*10, y=ld[,2]*10), 
                  label=rownames(ld),
                  color="#222222",
                  size=3)+
#  geom_text_repel(aes(x=Yld[,1], y=Yld[,2]), 
#            label=rownames(Yld), size=3, color="blue")+
  theme_cowplot() +
  scale_size(
    limits=c(0,1),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  ) +
  scale_x_continuous(expand=expansion(mult=.1))

#plot R2X, R2Y, Q2Y
pls2res <- rq_func(pls2)
R2X <- pls2res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls2res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls2res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

#plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
#plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
#plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
#plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
#matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
#    geom_vline(xintercept = comp, linetype="dashed") + 
    geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  theme(panel.background = element_rect(color = "#000000"))+
  scale_y_continuous(limits = c(0,100))
#ggsave("TCGA_PLSR_var-expl.png", width = 5, height = 4)


```

###neftel sets
```{r}
sets.gsva <- c(
   "NEFTEL_MES",
   "NEFTEL_PN"
               )

gsva.sub <- dplyr::select(gsva_res.df.t, one_of(sets.gsva)) %>% data.matrix()
x.mat <- data_for_gsva %>% t() %>% data.matrix()
colnames(x.mat) <- colnames(x.mat) %>% mapIds(org.Hs.eg.db, ., 'SYMBOL', 'ENTREZID')
#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(c(phillips_mes, phillips_pn, verhaak_m, verhaak_p))) %>% data.matrix()
#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(c(verhaak_m, verhaak_p))) %>% data.matrix()
#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(c(phillips_mes, phillips_pn))) %>% data.matrix()
x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(c(neftel_p, neftel_m))) %>% data.matrix()

#pls.dat1 <- data.frame(Y=I(gsva.sub), X=I(x.mat))
#pls1 <- plsr(Y~X, scale=F, data=pls.dat1, method="kernelpls")

#pls1res <- rq_func(pls1)
#Q2Y_cum <- pls1res$Q2Y.cum 
#comp <- which.max(Q2Y_cum)

###this is for interative PLSR using VIP scores to filter genes
#numgenes <- 50
#vip.scores <- VIP(pls1, comp) %>% data.frame()
#colnames(vip.scores)[1] <- "VIP.score"
#vip.scores <- rownames_to_column(vip.scores, var="gene")
#vip.scores <- vip.scores %>% filter(VIP.score>score.min)
#vip.scores <- vip.scores %>% top_n(numgenes, VIP.score)
#genes.VIP <- vip.scores %>% pull(gene)
#running PLSR again using only genes w/ high VIP scores

#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(genes.VIP)) %>% data.matrix()
pls.dat2 <- data.frame(Y=I(gsva.sub),X=I(x.mat))
pls1 <- plsr(Y~X, scale=T, data=pls.dat2, validation = "LOO", method="kernelpls")

pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

vip.scores <- VIP(pls1, comp) %>% data.frame()
sc <- scores(pls1)
ld <- loadings(pls1) %>% cbind(vip.scores$.)
colnames(ld)[ncol(ld)] <- "VIP"
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)
Yld <- data.frame(matrix(as.numeric(Yld), attributes(Yld)$dim, dimnames=attributes(Yld)$dimnames)) %>% rownames_to_column(var="gene")

ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], size = (ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)])), alpha=0.5)+
#  geom_point(aes(x=Yld[,2], y=Yld[,3]))+
  labs(x = "PC1", 
       y = "PC2"
       )+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#555555",
                  size=3, alpha = 1, max.overlaps = 10)+
  geom_text(data = Yld,
            aes(x=Yld[,2], y=Yld[,3]), 
            label=Yld[,1], 
            size=3, color="darkred")+
  theme_cowplot() +
  scale_size(
    limits=c(0,2),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  )

plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
    geom_vline(xintercept = comp, linetype="dashed") + geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  scale_x_continuous(limits = c(0,20))+
  scale_y_continuous(limits = c(0,100))
plot(RMSEP(pls1), legendpos="topleft")

ld %>% data.frame() %>%  rownames_to_column(var = "gene") %>% 
  top_n(50, VIP) %>% 
  arrange(VIP) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(gene = factor(gene, levels=gene)) %>%   # This trick update the factor levels
  ggplot(aes(x = VIP, y = gene, fill = Comp.1))+
  geom_bar(stat = "identity") + 
  scale_fill_gradient(low = "blue", high = "red")+
  theme_cowplot()
#ggsave("neftel_vip.png", width = 6, height = 10)
```
###verhaak
```{r}
sets.gsva <- c(
  "VERHAAK_GLIOBLASTOMA_MESENCHYMAL", 
  "VERHAAK_GLIOBLASTOMA_PRONEURAL"
               )

gsva.sub <- dplyr::select(gsva_res.df.t, one_of(sets.gsva)) %>% data.matrix()
x.mat <- data_for_gsva %>% t() %>% data.matrix()
colnames(x.mat) <- colnames(x.mat) %>% mapIds(org.Hs.eg.db, ., 'SYMBOL', 'ENTREZID')
#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(c(phillips_mes, phillips_pn, verhaak_m, verhaak_p))) %>% data.matrix()
x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(c(verhaak_m, verhaak_p))) %>% data.matrix()



#pls.dat1 <- data.frame(Y=I(gsva.sub), X=I(x.mat))
#pls1 <- plsr(Y~X, scale=F, data=pls.dat1, method="kernelpls")

#pls1res <- rq_func(pls1)
#Q2Y_cum <- pls1res$Q2Y.cum 
#comp <- which.max(Q2Y_cum)
###this is for interative PLSR using VIP scores to filter genes
#numgenes <- 50
#vip.scores <- VIP(pls1, comp) %>% data.frame()
#colnames(vip.scores)[1] <- "VIP.score"
#vip.scores <- rownames_to_column(vip.scores, var="gene")
#vip.scores <- vip.scores %>% filter(VIP.score>score.min)
#vip.scores <- vip.scores %>% top_n(numgenes, VIP.score)
#genes.VIP <- vip.scores %>% pull(gene)
#running PLSR again using only genes w/ high VIP scores

#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(genes.VIP)) %>% data.matrix()
pls.dat2 <- data.frame(Y=I(gsva.sub),X=I(x.mat))
pls1 <- plsr(Y~X, scale=T, data=pls.dat2, validation = "LOO", method="kernelpls")

pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

vip.scores <- VIP(pls1, comp) %>% data.frame()
sc <- scores(pls1)
ld <- loadings(pls1) %>% cbind(vip.scores$.)
colnames(ld)[ncol(ld)] <- "VIP"
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)
Yld <- data.frame(matrix(as.numeric(Yld), attributes(Yld)$dim, dimnames=attributes(Yld)$dimnames)) %>% rownames_to_column(var="gene")

ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], size = (ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)])), alpha=0.5)+
#  geom_point(aes(x=Yld[,2], y=Yld[,3]))+
  labs(x = "PC1", 
       y = "PC2"
       )+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#555555",
                  size=3, alpha = 1, max.overlaps = 10)+
  geom_text(data = Yld,
            aes(x=Yld[,2], y=Yld[,3]), 
            label=Yld[,1], 
            size=3, color="darkred")+
  theme_cowplot() +
  scale_size(
    limits=c(0,2),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  )

plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
    geom_vline(xintercept = comp, linetype="dashed") + geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  scale_x_continuous(limits = c(0,20))+
  scale_y_continuous(limits = c(0,100))
plot(RMSEP(pls1), legendpos="topleft")

ld %>% data.frame() %>%  rownames_to_column(var = "gene") %>% 
  top_n(50, VIP) %>% 
  arrange(VIP) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(gene = factor(gene, levels=gene)) %>%   # This trick update the factor levels
  ggplot(aes(x = VIP, y = gene, fill = Comp.1))+
  geom_bar(stat = "identity") + 
  scale_fill_gradient(low = "blue", high = "red")+
  theme_cowplot()
#ggsave("verhaak_vip.png", width = 6, height = 10)
```
###phillips
```{r}
sets.gsva <- c(
   "PHILLIPS_MES", 
   "PHILLIPS_PN"
               )

gsva.sub <- dplyr::select(gsva_res.df.t, one_of(sets.gsva)) %>% data.matrix()
x.mat <- data_for_gsva %>% t() %>% data.matrix()
colnames(x.mat) <- colnames(x.mat) %>% mapIds(org.Hs.eg.db, ., 'SYMBOL', 'ENTREZID')
x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(c(phillips_mes, phillips_pn))) %>% data.matrix()


#pls.dat1 <- data.frame(Y=I(gsva.sub), X=I(x.mat))
#pls1 <- plsr(Y~X, scale=F, data=pls.dat1, method="kernelpls")

#pls1res <- rq_func(pls1)
#Q2Y_cum <- pls1res$Q2Y.cum 
#comp <- which.max(Q2Y_cum)

###this is for interative PLSR using VIP scores to filter genes
#numgenes <- 50
#vip.scores <- VIP(pls1, comp) %>% data.frame()
#colnames(vip.scores)[1] <- "VIP.score"
#vip.scores <- rownames_to_column(vip.scores, var="gene")
#vip.scores <- vip.scores %>% filter(VIP.score>score.min)
#vip.scores <- vip.scores %>% top_n(numgenes, VIP.score)
#genes.VIP <- vip.scores %>% pull(gene)
#running PLSR again using only genes w/ high VIP scores

#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(genes.VIP)) %>% data.matrix()
pls.dat2 <- data.frame(Y=I(gsva.sub),X=I(x.mat))
pls1 <- plsr(Y~X, scale=T, data=pls.dat2, validation = "LOO", method="kernelpls")

pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

vip.scores <- VIP(pls1, comp) %>% data.frame()
sc <- scores(pls1)
ld <- loadings(pls1) %>% cbind(vip.scores$.)
colnames(ld)[ncol(ld)] <- "VIP"
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)
Yld <- data.frame(matrix(as.numeric(Yld), attributes(Yld)$dim, dimnames=attributes(Yld)$dimnames)) %>% rownames_to_column(var="gene")

ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], size = (ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)])), alpha=0.5)+
#  geom_point(aes(x=Yld[,2], y=Yld[,3]))+
  labs(x = "PC1", 
       y = "PC2"
       )+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#555555",
                  size=3, alpha = 1, max.overlaps = 10)+
  geom_text(data = Yld,
            aes(x=Yld[,2], y=Yld[,3]), 
            label=Yld[,1], 
            size=3, color="darkred")+
  theme_cowplot() +
  scale_size(
    limits=c(0,2),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  )

plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
    geom_vline(xintercept = comp, linetype="dashed") + geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  scale_x_continuous(limits = c(0,20))+
  scale_y_continuous(limits = c(0,100))
plot(RMSEP(pls1), legendpos="topleft")

ld %>% data.frame() %>%  rownames_to_column(var = "gene") %>% 
  top_n(50, VIP) %>% 
  arrange(VIP) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(gene = factor(gene, levels=gene)) %>%   # This trick update the factor levels
  ggplot(aes(x = VIP, y = gene, fill = Comp.1))+
  geom_bar(stat = "identity") + 
  scale_fill_gradient(low = "blue", high = "red")+
  theme_cowplot()
#ggsave("phillips_vip.png", width = 6, height = 10)
```
###combined mes/pn sets
```{r}
sets.gsva <- c(
  "VERHAAK_GLIOBLASTOMA_MESENCHYMAL", "VERHAAK_GLIOBLASTOMA_PRONEURAL",
  "PHILLIPS_MES", "PHILLIPS_PN",
  "NEFTEL_MES", "NEFTEL_PN"
               )

#genes.overlap <- intersect(verhaak_m, phillips_mes) %>% c(intersect(verhaak_p, phillips_pn))
genes.overlap <- intersect(verhaak_m, neftel_m) %>% intersect(phillips_mes) %>% c(intersect(verhaak_p, neftel_p) %>% intersect(phillips_pn))

gsva.sub <- dplyr::select(gsva_res.df.t, one_of(sets.gsva)) %>% data.matrix()
x.mat <- data_for_gsva %>% t() %>% data.matrix()
colnames(x.mat) <- colnames(x.mat) %>% mapIds(org.Hs.eg.db, ., 'SYMBOL', 'ENTREZID')
#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(c(verhaak_m, verhaak_p, phillips_mes, phillips_pn, neftel_m, neftel_p))) %>% data.matrix()
x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(genes.overlap)) %>% data.matrix()

##running PLSR the first time
#pls.dat1 <- data.frame(Y=I(gsva.sub), X=I(x.mat))
#pls1 <- plsr(Y~X, scale=F, data=pls.dat1, method="kernelpls")
#pls1res <- rq_func(pls1)
#Q2Y_cum <- pls1res$Q2Y.cum 
#comp <- which.max(Q2Y_cum)

###this is for iterative PLSR using VIP scores to filter genes
#numgenes <- 100
#vip.scores <- VIP(pls1, comp) %>% data.frame()
#colnames(vip.scores)[1] <- "VIP.score"
#vip.scores <- rownames_to_column(vip.scores, var="gene")

#vip.scores <- vip.scores %>% top_n(numgenes, VIP.score)
#genes.VIP <- vip.scores %>% pull(gene)
##running PLSR again using only genes w/ high VIP scores

#x.mat <- x.mat %>% data.frame() %>% dplyr::select(any_of(genes.VIP)) %>% data.matrix()
pls.dat2 <- data.frame(Y=I(gsva.sub),X=I(x.mat))
pls1 <- plsr(Y~X, scale=F, data=pls.dat2, validation = "LOO", method="kernelpls")

pls1res <- rq_func(pls1)
R2X <- pls1res$R2X; R2X_cum <- pls1res$R2X.cum
R2Y <- pls1res$R2Y; R2Y_cum <- pls1res$R2Y.cum
Q2Y_cum <- pls1res$Q2Y.cum 
rq2_cum <- cbind(R2X_cum,R2Y_cum,Q2Y_cum); colnames(rq2_cum) <- c("R2X","R2Y","Q2Y")
comp <- which.max(Q2Y_cum) # number of components for best predictive power (based on Q2Y)
R2X_cum[comp]
R2Y_cum[comp]
Q2Y_cum[comp]

vip.scores <- VIP(pls1, comp) %>% data.frame()
sc <- scores(pls1)
ld <- loadings(pls1) %>% cbind(vip.scores$.)
colnames(ld)[ncol(ld)] <- "VIP"
Ysc <-Yscores(pls1)
Yld <- Yloadings(pls1)
Yld <- data.frame(matrix(as.numeric(Yld), attributes(Yld)$dim, dimnames=attributes(Yld)$dimnames)) %>% rownames_to_column(var="gene")

ggplot()+
  geom_hline(yintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_vline(xintercept = 0, color = "#333333", linetype = "dashed", alpha = 0.7) +
  geom_point(aes(x=ld[,1], y=ld[,2], size = (ld[,ncol(ld)]-min(ld[,ncol(ld)]))/max(ld[,ncol(ld)])), alpha=0.5)+
#  geom_point(aes(x=Yld[,2], y=Yld[,3]))+
  labs(x = "PC1", 
       y = "PC2"
       )+
  geom_text_repel(aes(x=ld[,1], y=ld[,2]), 
                  label=rownames(ld),
                  color="#555555",
                  size=3, alpha = 1, max.overlaps = 10)+
  geom_text(data = Yld,
            aes(x=Yld[,2], y=Yld[,3]), 
            label=Yld[,1], 
            size=3, color="darkred")+
  theme_cowplot() +
  scale_size(
    limits=c(0,2),
  )+
  theme(
    panel.background = element_rect(color = "#000000"),
    legend.position = "none"
  )

plot(pls1, ncomp=comp, asp=1, line=T, xlab="Actual", ylab="Predicted", main="Parity plots")
plot(pls1res$R2X.cum, ylab="R2X", xlab="# PCs")
plot(pls1res$R2Y.cum, ylab="R2Y", xlab="# PCs")
plot(pls1res$Q2Y.cum, ylab="Q2Y", xlab="# PCs")
matplot(rq2_cum, pch=c("X","Y","Q"), cex=1, col=c("red","blue","green"), ylab="% var. expl.", xlab="# PCs")
rq2_cum %>% data.frame(ncomp=seq(1,nrow(.))) %>% pivot_longer(-ncomp) %>% mutate(name=factor(name,levels=colnames(rq2_cum))) %>% 
  ggplot(aes(ncomp, value, fill=name, color=name, shape=name)) +
    geom_vline(xintercept = comp, linetype="dashed") + geom_point(size=3) + #geom_line(size=1) +
    theme_cowplot(18) + scale_x_continuous(n.breaks=10) +
    labs(x="# PCs", y="% var. expl.", color="Metric", shape="Metric", fill="Metric")+
  scale_x_continuous(limits = c(0,20))+
  scale_y_continuous(limits = c(0,100))
plot(RMSEP(pls1), legendpos="topleft")

ld %>% data.frame() %>%  rownames_to_column(var = "gene") %>% 
  top_n(20, VIP) %>% 
  arrange(VIP) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(gene = factor(gene, levels=gene)) %>%   # This trick update the factor levels
  ggplot(aes(x = VIP, y = gene, fill = Comp.1))+
  geom_bar(stat = "identity") + 
  scale_fill_gradient(low = "red", high = "blue")+
  theme_cowplot()
#ggsave("all_gene_sets_vip.png", width = 6, height = 10)
```

```{r}
gsva_all <- tn.gsva.df

```

