---
title: "Analysis of CPTAC GBM data"
author: Paul J. Myers
output:
  html_document:
    fig_width: 9
    fig_height: 6
    theme: spacelab
    toc: yes
  pdf_document: 
    toc: yes
    fig_width: 12
    fig_height: 9
Date of creation: "9/27/2021"
R version: "4.1.0"
---

# Housekeeping Code
We start by loading some of the desired libraries for this script. Additional packages will be loaded later for specific analyses to make their use clear.

```{r Load packages, message=F, warning=F}
### Load packages:
library(tidyverse)
library(magrittr)
library(BiocManager)
library(paletteer)
library(scales)
library(cowplot)
library(ggrepel)
library(ggbeeswarm)
library(ggstatsplot)
library(pcaMethods)
library(ComplexHeatmap)
library(tidyHeatmap)
library(parallel)
```

```{r Clear workspace}
### Clear R workspace:
rm(list=ls())
```

```{r Check working directory}
### Check that current working directory is correct and the desired one:
cwd <- getwd()
cwd
```

```{r Define ni operator}
## Define 'not in' (ni) operator
`%ni%` <- Negate(`%in%`)
```

This function can be used to generate a ggplot with a blank grid and black box surrounding the plot area. The only input to the function specifies whether y tick marks and labels are desired or not (depending on the type of plot you are generating).
```{r Custom ggplot theme for plot background}
theme_cust <- function(ticks=T, axis="y", box=T, lnsz=1,
                       xtickangle=0, hjust=NULL, vjust=NULL, fill=NA,
                       legend.position="right"){
  theme_out <- theme(legend.position=legend.position)
  if (ticks == FALSE & axis == "y" ){
    theme_out <- theme_out + theme(
      axis.text.y = element_blank(), axis.ticks.y = element_blank(), # Make y axis ticks blank
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = "black", fill = fill),
      )
  } else if (ticks == FALSE & axis != "y"){
    theme_out <- theme_out + theme(
      axis.text.x = element_blank(), axis.ticks.x = element_blank(), # Make x axis ticks blank
      axis.line.y = element_blank(),
      axis.line.x = element_blank(),
      panel.background = element_rect(size = lnsz, color = "black", fill = fill),
      )
  } else {
     theme_out <- theme_out + theme(
       axis.line.y = element_blank(),
       axis.line.x = element_blank(),
       panel.background = element_rect(size = lnsz, color = "black", fill = fill),
       axis.text.x = element_text(angle=xtickangle, hjust = hjust, vjust=vjust),
       ) 
  }
  
  if (xtickangle==90){
    theme_out <- theme_out + theme(axis.text.x = element_text(angle=xtickangle, hjust = 1, vjust=0.4))
  }
  
  if (box != TRUE) {
    theme_out <- theme_out + theme(panel.background = element_blank(),
                                   axis.line.y = element_line(size = lnsz, color = "black"),
                                   axis.line.x = element_line(size = lnsz, color = "black")
                                   )
  } else {
    theme_out <- theme_out
  }
  return(theme_out)
}

# Julia's Plots.jl default color order:
julia_cols <- c("#009AFA","#E26E47","#3FA54E","#c270D2","#AD8F18","#01ABAE","#ED5F92","#C68324","#01A98C","#8F961E","#01A9CD","#9B7EE8","#618CF7","#F16072","#DC65B7","#6D9E33")
```

The function below is designed to manually calculate PTM-SEA/PTM-SVA enrichment scores using a similar approach to that described by [Krug et al. (2019)](http://doi.org/10.1074/mcp.TIR118.000943). The function accepts a data frame containing enrichment scores for phosphosite-specific signatures, the names of the signatures (must use the name `sig`), and the directionality of the signatures (must use the name `dir`). Signatures can be downloaded from the [PTMSigDB](http://prot-shiny-vm.broadinstitute.org:3838/ptmsigdb-app/). The output of the function can be a matrix (default) or data frame containing the PTM-SEA enrichment scores and the names of the signatures used.
```{r Custom PTM-SEA function}
ptmsea_func <- function(es_df, output="matrix"){
  es_df <- es_df %>% 
    mutate(dir = factor(dir, levels=c("u","d"))) %>% 
    group_by(sig) %>% 
    arrange(dir)
  numeric_cols <- unlist(lapply(es_df, is.numeric)) # indices of numeric columns
  scores <- es_df[,numeric_cols] # enrichment scores from ssGSEA/GSVA
  num_cols <- ncol(scores) # number of spls
  sigs <- unique(es_df$sig) # names of signatures used
  num_rows <- sigs %>% length() # number of unique signatures

  ptm_es <- matrix(0, nrow=num_rows, ncol=num_cols, dimnames=list(sigs)) # empty PTM set enrichment score matrix
  colnames(ptm_es) <- colnames(es_df[,numeric_cols]) # set sample (row) names

  for (i in 1:num_rows){
    sig_i = sigs[i] # signature of current iteration
    es_i = es_df %>% subset(sig %in% sig_i)

    ## Up direction scores:
    es_i.u = es_i %>% subset(dir %in% "u")
    if(nrow(es_i.u) > 0){
      up = es_i.u[,numeric_cols] %>% data.matrix()
    } else {
      up = 0
    }

    ## Down direction scores:
    es_i.d = es_i %>% subset(dir %in% "d")
    if(nrow(es_i.d) > 0){
      down = es_i.d[,numeric_cols] %>% data.matrix()
    } else {
      down = 0
    }
    score_i = up-down

    ## Store the results:
    ptm_es[i,] = score_i
    rownames(ptm_es)[i] <- sig_i # set geneset name
  }
  
  if(output == "df"){
    ## Convert to data frame and return PTM-SEA scores:
    ptm_es_df <- ptm_es %>% data.frame(sig=rownames(.))
    return(ptm_es_df)
  } else {
    return(ptm_es)
  }
}
```



# Load CPTAC data
All CPTAC GBM data were downloaded from the [CPTAC GBM Data Portal](https://cptac-data-portal.georgetown.edu/study-summary/S057) or from the supplemental data for the [associated publication by Wang et al.](https://doi.org/10.1016/j.ccell.2021.01.006).Clinical data for the tumors matching the proteomics data were also obtained from the CPTAC website; these data are loaded first.

```{r Sample phenotype and clinical data, message=F, warning=F}
## Mass spec sample information:
# fn_splinfo <- "../CPTAC GBM data/S048_CPTAC_GBM_Discovery_Cohort_TMT11_CaseID_SampleID_AliquotID_Map_Dec2019_r1.xlsx"
fn_splinfo <- "./Wang et al/Data/GBM_Metadata_CancerCell2021/S057_S048_CPTAC_GBM_Discovery_Cohort_TMT11_CaseID_SampleID_AliquotID_Map_Feb2021_r2.xlsx"
splinfo <- readxl::read_excel(fn_splinfo, skip=7) %>% 
  dplyr::select(contains(all_of(c("ID","Tumor","Withdrawn")))) %>%
  dplyr::rename(case_id = `Case ID (Participant ID)`) %>%
  mutate(case_id=make.names(case_id)) 
colnames(splinfo) <- colnames(splinfo) %>% make.names()

## Clinical/phenotype data:
# fn_clin <- "../CPTAC GBM data/S057_S048_CPTAC_GBM_Discovery_Cohort_Clinical_Data_Feb2021_r2.xlsx"
fn_clin <- "./Wang et al/Data/GBM_Metadata_CancerCell2021/S057_S048_CPTAC_GBM_Discovery_Cohort_Clinical_Data_Feb2021_r2.xlsx"
clin_data <- readxl::read_excel(fn_clin, sheet="Patient_Clinical_Attributes")
colnames(clin_data) <- colnames(clin_data) %>% make.names()
clin_data <- clin_data %>% mutate(case_id = make.names(case_id))

## Clinical, phenotype, and gene mutation data from Wang et al. supplement:
# fn_mut <- "../Wang et al/Table_S1.xlsx"
fn_mut <- "./Wang et al/Table_S1.xlsx"
mut_data <- readxl::read_excel(fn_mut, sheet="additional_annotations") %>% 
  dplyr::select(case, contains("_alter_status")) %>% 
  dplyr::rename(case_id=case) %>% 
  mutate(EGFR_SV = str_detect(EGFR_alter_status, "SV"),
         EGFR_AMP = str_detect(EGFR_alter_status, "Amp"),
         EGFR_WT = str_detect(EGFR_alter_status, "NA"),
         case_id = make.names(case_id),
         Aliquot.ID = match(case_id, splinfo$case_id) %>% splinfo$Aliquot.ID[.],
         ) %>% 
  na.omit()

## Manta structural variant data:
fn_sv <- "./Wang et al/Table_S2.xlsx"
sv_data <- readxl::read_excel(fn_sv, sheet="structural_variant_manta") %>% 
  mutate(case=make.names(case))

splinfo <- splinfo %>% 
  filter(case_id %in% clin_data$case_id) %>% 
  na.omit()


```

```{r Identifying EGFR vIII SV tumors}
## Attempt to pull out genome region where EGFRvIII mutation should show up:
sv_egfr <- sv_data %>% 
  filter(case %in% splinfo$case_id) %>% 
  filter(str_detect(CHROM_A, "chr7") | str_detect(CHROM_B, "chr7")) %>% 
  filter(TYPE=="DEL") %>% 
  filter((START_A>55019000 & END_A<55325000) | (START_B>55019000 & END_B<55325000)) %>% 
  mutate(
    END=str_extract(INFO, "END=(.*?);") %>% gsub("END=|;","",.) %>% as.numeric(),
    SVLEN=str_extract(INFO, "SVLEN=(.*?);") %>% gsub("SVLEN=|;","",.) %>% as.numeric(),
    ) %>% 
  filter(END<55325000 & SVLEN<=-801) # there are 801 total bps in EGFR exons 2-7
write.csv(sv_egfr, file="EGFR_Manta_SV_calls.csv")

## Annotate mutation data with EGFRvIII status:
vIII_spls <- sv_egfr$case %>% unique()
mut_data <- mut_data %>% 
  mutate(EGFRvIII = F,
         EGFRvIII = replace(EGFRvIII, case_id %in% vIII_spls, T)
  )

## Sample info without non-VIII EGFR SVs:
mut_vIII <- mut_data %>% 
  filter(EGFRvIII | EGFR_WT)
```

Next we load the global proteomics data.
```{r Load the total proteomics data, message=F}
## Load data:
# fn <- "../CPTAC GBM data/CPTAC3_Glioblastoma_Multiforme_Proteome.tmt11.tsv" # raw proteome data
fn <- "./Wang et al/Data/Proteome/proteome_mssm_per_gene_clean.v4.0.tsv" # cleaned proteome data
raw.data <- read_tsv(fn)
colnames(raw.data) <- colnames(raw.data) %>% make.names()

## Clean up data and get just the confirmed PDAC tumor samples:
data1 <- raw.data %>% # make a copy for use downstream
  dplyr::select(gene, 
                # contains(all_of(clin_data$case_id %>% match(splinfo$case_id) %>% na.omit() %>% splinfo$Aliquot.ID[.])),
                contains(all_of(splinfo$case_id)),
                # -contains(all_of(c("Unshared","QC1","Withdrawn","Organism","Authority"))), # remove unneeded columns
                )
colnames(data1) <- make.names(colnames(data1))
# colnames(data1) <- sub(".Log.Ratio","",colnames(data1))

incomplete_prots <- (!complete.cases(data1)) %>% data1$gene[.] %>% data.frame(incomplete_proteins=.)

## Dataframe of patient/sample metadata:
spls <- data.frame(case_id = colnames(data1)[-1],
                   spl = paste("SPL", 1:length(colnames(data1)[-1]), sep="")) %>%
  # mutate(case_id = match(aliquot_id, splinfo$Aliquot.ID) %>% splinfo$case_id[.]) %>% 
  inner_join(splinfo, by="case_id", suffix=c("",".x")) %>% 
  inner_join(clin_data, by="case_id", suffix=c("",".x")) %>% 
  inner_join(mut_data, by="case_id", suffix=c("",".x"))
```


Now loading and cleaning up the phosphoproteomics data.
```{r Load the phosphoproteomics data, message=F}
## Load data:
# fn.phos <- "../CPTAC GBM data/CPTAC3_Glioblastoma_Multiforme_Phosphoproteome.phosphosite.tmt11.tsv"
fn.phos <- "./Wang et al/Data/Proteome/phosphoproteome_mssm_per_gene_clean.v4.0.tsv" 
raw.phos <- read_tsv(fn.phos) %>%   
  mutate(site = sub("-", "_", site),
         site = sub("s|s$", "", site),
         site = sub("t|t$", "", site),
         site = sub("y|y$", "", site),
         site = toupper(site)
         )
colnames(raw.phos) <- colnames(raw.phos) %>% make.names()

## Clean up data:
phos1 <- raw.phos %>% # make a copy for use downstream
  dplyr::rename(Phosphosite=site) %>% 
  dplyr::select(gene, Phosphosite,
                # contains(all_of(clin_data$case_id %>% match(splinfo$case_id) %>% na.omit() %>% splinfo$Aliquot.ID[.])),
                contains(all_of(splinfo$case_id)),
                # -contains(all_of(c("Unshared","QC1","Withdrawn","Organism","Authority"))), # remove unneeded columns
                ) #%>% 
  # group_by(Phosphosite) %>%
  # mutate(Phosphosite = sub(".*:", paste(gene,"_",sep=""), Phosphosite)) %>%
  # summarise(across(where(is.numeric), ~mean(.x, na.rm = TRUE)))
phos1[is.na(phos1)] <- NA
colnames(phos1) <- make.names(colnames(phos1))
colnames(phos1) <- sub(".Log.Ratio","",colnames(phos1))

numna <- phos1[,-1] %>% is.na() %>% sum()
numel <- dim(phos1[,-1]) %>% prod()
fracna <- numna/numel; fracna
```




# Exploratory analysis of EGFR, SHP2, and PTEN activity/expression in EGFRvIII versus EGFR WT tumors
```{r EGFR mutation status analysis, message=F, warning=F}
genes_to_try <- c("EGFR","PTPN11","PTEN","LAMP1")

## Get data on EGFR, SHP2 (PTPN11), and PTEN total protein expression:
data2 <- data1 %>% 
  filter(gene %in% genes_to_try) %>% 
  pivot_longer(-gene, names_to = "case_id") %>% 
  inner_join(spls, by="case_id") %>% 
  dplyr::select(gene, value, contains(all_of(c("spl","EGFR","PTPN11","PTEN"))))

## Visualize expression across different EGFRvIII statuses:
data2 %>% 
  ggplot(aes(x=EGFRvIII, y=value, shape=EGFRvIII, color=EGFRvIII, fill=EGFRvIII)) +
    # geom_histogram(aes(y=..density..)) +
    # geom_density(alpha=1) +
    geom_quasirandom(size=3, alpha=0.7) +
    geom_violin(fill=NA) +
    scale_color_manual(values=julia_cols) +
    scale_fill_manual(values=julia_cols) +
    facet_wrap(~gene, scales="free") + 
    theme_cowplot() + 
    labs(title="CPTAC GBM", y="Expression") +
    guides(shape="none",color="none",fill="none")

grouped_ggbetweenstats(data2, x=EGFRvIII, y=value, grouping.var=gene,
                         plot.type="violin", type="np",
                         point.args = list(alpha=0.7, shape=16, size=2,
                                           position = position_quasirandom(width = 0.15, groupOnX=T)),
                         centrality.plotting=F, bf.message=F,
                         ggtheme = theme_cowplot(8), ggplot.component=list(scale_color_manual(values=julia_cols)),  
                         )


## Get data on EGFR, SHP2 (PTPN11), and PTEN phosphorylation data:
phos2 <- phos1 %>% 
  dplyr::select(-gene) %>% 
  filter(str_detect(Phosphosite, paste(genes_to_try, collapse="|"))) %>% 
  pivot_longer(-Phosphosite, names_to = "case_id") %>% 
  inner_join(spls, by="case_id") %>% 
  dplyr::select(Phosphosite, value, contains(all_of(c("spl","EGFR","PTPN11","PTEN"))))

## Visualize expression across different EGFRvIII statuses:
# phos2 %>% 
#   ggplot(aes(x=EGFRvIII, y=value, shape=EGFRvIII, color=EGFRvIII, fill=EGFRvIII)) +
#     # geom_histogram(aes(y=..density..)) +
#     # geom_density(alpha=1) +
#     geom_quasirandom(size=3, alpha=0.7) +
#     geom_violin(fill=NA) +
#     scale_color_manual(values=julia_cols) +
#     scale_fill_manual(values=julia_cols) +
#     facet_wrap(~Phosphosite, scales="free") + 
#     theme_cowplot() + 
#     labs(title="CPTAC GBM", y="Expression") +
#     guides(shape="none",color="none",fill="none")

pegfr_comps <- phos2 %>% 
  # filter(str_detect(Phosphosite, "EGFR_[A-Z][0-9]{3,4}$")) %>%
  filter(str_detect(Phosphosite, "EGFR_Y")) %>%
  grouped_ggbetweenstats(x=EGFRvIII, y=value, grouping.var=Phosphosite,
                         plot.type="violin", type="np",
                         point.args = list(alpha=0.7, shape=16, size=2,
                                           position = position_quasirandom(width = 0.15, groupOnX=T)),
                         centrality.plotting=F, bf.message=F,
                         ggplot.component=list(geom_hline(yintercept=0, linetype=8),
                                               scale_color_manual(values=julia_cols)),
                         ggtheme = theme_cowplot(8),
                         )
pegfr_comps
#ggsave("EGFR-phosphosite-comparisons_EGFR-WT-vs-vIII.png", pegfr_comps, width=20, height=10)

phos2 %>% 
  filter(str_detect(Phosphosite, "PTEN")) %>%
  grouped_ggbetweenstats(x=EGFRvIII, y=value, grouping.var=Phosphosite,
                         plot.type="violin", type="np",
                         point.args = list(alpha=0.7, shape=16, size=2,
                                           position = position_quasirandom(width = 0.15, groupOnX=T)),
                         centrality.plotting=F, bf.message=F,
                         ggplot.component=list(geom_hline(yintercept=0, linetype=8),
                                               scale_color_manual(values=julia_cols)),
                         ggtheme = theme_cowplot(8),
                         )

phos2 %>% 
  filter(str_detect(Phosphosite, "PTPN11")) %>%
  grouped_ggbetweenstats(x=EGFRvIII, y=value, grouping.var=Phosphosite,
                         plot.type="violin", type="np",
                         point.args = list(alpha=0.7, shape=16, size=2,
                                           position = position_quasirandom(width = 0.15, groupOnX=T)),
                         centrality.plotting=F, bf.message=F,
                         ggplot.component=list(geom_hline(yintercept=0, linetype=8),
                                               scale_color_manual(values=julia_cols)),
                         ggtheme = theme_cowplot(8),                       
                         )
```
```{r pEGFR heatmaps and correlations}
breaks1 <- seq(-5,5,length.out=100)
colors1 <- paletteer_d("RColorBrewer::RdBu", 100, direction=-1, type="continuous")
phos2 %>% 
  filter(str_detect(Phosphosite,"EGFR_Y[0-9]{3,4}$")) %>% 
  mutate(EGFRvIII = replace(EGFRvIII, EGFRvIII, "EGFRvIII"),
         EGFRvIII = replace(EGFRvIII, EGFRvIII=="FALSE", "Non-vIII"),
         ) %>%
  dplyr::select(EGFRvIII, Phosphosite, spl, value) %>%
  group_by(EGFRvIII) %>%
  arrange(value) %>% mutate(spl=factor(spl,levels=unique(spl))) %>% 
  unique() %>% 
  # tidyHeatmap::heatmap(.row=Phosphosite, .column=spl, .value=value,
  #         show_column_dend=F, show_row_dend=F, show_column_names=F, .scale="none",
  #         cluster_columns=T, clustering_distance_columns="euclidean", clustering_method_columns="ward.D2",
  #         cluster_rows=F, clustering_distance_rows="euclidean", clustering_method_rows="ward.D2",
  #         palette_value=circlize::colorRamp2(breaks1, colors1),
  #         column_title="CPTAC GBM: EGFR phospho-tyrosines", 
  #         column_title_gp=gpar(fontface="bold"),
  #         row_title=NULL,
  #         column_names_gp=gpar(fontsize=12),
  #         )
  ggplot(aes(spl, Phosphosite, fill=value)) + 
    geom_tile() +
    facet_wrap(~EGFRvIII, scales="free_x") +
    scale_fill_distiller(palette="RdBu") +
    theme_minimal_grid() + theme(axis.text.x=element_blank()) +
    labs(title="CPTAC GBM: EGFR phospho-tyrosines")
    
```


# GSVA: Global proteomics data
Here we load the necessary packages and gene sets from/related to the Molecular Signatures Database (MSigDB) for analyzing pathway and gene set enrichment in downstream analyses. The primary package we'll use for this is Gene Set Variation Analysis (`GSVA`), and we will load gene sets from the MSigDB programmatically using the `msigdbr` package.
```{r Load packages and MSigDB collections, message=F, warning=F}
## Load packages:
library(org.Hs.eg.db)
library(msigdbr)
library(GSVA)

### Define MSigDB gene set collection(s) to use --> retrieve with 'msigdbr' package:
species = "Homo sapiens"

## Retrieve Hallmark and canonical pathways collections in the database:
hall = msigdbr(species = species, category = "H")
cp = msigdbr(species = species, category = "C2", subcategory = "CP")
cp.b = msigdbr(species = species, category = "C2", subcategory = "CP:BIOCARTA")
cp.r = msigdbr(species = species, category = "C2", subcategory = "CP:REACTOME")
cp.p = msigdbr(species = species, category = "C2", subcategory = "CP:PID")
cp.k = msigdbr(species = species, category = "C2", subcategory = "CP:KEGG")
cp.w = msigdbr(species = species, category = "C2", subcategory = "CP:WIKIPATHWAYS")
gene_sets1 <- rbind(hall, cp, cp.b, cp.r, cp.p, cp.k, cp.w) %>% split(x = .$gene_symbol, f = .$gs_name)

## Go collections:
go.bp <- msigdbr(species = species, category = "C5", subcategory = "GO:BP")
go.cc <- msigdbr(species = species, category = "C5", subcategory = "GO:CC")
go.mf <- msigdbr(species = species, category = "C5", subcategory = "GO:MF")
gene_sets2 <- rbind(go.bp, go.cc, go.mf) %>% split(x = .$gene_symbol, f = .$gs_name)

## Transcription factor target collections:
go.tft <- msigdbr(species = species, category = "C3", subcategory = "TFT:GTRD")
go.tft_leg <- msigdbr(species = species, category = "C3", subcategory = "TFT:TFT_Legacy")
gene_sets3 <- rbind(go.tft, go.tft_leg) %>% split(x = .$gene_symbol, f = .$gs_name)

gene_sets_all <- append(gene_sets1, gene_sets2) %>% append(gene_sets3)
```

And now calculating GSVA scores for the desired gene sets.
```{r GSVA calculations}
#### Expression data prep ####
### Format data: Samples as columns, genes as rows
data_for_gsva <- data1[,-1] %>% data.matrix()
rownames(data_for_gsva) <- data1$gene

# ## Convert gene names to Entrez IDs:
# entrez_ids <- rownames(data_for_gsva) %>% mapIds(org.Hs.eg.db, ., 'ENTREZID', 'SYMBOL')
# rownames(data_for_gsva) <- entrez_ids
  
#### Calculate GSVA scores ####
fn_gsva <- "CPTAC-GBM_gsva_results.Rdata" # file name for storing the GSVA results
force_calculate_gsva <- F
if(!file.exists(fn_gsva) | force_calculate_gsva){
  gsva_res <- gsva(
    data_for_gsva, 
    gene_sets_all,
    method = "gsva", 
    kcdf = "Gaussian",
    min.sz = 5, # Minimum number of genes required to include a gene set
    parallel.sz = detectCores()#-1
    )
  ### Save GSVA results:
  save(list = c("gsva_res"), file = fn_gsva)
  
} else if (file.exists(fn_gsva) & !exists("gsva_res")){ 
  ## Load GSVA results, if they haven't been loaded already:
  load(fn_gsva) 
}
## GSVA results in dataframe format:
gsva_res.df <- gsva_res %>% data.frame(gene.set=rownames(.), .) # Convert to data frame
rownames(gsva_res.df) <- gsva_res.df$gene.set
gsva_res.df.t <- gsva_res %>%
  t() %>% 
  data.frame(Aliquot.ID=rownames(.), .)
```

## DEnA
Want to perform a first pass at comparing DNA damage-related protein expression based on whether tumors express WT EGFR or EGFRvIII. To do this, we'll perform differential enrichment analysis using `limma`.
```{r DEnA calculation}
library(limma)
## Get DNA damage-related GSVA scores:
dna_dmg_kw <- c("DNA_DAMAGE","LYSO")
dna_dmg <- gsva_res.df.t %>% dplyr::select(contains(all_of(dna_dmg_kw))) %>% t() # just the GSVA scores for DEnA
dna_dmg_df <- gsva_res.df.t %>%
  dplyr::select(Aliquot.ID, contains(all_of(dna_dmg_kw))) %>% 
  inner_join(mut_data, by="Aliquot.ID")


## =============== DEnA with limma: GSVA scores =============== ##
## Define statistical threshold:
PvalueCutoff2 <- 0.05 # significance threshold

## Define design matrix:
# design2 <- model.matrix(~ 0 + spls$EGFRvIII) 
design2 <- model.matrix(~ 0 + mut_vIII$EGFRvIII) 
colnames(design2) <- c("WT","EGFRvIII") # check that these are in the correct order
# rownames(design2) <- spls$Aliquot.ID
rownames(design2) <- mut_vIII$case_id %>% match(spls$case_id) %>% spls$Aliquot.ID[.]

## Perform DEA:
fit2 <- lmFit(gsva_res[,rownames(design2)], design2)
# fit2 <- lmFit(dna_dmg[,rownames(design2)], design2)
contr2 <- makeContrasts(HIvsLOW = EGFRvIII - WT, # Make contrasts --> compare groups of interest
                        levels = colnames(design2))
tmp2 <- contrasts.fit(fit2, contr2)
tmp2 <- eBayes(tmp2)
allGeneSets <- topTable(tmp2, sort.by = "P", number=Inf) %>% 
  mutate(ID = rownames(.),
         collection = sub("\\_.*", "", ID)
         )
DEgeneSets <- topTable(tmp2, sort.by = "P", number=Inf,
 p.value=PvalueCutoff2, adjust="fdr") %>% 
  mutate(ID=rownames(.))
res2 <- decideTests(tmp2, p.value=PvalueCutoff2)
summary(res2)
# DEgeneSets
```

## DNA damage and other comparisons across vIII versus WT EGFR tumors
```{r Plot DEnA results for comparisons of interest}
comps_of_interest <- c("DNA_DAMAGE","PINOCYTOSIS","LYSO","ENDOSOME","ENDOCYTOSIS","PTEN","GOCC","JNK","P38","STRESS","PTPN11|SHP2")

for(i in 1:length(comps_of_interest)){
  comp_kw <- comps_of_interest[i]
  lblsz <- 3
  if(comp_kw=="STRESS"){lblsz <- 1.75}
  comp_plot <- allGeneSets %>% 
    filter(str_detect(ID, comp_kw)) %>% 
    ggplot(aes(x=logFC, y=-log10(P.Value), color=logFC>0, label=ID)) +
      geom_hline(yintercept=-log10(PvalueCutoff2), linetype=8) + geom_vline(xintercept=0, linetype=8) +
      geom_point(alpha=0.7) + 
      geom_label_repel(
        data=function(x){subset(x, P.Value<PvalueCutoff2)}, label.size=0.15, label.padding=0.15,
        size=lblsz, min.segment.length=0, force=10) + 
      # scale_color_manual(values=c("grey60","red")) +
      scale_color_manual(values=julia_cols) +
      theme_cowplot() + theme_cust() +
      labs(title=paste0("CPTAC GBM, DEnA: EGFRvIII vs. WT\nKeyword: ",comp_kw)) + 
      guides(color="none")
  
  show(comp_plot)
  
  if(comp_kw=="PTPN11|SHP2"){comp_kw <- "PTPN11-SHP2"}
  #ggsave(paste("DEnA-volcanoplot_EGFRvIII-vs-WT_",comp_kw,".png",sep=""), comp_plot)
}

```

```{r Barplots of up-/down-regulated protein sets}
barplot_kw <- "STRESS"
## Down-regulated in vIII mutants:
allGeneSets %>% 
  mutate(sign=logFC<0) %>% 
  filter(str_detect(ID, barplot_kw) & P.Value<PvalueCutoff2 & logFC<0) %>% 
  arrange(logFC) %>% 
  head(20) %>% 
  arrange(desc(logFC)) %>% 
  mutate(ID=factor(ID,levels=ID)) %>%
  ggplot(aes(x=logFC, y=ID, fill=-log10(P.Value), color=-log10(P.Value))) +
    geom_col(color="black") +
    # facet_wrap(~sign, scales="free") + 
    scale_fill_viridis_c() + scale_color_viridis_c() +
    theme_minimal_grid(10) + 
    labs(title="CPTAC GBM, DEnA: EGFRvIII vs. WT")

## Up-regulated in vIII mutants:
allGeneSets %>% 
  mutate(sign=logFC<0) %>% 
  filter(str_detect(ID, barplot_kw) & P.Value<PvalueCutoff2 & logFC>0) %>% 
  arrange(desc(logFC)) %>% 
  head(20) %>% 
  arrange(logFC) %>% 
  mutate(ID=factor(ID,levels=ID)) %>%
  ggplot(aes(x=logFC, y=ID, fill=-log10(P.Value), color=-log10(P.Value))) +
    geom_col(color="black") +
    # facet_wrap(~sign, scales="free") + 
    scale_fill_viridis_c() + scale_color_viridis_c() +
    theme_minimal_grid(10) + 
    labs(title="CPTAC GBM, DEnA: EGFRvIII vs. WT")

```

# PTM-SVA

The [PTMSigDB](http://prot-shiny-vm.broadinstitute.org:3838/ptmsigdb-app/) "is a database comprised of modification site-specific signatures of perturbations, kinase activities and signaling pathways curated from more than 2,500 publications which provides the foundation to perform PTM-SEA." We now load in those signatures and format them for use with the phosphoprotein data to perform PTM set variation analysis (PTM-SVA -> using GSVA instead of ssGSEA as the base algorithm).
```{r Load PTMsigDB signatures}
## PTMsigDB collections:
fn.ptmsigs <- "data_PTMsigDB_all_sites_v1.9.0.xlsx"
ptmsigs <- readxl::read_xlsx(fn.ptmsigs) %>% 
  group_by(signature) %>% 
  mutate(sig_dir = paste(signature, site.direction, sep="_"),
         site.annotation = site.annotation %>% gsub(":.*","",.)
         ) %>% 
  # filter(category %>% str_detect("PATH|KINASE")) %>% # just the kinase and pathway signatures and Src
  # filter(category %>% str_detect("KINASE")) %>% # just the kinase and pathway signatures
  # filter(signature %>% str_detect("Pathway|PATHWAY|pathway")) %>% # just the pathway signatures
  filter(!(signature %>% str_detect("Pancreatic|cancer|Cancer|PERT|DISEASE|INFECTION"))) # get rid of cancer-specific signatures
ptm_sets <- ptmsigs %>% split(x = .$site.annotation, f = .$sig_dir)
```

Post-translational modification set enrichment analyssis (PTM-SEA) is described by [Krug et al. (2019)](http://doi.org/10.1074/mcp.TIR118.000943). Here, we use our custom approach, which follows the same general steps as Krug et al. except that we use the `GSVA` package to calculate the per-sample enrichment scores for the up/down directions of each PTM signature. We then use our custom function defined above to calculate a single enrichment score for each signature based on the difference between the scores for the up/down directions.

```{r PTM-SVA calculations}
library(GSVA)
## Format phosphosite data for gsva function:
phos3 <- phos1[,-1] %>%
  ungroup() %>% 
  dplyr::select(-c(Phosphosite)) %>% 
  data.matrix()
rownames(phos3) <- phos1$Phosphosite

## Calculate single-sample phosphosite-specific signature scores:
phos_ssgsea <- gsva(
  phos3, 
  ptm_sets, verbose=F,
  # method = "ssgsea", tau = 0.75,
  method = "gsva", #mx.diff = F, abs.ranking = T,
  min.sz = 1, # Minimum number of sites required to include a signature
  parallel.sz=detectCores()#-1
  )
phos_ssgsea.df <- phos_ssgsea %>%
  data.frame(sig_dir=rownames(.)) %>% 
  group_by(sig_dir) %>%
  mutate(
    dir = match(sig_dir, ptmsigs$sig_dir) %>% ptmsigs$site.direction[.] %>% unique()
    ) %>% 
  separate(sig_dir, into="sig", sep="_d$|_u$",remove=T)

## Calculate PTM-SEA enrichment scores:
use_ud_diff <- F # take difference between up/down directions of signatures?
if(use_ud_diff){
  ptmsva_res <- ptmsea_func(phos_ssgsea.df, output="matrix")
  ptmsva_res.df <- ptmsea_func(phos_ssgsea.df, output="df")
  ptmsva_res.df.t <- ptmsva_res %>% t() %>% data.frame(Aliquot.ID=rownames(.))
} else {
  ptmsva_res <- phos_ssgsea
  ptmsva_res.df <- ptmsva_res %>% data.frame(sig_dir=rownames(.))
  ptmsva_res.df.t <- ptmsva_res %>% t() %>% data.frame(Aliquot.ID=rownames(.))
}
```

## PTM-SVA DEnA
```{r PTM-SVA DEnA calculation}
library(limma)
## Get DNA damage-related GSVA scores:
stress_kw <- c("JNK","P38","PI3K","Akt","AKT","ASK1","MAP3K8")
stress_sig <- ptmsva_res.df.t %>% dplyr::select(contains(all_of(stress_kw))) %>% t()
stress_sig_df <- ptmsva_res.df.t %>%
  dplyr::select(Aliquot.ID, contains(all_of(stress_kw))) %>% 
  inner_join(mut_data, by="Aliquot.ID")


## =============== DEnA with limma: GSVA scores =============== ##
## Define statistical threshold:
PvalueCutoff3 <- 0.05 # significance threshold

## Define design matrix:
# design3 <- model.matrix(~ 0 + spls$EGFRvIII) 
design3 <- model.matrix(~ 0 + mut_vIII$EGFRvIII) 
colnames(design3) <- c("WT","EGFRvIII") # check that these are in the correct order
# rownames(design3) <- spls$case_id
rownames(design3) <- mut_vIII$case_id


## Perform DEA:
fit3 <- lmFit(ptmsva_res[,rownames(design3)], design3)
# fit3 <- lmFit(dna_dmg[,rownames(design3)], design3)
contr3 <- makeContrasts(HIvsLOW = EGFRvIII - WT, # Make contrasts --> compare groups of interest
                        levels = colnames(design3))
tmp3 <- contrasts.fit(fit3, contr3)
tmp3 <- eBayes(tmp3)
allGeneSets2 <- topTable(tmp3, sort.by = "P", number=Inf) %>% 
  mutate(ID = rownames(.),
         collection = sub("\\_.*", "", ID)
         )
DEgeneSets2 <- topTable(tmp3, sort.by = "P", number=Inf,
 p.value=PvalueCutoff3, adjust="fdr") %>% 
  mutate(ID=rownames(.))
res3 <- decideTests(tmp3, p.value=PvalueCutoff3)
summary(res3)
# DEgeneSets2

linesz <- 0.8
allGeneSets2 %>%
  filter(P.Value<PvalueCutoff3) %>% 
  arrange(logFC) %>% 
  mutate(ID=factor(ID,levels=ID)) %>% 
  ggplot(aes(x=logFC, y=ID, fill=-log10(P.Value))) +
    geom_vline(xintercept=0, size=linesz) +
    geom_col(color="black", size=linesz) + 
    scale_fill_viridis_c(option="inferno") +
    theme_minimal_grid(8) +
    labs(title="CPTAC GBM, PTM set DEnA\nEGFRvIII vs. WT", y="PTM set", x="PTM-SVA score logFC") +
    guides(color="none")
#ggsave("PTM-SVA_DEnA-barplot_all-signif-res.png")
```

```{r Stress/MAPK-related PTM sets}
lblsz <- 4
linesz <- 0.8
stress_kw <- c("JNK","P38","PI3K","Akt","AKT","ASK1","MAP3K8","Cell_Cycle","RAS","Ras","MAPK","MAP[0-9]K")


allGeneSets2 %>%
  filter(str_detect(ID, paste(stress_kw, collapse="|"))) %>%
  arrange(logFC) %>% 
  mutate(ID=factor(ID,levels=ID)) %>% 
  ggplot(aes(x=logFC, y=ID, fill=P.Value<PvalueCutoff3)) +
  # ggplot(aes(x=logFC, y=ID, fill=-log10(P.Value))) +
    geom_vline(xintercept=0, size=linesz) +
    geom_col(color="black", size=linesz) + 
    scale_fill_manual(values=julia_cols) +
    # theme_cowplot(line_size=linesz) + theme_cust() +
    theme_minimal_grid(10) +
    labs(title="CPTAC GBM, PTM set DEnA\nEGFRvIII vs. WT", y="PTM set", x="PTM-SVA score logFC", fill="p < 0.05") +
    guides(color="none")
#ggsave("PTM-SVA_DEnA-barplot_MAPK-RAS.png")


```
